---
title: "Session 1 -- Introduction to RStudio"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Session 1 -- Introduction to RStudio}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

> #### Learning objectives
>
> * Describe the main panes in RStudio
> * Generate and edit an R Markdown report
> * Import tabular data into R
> * Inspect data for missing values


# Introduction

This is a 'next-steps' course, designed for follow-on practice with cancer datasets after completing DataCamp's [Introduction to the Tidyverse](https://www.datacamp.com/courses/introduction-to-the-tidyverse) (or another general introductory R course). Therefore, we assume you know some R basics already, at least the content of the DataCamp course. [R for Data Science book](https://r4ds.had.co.nz/index.html) is an excellent reference book if you want to look up anything you don't recall or are not very familiar with. We also assume you know some reasons why learning R is good (as you've already started `r emo::ji("smile")`) but below are a list of reasons why learning R is useful and worth your time. 

## Why learn R?

* R involves creating & using code documents which makes the steps you used in your analysis clear and can be inspected by someone else for feedback and error-checking.

* R produces high-quality graphics suitable for publication in journals or the web.

* R is interdisciplinary and extensible and has thousands of installable packages to extend its capabilities. R has packages for image analysis, GIS, time series, population genetics, and a lot more.

* R scales well to work on data of all shapes and sizes.

* R can connect to spreadsheets, databases, and many other data formats, on your computer or on the web.

* R has a large and welcoming community - Thousands use R daily and many of them are willing to help you through websites such as [Stack Overflow](https://stackoverflow.com/) or the [RStudio community](https://community.rstudio.com/).

* Not only is R free, but it is also open-source and cross-platform.

* R code is great for reproducibility. An increasing number of journals and funding agencies expect analyses to be reproducible, so knowing R will give you an edge with these requirements.

* R integrates with other tools to generate manuscripts from your code. This document ([R Markdown](https://rmarkdown.rstudio.com/), a .Rmd file) is a case in point. 

```{block type = "rmdblock"}
We will be using R Markdown for the assignments in this course to help you learn how to use it. Information on creating R Markdown manuscripts for journals can be found [here](https://bookdown.org/yihui/rmarkdown/journals.html#ref-R-rticles) and in [this paper](https://www.sciencedirect.com/science/article/pii/S2667145X21000171). Some examples of R Markdown in action:  
  * [PhD thesis](https://sa-lee.github.io/thesis/preface.html)  
  * [Figures](https://atlassian.petermac.org.au/bitbucket/projects/DAW/repos/multisamplevariantcalling/browse/plotting/PaperFigures.Rmd) for a Peter Mac bioinformatics paper  
  * [Analysis results](http://oshlacklab.com/combes-organoid-paper/) using [workflowr](https://workflowr.github.io/workflowr/index.html) (used by a few groups at Peter Mac to share/collaborate)
```

---

# RStudio  a brief tour

RStudio provides us with a friendly interface to the R statistical programming language.

```{r, echo=FALSE, out.width = "100%"}
knitr::include_graphics("images/Rstudioataglance.png")
```

It consists of four main "Panes":

* the Editing Pane (Source) for your scripts and documents (top-left, in the default layout)
* the R Console (bottom-left).
* your Environment/History (top-right),
* your Files/Plots/Packages/Help/Viewer (bottom-right)

These can be re-sized and moved around to suit how
you like to work (see menu, Tools -> Global Options -> Pane Layout)

One of the advantages of using RStudio is that all the information you need to write code is available in a single window. Additionally, with many shortcuts, autocompletion, and highlighting for the major file types you use while developing in R, RStudio will make typing easier and less error-prone.


## Editing pane

By default the top left-hand pane is one for creating, editing & running R files e.g. scripts and reports.

```{r, echo=FALSE, out.width = "100%"}
knitr::include_graphics("images/Rstudio_scripting.png")
```

A script is an R program that you have written. A good practice is for that script to perform only one role in your analysis workflow and so you may have several R scripts which you call, in a particular sequence, to analyse your data. A script is basically a text file that contains R commands and (ideally) comments to explain what the codes function is (as a documentation process).

As well as R scripts, there are many types of RStudio document including [R Markdown files](https://r4ds.had.co.nz/r-markdown.html). We will use R Markdown in the teaching of this course. This is so you can learn how to generate reports summarising your analysis for collaborating or sharing your work with others.


## Console

Coming down the screen to the bottom left-hand pane we find the console window. This is where we can find output produced by running our R scripts.


```{r, echo=FALSE, out.width = "100%"}
knitr::include_graphics("images/Rstudio_console.png")
```


We can also try out snippets of R code here. Those of you who have only used graphical interfaces like Windows or MacOS where you click on commands using a mouse may find this aspect of R somewhat different. We type in commands to R using the command line.

This area can also be used like a calculator. Let's just type in something like `23 + 45` followed by the return key and see what happens. You should get the following:

```
> 23 + 45
[1] 68
```

Now 68 is clearly the answer but what is that 1 in brackets?

Here is another example to explain. If we type `1:36` and press enter, what happens?
R generates output counting from 1 to 36 but cannot fit all the output on one line and so starts another like this:

```
> 1:36
 [1]  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25
[26] 26 27 28 29 30 31 32 33 34 35 36
```

Now we have two lines beginning with a number in square brackets. Note that the
number of values displayed on each line may differ on your computer; it largely
depends on the width of your console pane and the font size. Try creating a
larger sequence of numbers, e.g. `1:100`, if all 36 numbers fit on a single line
in your case.

This is just R helping us to keep tabs on which number we are looking at. `[1]` denotes that the line starts with the first result and the `[26]` denotes that this line starts with the 26th number. Let's try another one and generate a sequence incrementing in steps of 2:

```
> 1:36 * 2
 [1]  2  4  6  8 10 12 14 16 18 20 22 24 26 28 30 32 34 36 38 40 42 44 46 48 50
[26] 52 54 56 58 60 62 64 66 68 70 72
```

There are other tabs on this pane but we shall not be covering these on this course.

## Environment

Next we move to the top right-hand corner pane. Here we have even more tabs (of which we will only consider two Environment and History).


```{r, echo=FALSE, out.width = "100%"}
knitr::include_graphics("images/Rstudio_env.png")
```


Environment keeps track on R variables which we create (more on those later) and their contents. History is like a tally role of all the R commands we have entered in our session.

## Files, Plots & Help

Our final bottom right-hand pane also has several tabs. The Files tab is a file explorer that enable us to move around our directories and select which files we wish to work on. We can also change the default working directory that Rstudio will use.


```{r, echo=FALSE, out.width = "100%"}
knitr::include_graphics("images/Rstudio_filexpl.png")
```


The Plots tab is where any graphs that we create in R will appear. We can move through them using the arrow buttons and the export button will convert them to different graphics formats e.g. for publication in a paper or for the web.

The Packages tab shows which R packages are installed (These expand R's functionality and again will be covered later) and can also install new packages.

The Help tab is a _massively useful_ tab which enables you to search R help index to get help pages on R functions and provide example code to help you use them in your R scripts.

The Viewer tab is where we can preview an R markdown report after we Knit, more on that shortly. 
---

# Our first look at an R Markdown report

Our overall goal for this course is to give you the ability to import your data into R, format it, select a subset
of the data most of interest for a given analysis, carry out an analysis to summarize these data,
create visualizations of the data, and generate a reproducible report.

So let's see an example of what we'll be learning and generate our first visualisation and report. By the end of this course we should be able to explain what each line of the code does. 

```{r echo = FALSE}
xfun::embed_file('../inst/example-report.Rmd')
```

 * Download the example report (Rmd) file from the link above
 * In your RStudio top menu go to File > Open File and open the Rmd file 
 * Click the Knit button and inspect the report in the Viewer pane (see below)
 * Try out the exercise in the Rmd file to edit the file. Knit again to see the changes in the report.
 
```{r, echo=FALSE, out.width = "100%"}
knitr::include_graphics("images/rmarkdown.png")
```

As described in the [R for Data Science book](https://r4ds.had.co.nz/r-markdown.html), an R Markdown, or .Rmd, document contains three types of content:

* An (optional) YAML header surrounded by `---`.
* Chunks of R code surrounded by ```` ``` ````.
* Text mixed with simple text formatting like `# heading` and `**bold**`.

When you open an .Rmd, you get a notebook interface where code and output are interleaved. You can run each code chunk by clicking the Run icon (it looks like a play button at the top of the chunk), or by pressing Cmd/Ctrl + Shift + Enter. RStudio executes the code and displays the results inline with the code. When you knit the document, R Markdown executes all of the code chunks and creates a document (whichever you specify: HTML/PDF/Word) which includes the code and its output. Here a HTML file will be generated in the folder where you have the example_report.Rmd file. To create an R Markdown file in RStudio you can use File > New File > R Markdown.

Next we'll play around with the example R Markdown file some more and explain some of the things you can see in it.

## Packages

There are thousands of R packages that extend R's core functionality to provide
specialized functions for doing all kinds of things. There are R packages that
provide various kinds of visualizations and plotting functions, others that let
us manipulate data in interesting and useful ways, and still others that give us
access to cutting-edge machine learning algorithms. Several packages have been
developed by bioinformaticians at Peter Mac, mainly focusing on
genomic and transcriptomic analysis methods.

You need to load a package using the **`library()`** function in order to be
able to use the functions and data sets that the package provides. We can load the core tidyverse packages in one go, as you can see in the example_report.Rmd, as follows.

```{r}
library(tidyverse)
```

The output from this command tells us which packages were loaded, one of which is `ggplot2` that provides excellent
plotting functions for visualizing our data and will be the main focus next
week.

We also got a warning about some conflicts. This tells us that there are
functions loaded as part of the `dplyr` package that have the same name as
functions that are in the `stats` package. The `stats` package was already
loaded, from when we first started our R session, and the functions referred to
in the warning are now masked; to use these instead of their `dplyr` namesakes
you need to specify them fully by adding the package name and `::` as shown in
the warning message.

### Installing packages

If you got the following error message when trying to load the tidyverse, then
you haven't yet installed it.

```
> library(tidyverse)
Error in library(tidyverse) : there is no package called ‘tidyverse’
```

Installing packages should be relatively straightforward using the
`install.packages()` function. Note that you only need to install a package once on your computer (unless you want to install a newer version), then you use `library()` to load it every time you start an R session.

```{r eval = FALSE}
install.packages("tidyverse")
```

If you're using R at a regular console within a terminal window, i.e. not
RStudio, you may be asked to select one of several mirrors of the CRAN R package
repository hosted at different locations around the world.

For some packages you may be asked to choose between a binary and a source
package, particularly if the latest version of the package is only available in
source form and that package contains code written in another language such as
C. This could be a problem if you don't have a C compiler installed on your
computer.

---

# Importing data into R

Another of the tidyverse packages loaded when we ran `library(tidyverse)` is the
`readr` package, which provides functions for reading tabular data into a tibble
and writing tibbles to tab-delimited and comma-separated value (CSV) files. A [tibble](https://r4ds.had.co.nz/tibbles.html) is tidyverse's table format, a modern version of R's data frame.

For example, the clinical data for the METABRIC breast cancer dataset generated
at CRUK CI can be downloaded as a tab-delimited file from
[cBioPortal](https://www.cbioportal.org/study/clinicalData?id=brca_metabric).
Targeted sequencing was performed for 2509 primary breast tumours, along with
548 matched normals, using a panel of 173 of the most frequently mutated breast
cancer genes as part of the METABRIC study
([Pereira *et al.*, Nature Communications 7:11479, 2016](https://www.ncbi.nlm.nih.gov/pubmed/27161491)).

This file contains information in tabular form in which values are separated by
tab characters so we'll use the **`read_tsv()`** to read these data.

Create a new chunk in the Rmd file you downloaded above. You can insert an R code chunk either using the RStudio toolbar (the Insert button) or the keyboard shortcut Ctrl + Alt + I (Cmd + Option + I on macOS). Copy/paste the line of code below into the chunk and then click the chunk’s Run button.

```{r}
metabric <- read_tsv("https://zenodo.org/record/6450144/files/brca_metabric_clinical_data.tsv")
```

`read_tsv()` prints a message telling us which types it has assigned to each of
the columns it has read, although in this case the table has quite a lot of
columns and it only outputs details about the first few.

To try out the rest of the code shown in this document below, for each line or code section, insert a chunk, copy/paste or type the commands and Run.

Run the line below to print the clinical data set. We can see that `read_tsv()` created a tibble.

```{r}
metabric
```

```{block type = "rmdblock"}
**`read_csv()`** and **`read_tsv()`**
  
`read_csv()` reads data from a comma-separated value (CSV) file into a tibble.
`read_tsv()` is the equivalent function that works on tab-delimited files.

These functions expect the first line to contain column names and try to make
sensible guesses at the type of data in each column. You can change this by
specifying various arguments, e.g. to skip comment lines beginning with a
specific character (e.g. "#") or to tell the function what the column types
are.

patients <- read_csv("patients.csv")

To export a CSV or TSV file from R you can use `write_csv()` or `write_tsv()`.
```

We are now in a position to explore these data. We can see all column names with R's built-in **`colnames()`**.

```{r}
colnames(metabric)
```

We can access individual columns with `table_name$column_name`, for example `metabric$Chemotherapy`

```{r}
metabric$Chemotherapy
```

We can use R's built-in **`table()`** function to count the number of patients in this cohort that had
chemotherapy and those that did not.

```{r}
table(metabric$Chemotherapy)
```

We can summarize the numbers of mutations found per patient using R's built-in **`summary()`** function.

```{r}
summary(metabric$`Mutation Count`)
```

Note that several of the column names have multiple words separated by spaces.
We have to use backticks ( **`** ) to refer to column names such as these. Tip: if you have lots of columns with spaces or special characters you could use the popular janitor package's [clean_names](https://albert-rapp.de/post/2022-01-12-janitor-showcase/) function to format the column names in a nicer way.

We can create a scatter plot to see if there is any relationship between the age
of the patient at diagnosis of breast cancer and the Nottingham prognostic index
(whatever that is) using R's built-in **`plot()`** function.

```{r metabric_plot}
plot(metabric$`Age at Diagnosis`, metabric$`Nottingham prognostic index`,
     xlab = "Age at diagnosis", ylab = "Nottingham prognostic index")
```

Hmm, what a horrible looking plot. Thank goodness we're moving on to
**`ggplot2`** next week.

---

# Missing values

The METABRIC clinical dataset has some missing values. If you look back to the
table we printed out earlier you can see 2 missing values for mutation counts
within the first 10 rows. We can specify rows of a column to view with **`[firstrow:lastrow]`**.

```{r}
metabric$`Mutation Count`[1:10]
```

```{block type = "rmdblock"}
**Missing values** (**`NA`**)

Missing values in R are represented as **`NA`**, which stands for 'not available'.
```

The `summary()` function reported that there are 152 missing mutation counts.
`summary()` reports the mean and median values for the remaining observations,
i.e. those for which a mutation count *is* available. Not all functions do this
though.

```{r}
mean(metabric$`Mutation Count`)
```

The `mean()` function, for example, takes a different approach in how it treats
missing values. It returns `NA`, i.e. a missing value, because it takes the view
that it cannot compute a mean for a set of values where some are not available.
You can specify the argument `na.rm = TRUE` when calling `mean()` to instruct it
to ignore the missing values and calculate the mean of the remaining values.

```{r}
mean(metabric$`Mutation Count`, na.rm = TRUE)
```

There are many such functions in R that would be similarly affected by missing
values and which offer the `na.rm` argument.

You can test for missing values using the `is.na()` function.

```{r}
first_ten_mutation_counts <- metabric$`Mutation Count`[1:10]
is.na(first_ten_mutation_counts)
```

One final handy tip: summing the result of `is.na()` will count the number of missing values in a vector. If you don't know what a vector is, [see here](https://r4ds.had.co.nz/vectors.html).

```{r}
sum(is.na(metabric$`Mutation Count`))
```

Can you see why this works? *(hint: look at the output of `is.na(metabric$`Mutation Count`)`, use the help page which you can get by typing '`?sum` to check what kind of vector `sum()` expects
and how it handles logical values)*.

## Working directory

By default, R looks for (and saves) files in the working directory. When importing the METABRIC dataset we specified where the file was located using a URL `https://zenodo.org/record/6450144/files/brca_metabric_clinical_data.tsv`
as it's located on the web. If you have a file on your computer you can specify the path to where it is located. For example, if we paste the URL above into our web browser to download the file we can then import it from our computer using the absolute path, which starts with `/`.

```{r eval = FALSE}
metabric_downloaded <- read_tsv("/home/maria/Downloads/brca_metabric_clinical_data.tsv")
```

Or if we were working inside the Downloads directory we could use the relative path.

```{r eval = FALSE}
metabric_downloaded <- read_tsv("brca_metabric_clinical_data.tsv")
```

The **`getwd()`** function will tell you what the current working
directory is.

```{r eval = FALSE}
getwd()
```

If you open RStudio by clicking on its icon in the task bar or its entry in the
start menu, your working directory will probably be your home folder. You can explicitly set your working directory using **`setwd("/path/to/your/working/directory")`** (or in the RStudio Files pane More > Set As Working Directory). However, the path will be specific to your computer and won't work if you have it in a script that you share with someone else.

Alternatively (and better) you can create an RStudio Project as described in more detail [here](https://r4ds.had.co.nz/workflow-projects.html#rstudio-projects) to keep track of and set the working directory.

Create an RStudio project for this course (File > New Project > New Directory > Project Type: New Project > Name: r-skills-course) and store your files (assignment Rmds etc) from the course in that r-skills-course folder. When you want to open the project, in RStudio menu you can use File > Open Project or click on the .Rproj file. This will automatically set the working directory to where the .Rproj file is located.

---

# Summary

In this session we have covered the following concepts:

* The RStudio environment and the functions associated with each pane
* Generating a visualisation and R Markdown report
* How to read tabular data into R
* Missing values and how R handles these


# Assignment

This week's assignment and the subsequent weeks' assignments are in R Markdown files that you can use to practice working with reports.

```{r echo = FALSE}
xfun::embed_file('../assignments/assignment1.Rmd')
```