<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Session 4 -- Grouping, combining and restructuring data • R4CancerSci</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Session 4 -- Grouping, combining and restructuring data">
<meta property="og:description" content="R4CancerSci">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">R4CancerSci</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Material
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/week0.html">Week 0 - Getting ready</a>
    </li>
    <li>
      <a href="../articles/week1.html">Week 1 - Introduction to RStudio</a>
    </li>
    <li>
      <a href="../articles/week2.html">Week 2 - Data visualization with ggplot2</a>
    </li>
    <li>
      <a href="../articles/week3.html">Week 3 - Data manipulation using dplyr</a>
    </li>
    <li>
      <a href="../articles/week4.html">Week 4 - Grouping, combining and restructuring data for analysis</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/pmacdasci/R4CancerSci" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Session 4 – Grouping, combining and
restructuring data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/R4CancerSci/R4CancerSci/blob/HEAD/vignettes/week4.Rmd" class="external-link"><code>vignettes/week4.Rmd</code></a></small>
      <div class="hidden name"><code>week4.Rmd</code></div>

    </div>

    
    
<blockquote>
<h4 id="learning-objectives">Learning objectives</h4>
<ul>
<li>Use <code><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by()</a></code> with <code><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise()</a></code> to compute
summary values for groups of observations</li>
<li>Use <code><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count()</a></code> to count the numbers of observations within
categories</li>
<li>Combine data from two tables based on a common identifier
(<code>join</code> operations)</li>
<li>Understand what makes a data set ‘tidy’ and why you’d want your data
to be structured this way</li>
<li>Use <code><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer()</a></code> and <code><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html" class="external-link">pivot_wider()</a></code>
operations to restructure data frames</li>
<li>Tease apart columns containing multiple variables using
<code><a href="https://tidyr.tidyverse.org/reference/separate.html" class="external-link">separate()</a></code>
</li>
<li>Modify character variables using string manipulation functions from
the stringr package</li>
<li>Customize plots created using ggplot2 by changing labels, scales and
colours and themes</li>
</ul>
</blockquote>
<hr>
<div class="rmdblock">
<p>
A video that goes through this session’s material can be found <a href="https://www.youtube.com/watch?v=6IOxhJIbiA8" class="external-link">here</a>.
</p>
</div>
<div class="section level2">
<h2 id="grouping-and-combining-data">Grouping and combining data<a class="anchor" aria-label="anchor" href="#grouping-and-combining-data"></a>
</h2>
<p>In this session, we’ll look at some more useful functions provided by
the <strong>dplyr</strong> package, the ‘workhorse’ in the tidyverse
family for manipulating tabular data. Continuing from last week, we’ll
see how we can summarise data for groups of observations within
different categories. We’ll also show how dplyr allows us to combine
data for the same observational unit, e.g. person or date, that comes
from different sources and is read into R in different tables.</p>
<p>We’ll also look at how to customize the plots we create using
<strong>ggplot2</strong>, in particular how we can add or change titles
and labels, how we can adjust the way the axes are displayed and how we
can use a colour scheme of our choosing.</p>
<p><strong>dplyr</strong> and <strong>ggplot2</strong> are core
component packages within the tidyverse and both get loaded as part of
the tidyverse.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## ── <span style="font-weight: bold;">Attaching packages</span> ─────────────────────────────────────── tidyverse 1.3.1 ──</span></code></pre>
<pre><code><span class="co">## <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">ggplot2</span> 3.3.6     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">purrr  </span> 0.3.4</span>
<span class="co">## <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">tibble </span> 3.1.7     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">dplyr  </span> 1.0.9</span>
<span class="co">## <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">tidyr  </span> 1.2.0     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">stringr</span> 1.4.0</span>
<span class="co">## <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">readr  </span> 2.1.2     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">forcats</span> 0.5.1</span></code></pre>
<pre><code><span class="co">## ── <span style="font-weight: bold;">Conflicts</span> ────────────────────────────────────────── tidyverse_conflicts() ──</span>
<span class="co">## <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">dplyr</span>::<span style="color: #00BB00;">filter()</span> masks <span style="color: #0000BB;">stats</span>::filter()</span>
<span class="co">## <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">dplyr</span>::<span style="color: #00BB00;">lag()</span>    masks <span style="color: #0000BB;">stats</span>::lag()</span></code></pre>
<p>To demonstrate how these grouping and combining functions work and to
illustrate customization of plots, we’ll again use the METABRIC data
set.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">metabric</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="st">"https://zenodo.org/record/6450144/files/metabric_clinical_and_expression_data.csv"</span><span class="op">)</span>
<span class="va">metabric</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 1,904 × 32</span></span>
<span class="co">##    Patient_ID Cohort Age_at_diagnosis Survival_time Survival_status Vital_status</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       </span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> MB-0000         1             75.6         140.  LIVING          Living      </span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> MB-0002         1             43.2          84.6 LIVING          Living      </span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> MB-0005         1             48.9         164.  DECEASED        Died of Dis…</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> MB-0006         1             47.7         165.  LIVING          Living      </span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> MB-0008         1             77.0          41.4 DECEASED        Died of Dis…</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> MB-0010         1             78.8           7.8 DECEASED        Died of Dis…</span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> MB-0014         1             56.4         164.  LIVING          Living      </span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> MB-0022         1             89.1          99.5 DECEASED        Died of Oth…</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> MB-0028         1             86.4          36.6 DECEASED        Died of Oth…</span>
<span class="co">## <span style="color: #BCBCBC;">10</span> MB-0035         1             84.2          36.3 DECEASED        Died of Dis…</span>
<span class="co">## <span style="color: #949494;"># … with 1,894 more rows, and 26 more variables: Chemotherapy &lt;chr&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   Radiotherapy &lt;chr&gt;, Tumour_size &lt;dbl&gt;, Tumour_stage &lt;dbl&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   Neoplasm_histologic_grade &lt;dbl&gt;, Lymph_nodes_examined_positive &lt;dbl&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   Lymph_node_status &lt;dbl&gt;, Cancer_type &lt;chr&gt;, ER_status &lt;chr&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   PR_status &lt;chr&gt;, HER2_status &lt;chr&gt;, HER2_status_measured_by_SNP6 &lt;chr&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   PAM50 &lt;chr&gt;, `3-gene_classifier` &lt;chr&gt;, Nottingham_prognostic_index &lt;dbl&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   Cellularity &lt;chr&gt;, Integrative_cluster &lt;chr&gt;, Mutation_count &lt;dbl&gt;, …</span></span></code></pre>
<hr>
</div>
<div class="section level2">
<h2 id="grouping-observations">Grouping observations<a class="anchor" aria-label="anchor" href="#grouping-observations"></a>
</h2>
<div class="section level3">
<h3 id="summaries-for-groups">Summaries for groups<a class="anchor" aria-label="anchor" href="#summaries-for-groups"></a>
</h3>
<p>In the previous session we introduced the <code><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise()</a></code>
function for computing a summary value for one or more variables from
all rows in a table (data frame or tibble). For example, we computed the
mean expression of ESR1, the estrogen receptor alpha gene, as
follows.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span><span class="va">metabric</span>, <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">ESR1</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 1 × 1</span></span>
<span class="co">##   `mean(ESR1)`</span>
<span class="co">##          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span>         9.61</span></code></pre>
<p>While the <code><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise()</a></code> function is useful on its own, it
becomes really powerful when applied to groups of observations within a
dataset. For example, we might be more interested in the mean ESR1
expression calculated separately for ER positive and ER negative
tumours. We could take each group in turn, filter the data frame to
contain only the rows for a given ER status, then apply the
<code><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise()</a></code> function to compute the mean expression, but
that would be somewhat cumbersome. Even more so if we chose to do this
for a categorical variable with more than two states, e.g. for each of
the integrative clusters. Fortunately, the
<strong><code><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by()</a></code></strong> function allows this to be done
in one simple step.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">metabric</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">ER_status</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">ESR1</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 2 × 2</span></span>
<span class="co">##   ER_status `mean(ESR1)`</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Negative          6.21</span>
<span class="co">## <span style="color: #BCBCBC;">2</span> Positive         10.6</span></code></pre>
<p>We get an additional column in our output for the categorical
variable, <code>ER_status</code>, and a row for each category.</p>
<p>Incidentally, we should expect this result of ER-positive tumours
having a higher expression of ESR1 on average than ER-negative tumours.
Simple summaries like this are a good way of checking that what we think
we know actually holds true in the data we’re looking at. Note that the
expression values are on a log<sub>2</sub>scale so ER-positive breast
cancers express ESR1 at a level that is approximately 20 times greater,
on average, than that of ER-negative tumours.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fl">2</span> <span class="op">**</span> <span class="op">(</span><span class="fl">10.6</span> <span class="op">-</span> <span class="fl">6.21</span><span class="op">)</span>  <span class="co"># equivalent to (2 ** 10.6) / (2 ** 6.21)</span></code></pre></div>
<pre><code><span class="co">## [1] 20.96629</span></code></pre>
<p>Let’s have a look at how ESR1 expression varies between the
integrative cluster subtypes defined by the METABRIC study.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">metabric</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">Integrative_cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>ESR1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">ESR1</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 11 × 2</span></span>
<span class="co">##    Integrative_cluster  ESR1</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> 1                   10.3 </span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> 10                   6.39</span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> 2                   10.9 </span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> 3                   10.5 </span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> 4ER-                 6.55</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> 4ER+                 9.78</span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> 5                    7.78</span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> 6                   10.9 </span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> 7                   10.9 </span>
<span class="co">## <span style="color: #BCBCBC;">10</span> 8                   11.1 </span>
<span class="co">## <span style="color: #BCBCBC;">11</span> 9                    9.96</span></code></pre>
<p>The following schematic contains another example using a simplified
subset of the METABRIC tumour samples to show what’s going on.</p>
<p><img src="images/group_and_summarise.png"></p>
<p>As we saw last session, we can summarize multiple observations,
e.g. the mean expression for other genes of interest, with
<code><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise()</a></code> and <code><a href="https://dplyr.tidyverse.org/reference/across.html" class="external-link">across()</a></code>, this time using the
PAM50 classification to define the groups.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">metabric</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">PAM50</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html" class="external-link">across</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">ESR1</span>, <span class="va">PGR</span>, <span class="va">ERBB2</span><span class="op">)</span>, <span class="va">mean</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 7 × 4</span></span>
<span class="co">##   PAM50        ESR1   PGR ERBB2</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Basal        6.42  5.46 10.2 </span>
<span class="co">## <span style="color: #BCBCBC;">2</span> claudin-low  7.47  5.60  9.85</span>
<span class="co">## <span style="color: #BCBCBC;">3</span> Her2         7.81  5.62 12.6 </span>
<span class="co">## <span style="color: #BCBCBC;">4</span> LumA        10.8   6.75 10.7 </span>
<span class="co">## <span style="color: #BCBCBC;">5</span> LumB        11.0   6.39 10.6 </span>
<span class="co">## <span style="color: #BCBCBC;">6</span> NC          10.9   6.47 10.3 </span>
<span class="co">## <span style="color: #BCBCBC;">7</span> Normal       9.50  6.21 10.8</span></code></pre>
<p>We can also refine our groups by using more than one categorical
variable. Let’s subdivide the PAM50 groups by HER2 status to illustrate
this.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">metabric</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">PAM50</span>, <span class="va">HER2_status</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>ESR1_mean <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">ESR1</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## `summarise()` has grouped output by 'PAM50'. You can override using the</span>
<span class="co">## `.groups` argument.</span></code></pre>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 13 × 3</span></span>
<span class="co">## <span style="color: #949494;"># Groups:   PAM50 [7]</span></span>
<span class="co">##    PAM50       HER2_status ESR1_mean</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> Basal       Negative         6.39</span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> Basal       Positive         6.71</span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> claudin-low Negative         7.52</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> claudin-low Positive         6.80</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> Her2        Negative         8.82</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> Her2        Positive         7.04</span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> LumA        Negative        10.8 </span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> LumA        Positive        10.1 </span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> LumB        Negative        11.1 </span>
<span class="co">## <span style="color: #BCBCBC;">10</span> LumB        Positive        10.2 </span>
<span class="co">## <span style="color: #BCBCBC;">11</span> NC          Negative        10.9 </span>
<span class="co">## <span style="color: #BCBCBC;">12</span> Normal      Negative         9.68</span>
<span class="co">## <span style="color: #BCBCBC;">13</span> Normal      Positive         7.77</span></code></pre>
<p>It can be quite useful to know how many observations are within each
group. We can use a special function, <strong><code><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n()</a></code></strong>,
that counts the number of rows rather than computing a summary value
from one of the columns.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">metabric</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">PAM50</span>, <span class="va">HER2_status</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>N <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span>, ESR1_mean <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">ESR1</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## `summarise()` has grouped output by 'PAM50'. You can override using the</span>
<span class="co">## `.groups` argument.</span></code></pre>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 13 × 4</span></span>
<span class="co">## <span style="color: #949494;"># Groups:   PAM50 [7]</span></span>
<span class="co">##    PAM50       HER2_status     N ESR1_mean</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> Basal       Negative      179      6.39</span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> Basal       Positive       20      6.71</span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> claudin-low Negative      184      7.52</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> claudin-low Positive       15      6.80</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> Her2        Negative       95      8.82</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> Her2        Positive      125      7.04</span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> LumA        Negative      658     10.8 </span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> LumA        Positive       21     10.1 </span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> LumB        Negative      419     11.1 </span>
<span class="co">## <span style="color: #BCBCBC;">10</span> LumB        Positive       42     10.2 </span>
<span class="co">## <span style="color: #BCBCBC;">11</span> NC          Negative        6     10.9 </span>
<span class="co">## <span style="color: #BCBCBC;">12</span> Normal      Negative      127      9.68</span>
<span class="co">## <span style="color: #BCBCBC;">13</span> Normal      Positive       13      7.77</span></code></pre>
</div>
<div class="section level3">
<h3 id="counts">Counts<a class="anchor" aria-label="anchor" href="#counts"></a>
</h3>
<p>Counting observations within groups is such a common operation that
dplyr provides a <strong><code><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count()</a></code></strong> function to do
just that. So we could count the number of patient samples in each of
the PAM50 classes as follows.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">metabric</span>, <span class="va">PAM50</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 7 × 2</span></span>
<span class="co">##   PAM50           n</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Basal         199</span>
<span class="co">## <span style="color: #BCBCBC;">2</span> claudin-low   199</span>
<span class="co">## <span style="color: #BCBCBC;">3</span> Her2          220</span>
<span class="co">## <span style="color: #BCBCBC;">4</span> LumA          679</span>
<span class="co">## <span style="color: #BCBCBC;">5</span> LumB          461</span>
<span class="co">## <span style="color: #BCBCBC;">6</span> NC              6</span>
<span class="co">## <span style="color: #BCBCBC;">7</span> Normal        140</span></code></pre>
<p>This is much like the <code><a href="https://rdrr.io/r/base/table.html" class="external-link">table()</a></code> function we’ve used
several times already to take a quick look at what values are contained
in one of the columns in a data frame. They return different data
structures however, with <code><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count()</a></code> always returning a data
frame (or tibble) that can then be passed to subsequent steps in a
‘piped’ workflow.</p>
<p>If we wanted to subdivide our categories by HER2 status, we can add
this as an additional categorical variable just as we did with the
previous <code><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by()</a></code> examples.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">metabric</span>, <span class="va">PAM50</span>, <span class="va">HER2_status</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 13 × 3</span></span>
<span class="co">##    PAM50       HER2_status     n</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> Basal       Negative      179</span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> Basal       Positive       20</span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> claudin-low Negative      184</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> claudin-low Positive       15</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> Her2        Negative       95</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> Her2        Positive      125</span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> LumA        Negative      658</span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> LumA        Positive       21</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> LumB        Negative      419</span>
<span class="co">## <span style="color: #BCBCBC;">10</span> LumB        Positive       42</span>
<span class="co">## <span style="color: #BCBCBC;">11</span> NC          Negative        6</span>
<span class="co">## <span style="color: #BCBCBC;">12</span> Normal      Negative      127</span>
<span class="co">## <span style="color: #BCBCBC;">13</span> Normal      Positive       13</span></code></pre>
<p>The count column is named ‘n’ by default but you can change this.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">metabric</span>, <span class="va">PAM50</span>, <span class="va">HER2_status</span>, name <span class="op">=</span> <span class="st">"Samples"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 13 × 3</span></span>
<span class="co">##    PAM50       HER2_status Samples</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> Basal       Negative        179</span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> Basal       Positive         20</span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> claudin-low Negative        184</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> claudin-low Positive         15</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> Her2        Negative         95</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> Her2        Positive        125</span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> LumA        Negative        658</span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> LumA        Positive         21</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> LumB        Negative        419</span>
<span class="co">## <span style="color: #BCBCBC;">10</span> LumB        Positive         42</span>
<span class="co">## <span style="color: #BCBCBC;">11</span> NC          Negative          6</span>
<span class="co">## <span style="color: #BCBCBC;">12</span> Normal      Negative        127</span>
<span class="co">## <span style="color: #BCBCBC;">13</span> Normal      Positive         13</span></code></pre>
<p><code><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count()</a></code> is equivalent to grouping observations with
<code><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by()</a></code> and calling <code><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize()</a></code> using the
special <code><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n()</a></code> function to count the number of rows. So the
above statement could have been written in a more long-winded way as
follows.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">metabric</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">PAM50</span>, <span class="va">HER2_status</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>Samples <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Summarizing with <code><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n()</a></code> is useful when showing the number
of observations in a group alongside a summary value, as we did earlier
looking at the mean ESR1 expression within specified groups; it allows
you to see if you’re drawing conclusions from only a few data
points.</p>
</div>
<div class="section level3">
<h3 id="missing-values">Missing values<a class="anchor" aria-label="anchor" href="#missing-values"></a>
</h3>
<p>Many summarization functions return <code>NA</code> if any of the
values are missing, i.e. the column contains <code>NA</code> values. As
an example, we’ll compute the average size of ER-negative and
ER-positive tumours.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">metabric</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">ER_status</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>N <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span>, `Average tumour size` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">Tumour_size</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 2 × 3</span></span>
<span class="co">##   ER_status     N `Average tumour size`</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>                 <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Negative    445                    <span style="color: #BB0000;">NA</span></span>
<span class="co">## <span style="color: #BCBCBC;">2</span> Positive   <span style="text-decoration: underline;">1</span>459                    <span style="color: #BB0000;">NA</span></span></code></pre>
<p>The <code><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean()</a></code> function, along with many similar
summarization functions, has an <code>na.rm</code> argument that can be
set to <code>TRUE</code> to exclude those missing values from the
calculation.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">metabric</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">ER_status</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>N <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span>, `Average tumour size` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">Tumour_size</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 2 × 3</span></span>
<span class="co">##   ER_status     N `Average tumour size`</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>                 <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Negative    445                  28.5</span>
<span class="co">## <span style="color: #BCBCBC;">2</span> Positive   <span style="text-decoration: underline;">1</span>459                  25.6</span></code></pre>
<p>An alternative would be to filter out the observations with missing
values but then the number of samples in each ER status group would take
on a different meaning, which may or may not be what we actually
want.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">metabric</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">Tumour_size</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">ER_status</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>N <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span>, `Average tumour size` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">Tumour_size</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 2 × 3</span></span>
<span class="co">##   ER_status     N `Average tumour size`</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>                 <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Negative    438                  28.5</span>
<span class="co">## <span style="color: #BCBCBC;">2</span> Positive   <span style="text-decoration: underline;">1</span>446                  25.6</span></code></pre>
</div>
<div class="section level3">
<h3 id="counts-and-proportions">Counts and proportions<a class="anchor" aria-label="anchor" href="#counts-and-proportions"></a>
</h3>
<p>The <code><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum()</a></code> and <code><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean()</a></code> summarization
functions are often used with logical values. It might seem surprising
to compute a summary for a logical variable but but this turns out to be
quite a useful thing to do, for counting the number of <code>TRUE</code>
values or obtaining the proportion of values that are
<code>TRUE</code>.</p>
<p>Following on from the previous example we could add a column to our
summary of average tumour size for ER-positive and ER-negative patients
that contains the number of missing values.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">metabric</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">ER_status</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>N <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span>, Missing <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">Tumour_size</span><span class="op">)</span><span class="op">)</span>, `Average tumour size` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">Tumour_size</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 2 × 4</span></span>
<span class="co">##   ER_status     N Missing `Average tumour size`</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>                 <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Negative    445       7                  28.5</span>
<span class="co">## <span style="color: #BCBCBC;">2</span> Positive   <span style="text-decoration: underline;">1</span>459      13                  25.6</span></code></pre>
<p>Why does this work? Well, the <code><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na()</a></code> function takes a
vector and sees which values are <code>NA</code>, returning a logical
vector of <code>TRUE</code> where the value was <code>NA</code> and
<code>FALSE</code> if not.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">test_vector</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span>, <span class="fl">2</span>, <span class="cn">NA</span>, <span class="fl">6</span>, <span class="fl">5</span>, <span class="cn">NA</span>, <span class="fl">10</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">test_vector</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] FALSE FALSE FALSE  TRUE FALSE FALSE  TRUE FALSE</span></code></pre>
<p>The <code><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum()</a></code> function treats the logical vector as a set of
<code>0</code>s and <code>1</code>s where <code>FALSE</code> is
<code>0</code> and <code>TRUE</code> is <code>1</code>. In effect
<code><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum()</a></code> counts the number of <code>TRUE</code> values.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">test_vector</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 2</span></code></pre>
<p>Similarly, <code><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean()</a></code> will compute the proportion of the
values that are <code>TRUE</code>.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">test_vector</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 0.25</span></code></pre>
<p>So let’s calculate the number and proportion of samples that do not
have a recorded tumour size in each of the ER-negative and ER-positive
groups.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">metabric</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">ER_status</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>N <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span>, `Missing tumour size` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">Tumour_size</span><span class="op">)</span><span class="op">)</span>, `Proportion missing` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">Tumour_size</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 2 × 4</span></span>
<span class="co">##   ER_status     N `Missing tumour size` `Proportion missing`</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>                 <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Negative    445                     7              0.015<span style="text-decoration: underline;">7</span> </span>
<span class="co">## <span style="color: #BCBCBC;">2</span> Positive   <span style="text-decoration: underline;">1</span>459                    13              0.008<span style="text-decoration: underline;">91</span></span></code></pre>
<p>We can use <code><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum()</a></code> and <code><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean()</a></code> for any
condition that returns a logical vector. We could, for example, find the
number and proportion of patients that survived longer than 10 years
(120 months) in each of the ER-negative and ER-positive groups.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">metabric</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">Survival_status</span> <span class="op">==</span> <span class="st">"DECEASED"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">ER_status</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>N <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span>, N_long_survival <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">Survival_time</span> <span class="op">&gt;</span> <span class="fl">120</span><span class="op">)</span>, Proportion_long_survival <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">Survival_time</span> <span class="op">&gt;</span> <span class="fl">120</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 2 × 4</span></span>
<span class="co">##   ER_status     N N_long_survival Proportion_long_survival</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>                    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Negative    250              40                    0.16 </span>
<span class="co">## <span style="color: #BCBCBC;">2</span> Positive    853             325                    0.381</span></code></pre>
</div>
<div class="section level3">
<h3 id="selecting-or-counting-distinct-things">Selecting or counting distinct things<a class="anchor" aria-label="anchor" href="#selecting-or-counting-distinct-things"></a>
</h3>
<p>There are occasions when we want to count the number of distinct
values in a variable or a combination of variables. In this week’s
assignment, we introduce another set of data from the METABRIC study
which contains details of the mutations detected by targeted sequencing
of a panel of 173 genes. We’ll read this data into R now as this
provides a good example of having multiple observations in different
rows for a single observational unit, in this case several mutations
detected in each tumour sample.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mutations</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="st">"https://zenodo.org/record/6450144/files/metabric_mutations.csv"</span><span class="op">)</span>
<span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">mutations</span>, <span class="va">Patient_ID</span>, <span class="va">Chromosome</span>, Position <span class="op">=</span> <span class="va">Start_Position</span>, Ref <span class="op">=</span> <span class="va">Reference_Allele</span>, Alt <span class="op">=</span> <span class="va">Tumor_Seq_Allele1</span>, Type <span class="op">=</span> <span class="va">Variant_Type</span>, <span class="va">Gene</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 17,272 × 7</span></span>
<span class="co">##    Patient_ID Chromosome  Position Ref   Alt   Type  Gene </span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> MTS-T0058  17           7<span style="text-decoration: underline;">579</span>344 <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    INS   TP53 </span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> MTS-T0058  17           7<span style="text-decoration: underline;">579</span>346 <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    INS   TP53 </span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> MTS-T0058  6          168<span style="text-decoration: underline;">299</span>111 G     G     SNP   MLLT4</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> MTS-T0058  22          29<span style="text-decoration: underline;">999</span>995 G     G     SNP   NF2  </span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> MTS-T0059  2          198<span style="text-decoration: underline;">288</span>682 A     A     SNP   SF3B1</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> MTS-T0059  6           86<span style="text-decoration: underline;">195</span>125 T     T     SNP   NT5E </span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> MTS-T0059  7           55<span style="text-decoration: underline;">241</span>717 C     C     SNP   EGFR </span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> MTS-T0059  10           6<span style="text-decoration: underline;">556</span>986 C     C     SNP   PRKCQ</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> MTS-T0059  11          62<span style="text-decoration: underline;">300</span>529 T     T     SNP   AHNAK</span>
<span class="co">## <span style="color: #BCBCBC;">10</span> MTS-T0059  15          74<span style="text-decoration: underline;">912</span>475 G     G     SNP   CLK3 </span>
<span class="co">## <span style="color: #949494;"># … with 17,262 more rows</span></span></code></pre>
<p>We can see from just these few rows that each patient sample has
multiple mutations and sometimes there are more than one mutation in the
same gene within a sample, as can be seen in the first two rows at the
top of the table above.</p>
<p>If we want to count the number of patients in which mutations were
detected we could select the distinct set of patient identifiers using
the <strong><code><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct()</a></code></strong> function.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mutations</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="va">Patient_ID</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 2369</span></code></pre>
<p>Similarly, we could select the distinct set of mutated genes for each
patient as follows.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mutations</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="va">Patient_ID</span>, <span class="va">Gene</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 15,656 × 2</span></span>
<span class="co">##    Patient_ID Gene </span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> MTS-T0058  TP53 </span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> MTS-T0058  MLLT4</span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> MTS-T0058  NF2  </span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> MTS-T0059  SF3B1</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> MTS-T0059  NT5E </span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> MTS-T0059  EGFR </span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> MTS-T0059  PRKCQ</span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> MTS-T0059  AHNAK</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> MTS-T0059  CLK3 </span>
<span class="co">## <span style="color: #BCBCBC;">10</span> MTS-T0059  TP53 </span>
<span class="co">## <span style="color: #949494;"># … with 15,646 more rows</span></span></code></pre>
<p>This has reduced the number of rows as only distinct combinations of
patient and gene are retained. This would be necessary if we wanted to
count the number of patients that have mutations in each gene rather
than the number of mutations for that gene regardless of the
patient.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># number of mutations for each gene</span>
<span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">mutations</span>, <span class="va">Gene</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 173 × 2</span></span>
<span class="co">##    Gene       n</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> ACVRL1    20</span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> AFF2      70</span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> AGMO      43</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> AGTR2     14</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> AHNAK    327</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> AHNAK2   859</span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> AKAP9    182</span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> AKT1     115</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> AKT2      23</span>
<span class="co">## <span style="color: #BCBCBC;">10</span> ALK       98</span>
<span class="co">## <span style="color: #949494;"># … with 163 more rows</span></span></code></pre>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># number of tumour samples in which each gene is mutated</span>
<span class="va">mutations</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="va">Patient_ID</span>, <span class="va">Gene</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">Gene</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 173 × 2</span></span>
<span class="co">##    Gene       n</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> ACVRL1    20</span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> AFF2      69</span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> AGMO      43</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> AGTR2     14</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> AHNAK    272</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> AHNAK2   530</span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> AKAP9    173</span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> AKT1     115</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> AKT2      23</span>
<span class="co">## <span style="color: #BCBCBC;">10</span> ALK       95</span>
<span class="co">## <span style="color: #949494;"># … with 163 more rows</span></span></code></pre>
<p>The genes that differ in these two tables are those that have more
than one mutation within a patient tumour sample.</p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="joining-data">Joining data<a class="anchor" aria-label="anchor" href="#joining-data"></a>
</h2>
<p>In many real life situations, data are spread across multiple tables
or spreadsheets. Usually this occurs because different types of
information about a subject, e.g. a patient, are collected from
different sources. It may be desirable for some analyses to combine data
from two or more tables into a single data frame based on a common
column, for example, an attribute that uniquely identifies the subject
such as a patient identifier.</p>
<p>dplyr provides a set of join functions for combining two data frames
based on matches within specified columns. These operations are very
similar to carrying out join operations between tables in a relational
database using SQL.</p>
<div class="section level3">
<h3 id="left_join">
<code>left_join</code><a class="anchor" aria-label="anchor" href="#left_join"></a>
</h3>
<p>To illustrate join operations we’ll first consider the most common
type, a “left join”. In the schematic below the two data frames share a
common column, V1. We can combine the two data frames into a single data
frame by matching rows in the first data frame with those in the second
data frame that share the same value of variable V1.</p>
<div class="figure">
<img src="images/dplyr_left_join.png" alt=""><p class="caption">dplyr left join</p>
</div>
<p><code><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join()</a></code> returns all rows from the first data frame
regardless of whether there is a match in the second data frame. Rows
with no match are included in the resulting data frame but have
<code>NA</code> values in the additional columns coming from the second
data frame.</p>
<p>Here’s an example in which details about members of the Beatles and
Rolling Stones are contained in two tables, using data frames
conveniently provided by dplyr (we’ll look at a real example
shortly).</p>
<p>The name column identifies each of the band members and is used for
matching rows from the two tables.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">band_members</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 3 × 2</span></span>
<span class="co">##   name  band   </span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Mick  Stones </span>
<span class="co">## <span style="color: #BCBCBC;">2</span> John  Beatles</span>
<span class="co">## <span style="color: #BCBCBC;">3</span> Paul  Beatles</span></code></pre>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">band_instruments</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 3 × 2</span></span>
<span class="co">##   name  plays </span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span>
<span class="co">## <span style="color: #BCBCBC;">1</span> John  guitar</span>
<span class="co">## <span style="color: #BCBCBC;">2</span> Paul  bass  </span>
<span class="co">## <span style="color: #BCBCBC;">3</span> Keith guitar</span></code></pre>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">band_members</span>, <span class="va">band_instruments</span>, by <span class="op">=</span> <span class="st">"name"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 3 × 3</span></span>
<span class="co">##   name  band    plays </span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Mick  Stones  <span style="color: #BB0000;">NA</span>    </span>
<span class="co">## <span style="color: #BCBCBC;">2</span> John  Beatles guitar</span>
<span class="co">## <span style="color: #BCBCBC;">3</span> Paul  Beatles bass</span></code></pre>
<p>We have joined the band members and instruments tables based on the
common <code>name</code> column. Because this is a left join, only
observations for band members in the ‘left’ table
(<code>band_members</code>) are included with information brought in
from the ‘right’ table (<code>band_instruments</code>) where such
exists. There is no entry in <code>band_instruments</code> for Mick so
an <code>NA</code> value is inserted into the <code>plays</code> column
that gets added in the combined data frame. Keith is only included in
the <code>band_instruments</code> data frame so doesn’t make it into the
final output as this is based on those band members in the ‘left’
table.</p>
<p><code><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">right_join()</a></code> is similar but returns all rows from the
second data frame, i.e. the ‘right’ data frame, that have a match with
rows in the first data frame.</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">right_join</a></span><span class="op">(</span><span class="va">band_members</span>, <span class="va">band_instruments</span>, by <span class="op">=</span> <span class="st">"name"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 3 × 3</span></span>
<span class="co">##   name  band    plays </span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span>
<span class="co">## <span style="color: #BCBCBC;">1</span> John  Beatles guitar</span>
<span class="co">## <span style="color: #BCBCBC;">2</span> Paul  Beatles bass  </span>
<span class="co">## <span style="color: #BCBCBC;">3</span> Keith <span style="color: #BB0000;">NA</span>      guitar</span></code></pre>
<p><code><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">right_join()</a></code> is used very infrequently compared with
<code><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join()</a></code>.</p>
</div>
<div class="section level3">
<h3 id="inner_join">
<code>inner_join</code><a class="anchor" aria-label="anchor" href="#inner_join"></a>
</h3>
<p>Another joining operation is the “inner join” in which only
observations that are common to both data frames are included.</p>
<div class="figure">
<img src="images/dplyr_inner_join.png" alt=""><p class="caption">dplyr inner join</p>
</div>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">inner_join</a></span><span class="op">(</span><span class="va">band_members</span>, <span class="va">band_instruments</span>, by <span class="op">=</span> <span class="st">"name"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 2 × 3</span></span>
<span class="co">##   name  band    plays </span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span>
<span class="co">## <span style="color: #BCBCBC;">1</span> John  Beatles guitar</span>
<span class="co">## <span style="color: #BCBCBC;">2</span> Paul  Beatles bass</span></code></pre>
<p>In this case when considering observations identified by
<code>name</code>, only John and Paul are contained in both the
<code>band_members</code> and <code>band_instruments</code> tables, so
only these make it into the combined table.</p>
</div>
<div class="section level3">
<h3 id="full_join">
<code>full_join</code><a class="anchor" aria-label="anchor" href="#full_join"></a>
</h3>
<p>We’ve seen how missing rows from one table can be retained in the
joined data frame using <code>left_join</code> or
<code>right_join</code> but sometimes data for a given subject may be
missing from either of the tables and we still want that subject to
appear in the combined table. A <code>full_join</code> will return all
rows and all columns from the two tables and where there are no matching
values, <code>NA</code> values are used to fill in the missing
values.</p>
<div class="figure">
<img src="images/dplyr_full_join.png" alt=""><p class="caption">dplyr full join</p>
</div>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">full_join</a></span><span class="op">(</span><span class="va">band_members</span>, <span class="va">band_instruments</span>, by <span class="op">=</span> <span class="st">"name"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 4 × 3</span></span>
<span class="co">##   name  band    plays </span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Mick  Stones  <span style="color: #BB0000;">NA</span>    </span>
<span class="co">## <span style="color: #BCBCBC;">2</span> John  Beatles guitar</span>
<span class="co">## <span style="color: #BCBCBC;">3</span> Paul  Beatles bass  </span>
<span class="co">## <span style="color: #BCBCBC;">4</span> Keith <span style="color: #BB0000;">NA</span>      guitar</span></code></pre>
<p>Now, with <code><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">full_join()</a></code>, we have rows for both Mick and
Keith even though they are only in one or other of the tables being
joined.</p>
</div>
<div class="section level3">
<h3 id="joining-on-columns-with-different-headers">Joining on columns with different headers<a class="anchor" aria-label="anchor" href="#joining-on-columns-with-different-headers"></a>
</h3>
<p>It isn’t uncommon for the columns used for joining two tables to have
different names in each table. Of course we could rename one of the two
columns, e.g. using the dplyr <code><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename()</a></code> function, but the
dplyr join functions allow you to match using differently-named columns
as illustrated using another version of the
<code>band_instruments</code> data frame.</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">band_instruments2</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 3 × 2</span></span>
<span class="co">##   artist plays </span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span>
<span class="co">## <span style="color: #BCBCBC;">1</span> John   guitar</span>
<span class="co">## <span style="color: #BCBCBC;">2</span> Paul   bass  </span>
<span class="co">## <span style="color: #BCBCBC;">3</span> Keith  guitar</span></code></pre>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">band_members</span>, <span class="va">band_instruments2</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"name"</span> <span class="op">=</span> <span class="st">"artist"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 3 × 3</span></span>
<span class="co">##   name  band    plays </span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Mick  Stones  <span style="color: #BB0000;">NA</span>    </span>
<span class="co">## <span style="color: #BCBCBC;">2</span> John  Beatles guitar</span>
<span class="co">## <span style="color: #BCBCBC;">3</span> Paul  Beatles bass</span></code></pre>
<p>The name for the column used for joining is the one given in the
first table, i.e. the ‘left’ table, so <code>name</code> rather than
<code>artist</code> in this case.</p>
</div>
<div class="section level3">
<h3 id="multiple-matches-in-join-operations">Multiple matches in join operations<a class="anchor" aria-label="anchor" href="#multiple-matches-in-join-operations"></a>
</h3>
<p>You may be wondering what happens if there are multiple rows in one
of both of the two tables for the thing that is being joined, for
example what would happen if our second table had two entries for
instruments that Paul plays.</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">band_instruments3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>
  name <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"John"</span>,   <span class="st">"Paul"</span>, <span class="st">"Paul"</span>,   <span class="st">"Keith"</span><span class="op">)</span>,
  plays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"guitar"</span>, <span class="st">"bass"</span>, <span class="st">"guitar"</span>, <span class="st">"guitar"</span><span class="op">)</span>
<span class="op">)</span>
<span class="va">band_instruments3</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 4 × 2</span></span>
<span class="co">##   name  plays </span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span>
<span class="co">## <span style="color: #BCBCBC;">1</span> John  guitar</span>
<span class="co">## <span style="color: #BCBCBC;">2</span> Paul  bass  </span>
<span class="co">## <span style="color: #BCBCBC;">3</span> Paul  guitar</span>
<span class="co">## <span style="color: #BCBCBC;">4</span> Keith guitar</span></code></pre>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">band_members</span>, <span class="va">band_instruments3</span>, by <span class="op">=</span> <span class="st">"name"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 4 × 3</span></span>
<span class="co">##   name  band    plays </span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Mick  Stones  <span style="color: #BB0000;">NA</span>    </span>
<span class="co">## <span style="color: #BCBCBC;">2</span> John  Beatles guitar</span>
<span class="co">## <span style="color: #BCBCBC;">3</span> Paul  Beatles bass  </span>
<span class="co">## <span style="color: #BCBCBC;">4</span> Paul  Beatles guitar</span></code></pre>
<p>We get both entries from the second table added to the first
table.</p>
<p>Let’s add an entry for Paul being in a second band and see what
happens then when we combine the two tables, each with two entries for
Paul.</p>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">band_members3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>
  name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Mick"</span>,   <span class="st">"John"</span>,    <span class="st">"Paul"</span>,    <span class="st">"Paul"</span><span class="op">)</span>,
  band <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Stones"</span>, <span class="st">"Beatles"</span>, <span class="st">"Beatles"</span>, <span class="st">"Wings"</span><span class="op">)</span>
<span class="op">)</span>
<span class="va">band_members3</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 4 × 2</span></span>
<span class="co">##   name  band   </span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Mick  Stones </span>
<span class="co">## <span style="color: #BCBCBC;">2</span> John  Beatles</span>
<span class="co">## <span style="color: #BCBCBC;">3</span> Paul  Beatles</span>
<span class="co">## <span style="color: #BCBCBC;">4</span> Paul  Wings</span></code></pre>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">band_members3</span>, <span class="va">band_instruments3</span>, by <span class="op">=</span> <span class="st">"name"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 6 × 3</span></span>
<span class="co">##   name  band    plays </span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Mick  Stones  <span style="color: #BB0000;">NA</span>    </span>
<span class="co">## <span style="color: #BCBCBC;">2</span> John  Beatles guitar</span>
<span class="co">## <span style="color: #BCBCBC;">3</span> Paul  Beatles bass  </span>
<span class="co">## <span style="color: #BCBCBC;">4</span> Paul  Beatles guitar</span>
<span class="co">## <span style="color: #BCBCBC;">5</span> Paul  Wings   bass  </span>
<span class="co">## <span style="color: #BCBCBC;">6</span> Paul  Wings   guitar</span></code></pre>
<p>The resulting table includes all combinations of band and instrument
for Paul.</p>
</div>
<div class="section level3">
<h3 id="joining-by-matching-on-multiple-columns">Joining by matching on multiple columns<a class="anchor" aria-label="anchor" href="#joining-by-matching-on-multiple-columns"></a>
</h3>
<p>Sometimes the observations being combined are identified by multiple
columns, for example, a forename and a surname. We can specify a vector
of column names to be used in the join operation.</p>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">band_members4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>
  forename <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Mick"</span>,   <span class="st">"John"</span>,    <span class="st">"Paul"</span>,      <span class="st">"Mick"</span>,  <span class="st">"John"</span><span class="op">)</span>,
  surname <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Jagger"</span>, <span class="st">"Lennon"</span>,  <span class="st">"McCartney"</span>, <span class="st">"Avory"</span>, <span class="st">"Squire"</span><span class="op">)</span>,
  band <span class="op">=</span>     <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Stones"</span>, <span class="st">"Beatles"</span>, <span class="st">"Beatles"</span>,   <span class="st">"Kinks"</span>, <span class="st">"Roses"</span><span class="op">)</span>
<span class="op">)</span>
<span class="va">band_instruments4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>
  forename <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"John"</span>,   <span class="st">"Paul"</span>,      <span class="st">"Keith"</span>,    <span class="st">"Mick"</span>,  <span class="st">"John"</span><span class="op">)</span>,
  surname <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Lennon"</span>, <span class="st">"McCartney"</span>, <span class="st">"Richards"</span>, <span class="st">"Avory"</span>, <span class="st">"Squire"</span><span class="op">)</span>,
  plays <span class="op">=</span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"guitar"</span>, <span class="st">"bass"</span>,      <span class="st">"guitar"</span>,   <span class="st">"drums"</span>, <span class="st">"guitar"</span><span class="op">)</span>
<span class="op">)</span>
<span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">full_join</a></span><span class="op">(</span><span class="va">band_members4</span>, <span class="va">band_instruments4</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"forename"</span>, <span class="st">"surname"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 6 × 4</span></span>
<span class="co">##   forename surname   band    plays </span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Mick     Jagger    Stones  <span style="color: #BB0000;">NA</span>    </span>
<span class="co">## <span style="color: #BCBCBC;">2</span> John     Lennon    Beatles guitar</span>
<span class="co">## <span style="color: #BCBCBC;">3</span> Paul     McCartney Beatles bass  </span>
<span class="co">## <span style="color: #BCBCBC;">4</span> Mick     Avory     Kinks   drums </span>
<span class="co">## <span style="color: #BCBCBC;">5</span> John     Squire    Roses   guitar</span>
<span class="co">## <span style="color: #BCBCBC;">6</span> Keith    Richards  <span style="color: #BB0000;">NA</span>      guitar</span></code></pre>
</div>
<div class="section level3">
<h3 id="clashing-column-names">Clashing column names<a class="anchor" aria-label="anchor" href="#clashing-column-names"></a>
</h3>
<p>Occasionally we may find that there are duplicated columns in the two
tables we want to join, columns that aren’t those used for joining.
These variables may even contain different data but happen to have the
same name. In such cases dplyr joins add a suffix to each column in the
combined table.</p>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">band_members5</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>
  name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"John"</span>, <span class="st">"Paul"</span>, <span class="st">"Ringo"</span>, <span class="st">"George"</span>, <span class="st">"Mick"</span><span class="op">)</span>,
  birth_year <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1940</span>, <span class="fl">1942</span>, <span class="fl">1940</span>, <span class="fl">1943</span>, <span class="fl">1943</span><span class="op">)</span>,
  band <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Beatles"</span>, <span class="st">"Beatles"</span>, <span class="st">"Beatles"</span>, <span class="st">"Beatles"</span>, <span class="st">"Stones"</span><span class="op">)</span>
<span class="op">)</span>
<span class="va">band_instruments5</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>
  name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"John"</span>, <span class="st">"Paul"</span>, <span class="st">"Ringo"</span>, <span class="st">"George"</span><span class="op">)</span>,
  birth_year <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1940</span>, <span class="fl">1942</span>, <span class="fl">1940</span>, <span class="fl">1943</span><span class="op">)</span>,
  instrument <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"guitar"</span>, <span class="st">"bass"</span>, <span class="st">"drums"</span>, <span class="st">"guitar"</span><span class="op">)</span>
<span class="op">)</span>
<span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">band_members5</span>, <span class="va">band_instruments5</span>, by <span class="op">=</span> <span class="st">"name"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 5 × 5</span></span>
<span class="co">##   name   birth_year.x band    birth_year.y instrument</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span>
<span class="co">## <span style="color: #BCBCBC;">1</span> John           <span style="text-decoration: underline;">1</span>940 Beatles         <span style="text-decoration: underline;">1</span>940 guitar    </span>
<span class="co">## <span style="color: #BCBCBC;">2</span> Paul           <span style="text-decoration: underline;">1</span>942 Beatles         <span style="text-decoration: underline;">1</span>942 bass      </span>
<span class="co">## <span style="color: #BCBCBC;">3</span> Ringo          <span style="text-decoration: underline;">1</span>940 Beatles         <span style="text-decoration: underline;">1</span>940 drums     </span>
<span class="co">## <span style="color: #BCBCBC;">4</span> George         <span style="text-decoration: underline;">1</span>943 Beatles         <span style="text-decoration: underline;">1</span>943 guitar    </span>
<span class="co">## <span style="color: #BCBCBC;">5</span> Mick           <span style="text-decoration: underline;">1</span>943 Stones            <span style="color: #BB0000;">NA</span> <span style="color: #BB0000;">NA</span></span></code></pre>
<p>It is advisable to rename or remove the duplicated columns that
aren’t used for joining.</p>
</div>
<div class="section level3">
<h3 id="filtering-joins">Filtering joins<a class="anchor" aria-label="anchor" href="#filtering-joins"></a>
</h3>
<p>A variation on the join operations we’ve considered are
<code><a href="https://dplyr.tidyverse.org/reference/filter-joins.html" class="external-link">semi_join()</a></code> and <code><a href="https://dplyr.tidyverse.org/reference/filter-joins.html" class="external-link">anti_join()</a></code> that filter the
rows in one table based on matches or lack of matches to rows in another
table.</p>
<p><code><a href="https://dplyr.tidyverse.org/reference/filter-joins.html" class="external-link">semi_join()</a></code> returns all rows from the first table where
there are matches in the other table.</p>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter-joins.html" class="external-link">semi_join</a></span><span class="op">(</span><span class="va">band_members</span>, <span class="va">band_instruments</span>, by <span class="op">=</span> <span class="st">"name"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 2 × 2</span></span>
<span class="co">##   name  band   </span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span>
<span class="co">## <span style="color: #BCBCBC;">1</span> John  Beatles</span>
<span class="co">## <span style="color: #BCBCBC;">2</span> Paul  Beatles</span></code></pre>
<p><code><a href="https://dplyr.tidyverse.org/reference/filter-joins.html" class="external-link">anti_join()</a></code> returns all rows where there is no match in
the other table, i.e. those that are unique to the first table.</p>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter-joins.html" class="external-link">anti_join</a></span><span class="op">(</span><span class="va">band_members</span>, <span class="va">band_instruments</span>, by <span class="op">=</span> <span class="st">"name"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 1 × 2</span></span>
<span class="co">##   name  band  </span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Mick  Stones</span></code></pre>
</div>
<div class="section level3">
<h3 id="a-real-example-joining-the-metabric-clinical-and-mrna-expression-data">A real example: joining the METABRIC clinical and mRNA expression
data<a class="anchor" aria-label="anchor" href="#a-real-example-joining-the-metabric-clinical-and-mrna-expression-data"></a>
</h3>
<p>Let’s move on to a real example of joining data from two different
tables that we used in putting together the combined METABRIC clinical
and expression data set.</p>
<p>We first read the clinical data into R and then just select a small
number of columns to make it easier to see what is going on when
combining the data.</p>
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">clinical_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="st">"https://zenodo.org/record/6450144/files/metabric_clinical_data.csv"</span><span class="op">)</span>
<span class="va">clinical_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">clinical_data</span>, <span class="va">Patient_ID</span>, <span class="va">ER_status</span>, <span class="va">PAM50</span><span class="op">)</span>
<span class="va">clinical_data</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 2,509 × 3</span></span>
<span class="co">##    Patient_ID ER_status PAM50      </span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      </span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> MB-0000    Positive  claudin-low</span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> MB-0002    Positive  LumA       </span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> MB-0005    Positive  LumB       </span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> MB-0006    Positive  LumB       </span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> MB-0008    Positive  LumB       </span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> MB-0010    Positive  LumB       </span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> MB-0014    Positive  LumB       </span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> MB-0020    Negative  Normal     </span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> MB-0022    Positive  claudin-low</span>
<span class="co">## <span style="color: #BCBCBC;">10</span> MB-0025    Positive  <span style="color: #BB0000;">NA</span>         </span>
<span class="co">## <span style="color: #949494;"># … with 2,499 more rows</span></span></code></pre>
<p>We then read in the mRNA expression data that was downloaded
separately from <a href="https://www.cbioportal.org/study/summary?id=brca_metabric" class="external-link">cBioPortal</a>.</p>
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mrna_expression_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="st">"https://zenodo.org/record/6450144/files/metabric_mrna_expression.txt"</span><span class="op">)</span>
<span class="va">mrna_expression_data</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 2,509 × 10</span></span>
<span class="co">##    STUDY_ID      SAMPLE_ID  ESR1 ERBB2   PGR  TP53 PIK3CA GATA3 FOXA1  MLPH</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> brca_metabric MB-0000    8.93  9.33  5.68  6.34   5.70  6.93  7.95  9.73</span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> brca_metabric MB-0002   10.0   9.73  7.51  6.19   5.76 11.3  11.8  12.5 </span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> brca_metabric MB-0005   10.0   9.73  7.38  6.40   6.75  9.29 11.7  10.3 </span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> brca_metabric MB-0006   10.4  10.3   6.82  6.87   7.22  8.67 11.9  10.5 </span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> brca_metabric MB-0008   11.3   9.96  7.33  6.34   5.82  9.72 11.6  12.2 </span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> brca_metabric MB-0010   11.2   9.74  5.95  5.42   6.12  9.79 12.1  11.4 </span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> brca_metabric MB-0014   10.8   9.28  7.72  5.99   7.48  8.37 11.5  10.8 </span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> brca_metabric MB-0020   <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>   </span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> brca_metabric MB-0022   10.4   8.61  5.59  6.17   7.59  7.87 10.7   9.95</span>
<span class="co">## <span style="color: #BCBCBC;">10</span> brca_metabric MB-0025   <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>   </span>
<span class="co">## <span style="color: #949494;"># … with 2,499 more rows</span></span></code></pre>
<p>Now we have both sets of data loaded into R as data frames, we can
combine them into a single data frame using an
<code><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">inner_join()</a></code>. Our resulting table will only contain entries
for the patients for which expression data are available.</p>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">combined_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">inner_join</a></span><span class="op">(</span><span class="va">clinical_data</span>, <span class="va">mrna_expression_data</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Patient_ID"</span> <span class="op">=</span> <span class="st">"SAMPLE_ID"</span><span class="op">)</span><span class="op">)</span>
<span class="va">combined_data</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 2,509 × 12</span></span>
<span class="co">##    Patient_ID ER_status PAM50      STUDY_ID  ESR1 ERBB2   PGR  TP53 PIK3CA GATA3</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> MB-0000    Positive  claudin-l… brca_me…  8.93  9.33  5.68  6.34   5.70  6.93</span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> MB-0002    Positive  LumA       brca_me… 10.0   9.73  7.51  6.19   5.76 11.3 </span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> MB-0005    Positive  LumB       brca_me… 10.0   9.73  7.38  6.40   6.75  9.29</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> MB-0006    Positive  LumB       brca_me… 10.4  10.3   6.82  6.87   7.22  8.67</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> MB-0008    Positive  LumB       brca_me… 11.3   9.96  7.33  6.34   5.82  9.72</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> MB-0010    Positive  LumB       brca_me… 11.2   9.74  5.95  5.42   6.12  9.79</span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> MB-0014    Positive  LumB       brca_me… 10.8   9.28  7.72  5.99   7.48  8.37</span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> MB-0020    Negative  Normal     brca_me… <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>   </span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> MB-0022    Positive  claudin-l… brca_me… 10.4   8.61  5.59  6.17   7.59  7.87</span>
<span class="co">## <span style="color: #BCBCBC;">10</span> MB-0025    Positive  <span style="color: #BB0000;">NA</span>         brca_me… <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>   </span>
<span class="co">## <span style="color: #949494;"># … with 2,499 more rows, and 2 more variables: FOXA1 &lt;dbl&gt;, MLPH &lt;dbl&gt;</span></span></code></pre>
<p>Having combined the data, we can carry out exploratory data analysis
using elements from both data sets.</p>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">combined_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">PAM50</span><span class="op">)</span>, <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">ESR1</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PAM50</span>, y <span class="op">=</span> <span class="va">ESR1</span>, colour <span class="op">=</span> <span class="va">PAM50</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span>show.legend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p><img src="week4_files/figure-html/box_plot-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="restructuring-data">Restructuring data<a class="anchor" aria-label="anchor" href="#restructuring-data"></a>
</h2>
<p>The data you collect or obtain from a third party is not always in a
form that is suitable for exploratory analysis and visualization and may
need to be restructured before you can fully make use of it.</p>
<p>This is particularly true of the plotting and summarizing tools we’ve
been looking at in this course, which are designed specifically to work
on data in a format referred to as ‘tidy’. This is where the tidyverse
gets its name.</p>
<p>In this part of the session, we will look at what it means for data
to be ‘tidy’ and how you can transform your data, if necessary, into
this form. We’ll also look at useful functions for handling compound
variables, that is columns that contain more than one type of
measurement or attribute (you’d be surprised how common this is) and
some of the string manipulation functions from the stringr package that
can help with cleaning up values within a data frame.</p>
<p>Finally, we’ll take another look at customizing plots created with
ggplot2 by changing various non-data components that are largely
cosmetic.</p>
<p>The functions we’re mostly focusing on in this part are from the
<strong>tidyr</strong> and <strong>stringr</strong> packages both of
which get loaded as part of the tidyverse.</p>
</div>
<div class="section level2">
<h2 id="tidy-data">Tidy data<a class="anchor" aria-label="anchor" href="#tidy-data"></a>
</h2>
<p>So what is ‘tidy data’ and why should we care about it?</p>
<p>To answer these questions, we’ll look at different ways in which a
simple data set can be represented and consider the challenges
associated with each. The data set in question is a subset of data from
the <a href="https://www.who.int/health-topics/tuberculosis" class="external-link">World
Health Organization Global Tuberculosis Report</a>.</p>
<p>The tidyr package contains a series of tibbles that represent the
same set of information on the number of new cases of tuberculosis (TB)
recorded in each of 3 countries in the years 1999 and 2000 as well as
the populations of those countries.</p>
<p>Here is the first table, <code>table1</code>, that contains a row for
every combination of country and year and separate columns containing
the numbers of TB cases and the population.</p>
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">table1</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 6 × 4</span></span>
<span class="co">##   country      year  cases population</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Afghanistan  <span style="text-decoration: underline;">1</span>999    745   19<span style="text-decoration: underline;">987</span>071</span>
<span class="co">## <span style="color: #BCBCBC;">2</span> Afghanistan  <span style="text-decoration: underline;">2</span>000   <span style="text-decoration: underline;">2</span>666   20<span style="text-decoration: underline;">595</span>360</span>
<span class="co">## <span style="color: #BCBCBC;">3</span> Brazil       <span style="text-decoration: underline;">1</span>999  <span style="text-decoration: underline;">37</span>737  172<span style="text-decoration: underline;">006</span>362</span>
<span class="co">## <span style="color: #BCBCBC;">4</span> Brazil       <span style="text-decoration: underline;">2</span>000  <span style="text-decoration: underline;">80</span>488  174<span style="text-decoration: underline;">504</span>898</span>
<span class="co">## <span style="color: #BCBCBC;">5</span> China        <span style="text-decoration: underline;">1</span>999 <span style="text-decoration: underline;">212</span>258 <span style="text-decoration: underline;">1</span>272<span style="text-decoration: underline;">915</span>272</span>
<span class="co">## <span style="color: #BCBCBC;">6</span> China        <span style="text-decoration: underline;">2</span>000 <span style="text-decoration: underline;">213</span>766 <span style="text-decoration: underline;">1</span>280<span style="text-decoration: underline;">428</span>583</span></code></pre>
<p>Two alternative ways of representing the same information are given
in <code>table2</code> and <code>table3</code>. We’ll consider each of
these in turn, shortly.</p>
<div class="sourceCode" id="cb99"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">table2</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 12 × 4</span></span>
<span class="co">##    country      year type            count</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> Afghanistan  <span style="text-decoration: underline;">1</span>999 cases             745</span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> Afghanistan  <span style="text-decoration: underline;">1</span>999 population   19<span style="text-decoration: underline;">987</span>071</span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> Afghanistan  <span style="text-decoration: underline;">2</span>000 cases            <span style="text-decoration: underline;">2</span>666</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> Afghanistan  <span style="text-decoration: underline;">2</span>000 population   20<span style="text-decoration: underline;">595</span>360</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> Brazil       <span style="text-decoration: underline;">1</span>999 cases           <span style="text-decoration: underline;">37</span>737</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> Brazil       <span style="text-decoration: underline;">1</span>999 population  172<span style="text-decoration: underline;">006</span>362</span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> Brazil       <span style="text-decoration: underline;">2</span>000 cases           <span style="text-decoration: underline;">80</span>488</span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> Brazil       <span style="text-decoration: underline;">2</span>000 population  174<span style="text-decoration: underline;">504</span>898</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> China        <span style="text-decoration: underline;">1</span>999 cases          <span style="text-decoration: underline;">212</span>258</span>
<span class="co">## <span style="color: #BCBCBC;">10</span> China        <span style="text-decoration: underline;">1</span>999 population <span style="text-decoration: underline;">1</span>272<span style="text-decoration: underline;">915</span>272</span>
<span class="co">## <span style="color: #BCBCBC;">11</span> China        <span style="text-decoration: underline;">2</span>000 cases          <span style="text-decoration: underline;">213</span>766</span>
<span class="co">## <span style="color: #BCBCBC;">12</span> China        <span style="text-decoration: underline;">2</span>000 population <span style="text-decoration: underline;">1</span>280<span style="text-decoration: underline;">428</span>583</span></code></pre>
<div class="sourceCode" id="cb101"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">table3</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 6 × 3</span></span>
<span class="co">##   country      year rate             </span>
<span class="co">## <span style="color: #BCBCBC;">*</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            </span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Afghanistan  <span style="text-decoration: underline;">1</span>999 745/19987071     </span>
<span class="co">## <span style="color: #BCBCBC;">2</span> Afghanistan  <span style="text-decoration: underline;">2</span>000 2666/20595360    </span>
<span class="co">## <span style="color: #BCBCBC;">3</span> Brazil       <span style="text-decoration: underline;">1</span>999 37737/172006362  </span>
<span class="co">## <span style="color: #BCBCBC;">4</span> Brazil       <span style="text-decoration: underline;">2</span>000 80488/174504898  </span>
<span class="co">## <span style="color: #BCBCBC;">5</span> China        <span style="text-decoration: underline;">1</span>999 212258/1272915272</span>
<span class="co">## <span style="color: #BCBCBC;">6</span> China        <span style="text-decoration: underline;">2</span>000 213766/1280428583</span></code></pre>
<p>The final representation has the data split across two tables, a
scenario that is actually quite likely given that population data will
almost certainly have been collected separately from the recording of TB
cases.</p>
<div class="sourceCode" id="cb103"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">table4a</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 3 × 3</span></span>
<span class="co">##   country     `1999` `2000`</span>
<span class="co">## <span style="color: #BCBCBC;">*</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Afghanistan    745   <span style="text-decoration: underline;">2</span>666</span>
<span class="co">## <span style="color: #BCBCBC;">2</span> Brazil       <span style="text-decoration: underline;">37</span>737  <span style="text-decoration: underline;">80</span>488</span>
<span class="co">## <span style="color: #BCBCBC;">3</span> China       <span style="text-decoration: underline;">212</span>258 <span style="text-decoration: underline;">213</span>766</span></code></pre>
<div class="sourceCode" id="cb105"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">table4b</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 3 × 3</span></span>
<span class="co">##   country         `1999`     `2000`</span>
<span class="co">## <span style="color: #BCBCBC;">*</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Afghanistan   19<span style="text-decoration: underline;">987</span>071   20<span style="text-decoration: underline;">595</span>360</span>
<span class="co">## <span style="color: #BCBCBC;">2</span> Brazil       172<span style="text-decoration: underline;">006</span>362  174<span style="text-decoration: underline;">504</span>898</span>
<span class="co">## <span style="color: #BCBCBC;">3</span> China       <span style="text-decoration: underline;">1</span>272<span style="text-decoration: underline;">915</span>272 <span style="text-decoration: underline;">1</span>280<span style="text-decoration: underline;">428</span>583</span></code></pre>
<p>Time series data like this are very commonly represented as they are
in <code>table4a</code> and <code>table4b</code> with a series of dates
or years as columns extending across a spreadsheet. You will find
numerous examples of this if you seek out various data sets made
available by the UK Office for National Statistics and various other
national and international organizations.</p>
<p>Tables 1 to 4 are all different representations of the same
underlying data but one of these tables is structured in such a way as
to be most readily used in the tidyverse.</p>
<div class="section level3">
<h3 id="rules-for-tidy-data">Rules for tidy data<a class="anchor" aria-label="anchor" href="#rules-for-tidy-data"></a>
</h3>
<div class="rmdblock">
<p>
<strong>Tidy data</strong>
</p>
<p>
A tidy data set is a data frame (or table) for which the following are
true:
</p>
<ol style="list-style-type: decimal">
<li>
Each <strong>variable</strong> has its own column
</li>
<li>
Each <strong>observation</strong> has its own row
</li>
<li>
Each <strong>value</strong> has its own cell
</li>
</ol>
<p>
<img src="images/tidy_data.png"></p>
<p>
A <strong>variable</strong> contains all values that measure the same
underlying attribute (like height, temperature, duration) across units.
</p>
<p>
An <strong>observation</strong> contains all values measured on the same
unit (like a person or a day) across attributes.
</p>
<p>
<em>From <a href="https://www.jstatsoft.org/index.php/jss/article/view/v059i10/v59i10.pdf" class="external-link">‘Tidy
Data’</a> by Hadley Wickham.</em>
</p>
</div>
<p><strong>Question:</strong> <em>Which of the above representations of
TB cases is tidy?</em></p>
<p>Another way of framing the question is to consider what are the
variables in the TB data set, i.e. what are the things that vary and for
which we can attach a value for each observation?</p>
<p>Take another look at tables 4a and 4b. Do each of the columns
correspond to a variable? The country is certainly a variable. In this
data set it takes one of three values: Afghanistan, Brazil or China.</p>
<p>But what about the other two columns, ‘1999’ and ‘2000’? These are
values, not variables. The variable in this case would be ‘year’ and
could take a value of 1999 or 2000. Tables 4a and 4b are not tidy.</p>
<p>There is also another rather obvious problem with tables 4a and 4b –
the data are contained in two separate data frames. The data would
almost certainly have been collected separately, so it’s hardly
surprising, but whenever numbers of people affected by a disease, or
engaging in an activity, are compared between countries we almost always
want to be comparing the rate (the percentage within the population) and
not the absolute number, so that the comparison is fair. We need to
combine the data from these two tables in order to calculate the
rate.</p>
<p>The only table that is truly tidy is <code>table1</code>. It contains
one column for each of the variables in the data set, namely country,
year, the number of new TB cases and the population. We’ll look at
tables 2 and 3 shortly and why these aren’t tidy and what we can do
about it, but first we’ll see how we can convert tables 4a and 4b into
the tidy format.</p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="pivoting-operations">Pivoting operations<a class="anchor" aria-label="anchor" href="#pivoting-operations"></a>
</h2>
<div class="section level3">
<h3 id="pivot_longer">
<code>pivot_longer()</code><a class="anchor" aria-label="anchor" href="#pivot_longer"></a>
</h3>
<p>Tables 4a and 4b are examples of what is often referred to as ‘wide
format’. While neither table looks especially wide, you can imagine that
the more complete WHO data set contains data for very many years and
that if each had its own column, the table would be very much wider.</p>
<p>What we need is a column for ‘year’ so that we have a count, whether
it is the number of TB cases or the population, for each unique
combination of country and year. Transforming the table in this way is
known as ‘pivoting’ and the tidyr package provides the
<code><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer()</a></code> function for just such an operation.</p>
<div class="sourceCode" id="cb107"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">table4a</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 3 × 3</span></span>
<span class="co">##   country     `1999` `2000`</span>
<span class="co">## <span style="color: #BCBCBC;">*</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Afghanistan    745   <span style="text-decoration: underline;">2</span>666</span>
<span class="co">## <span style="color: #BCBCBC;">2</span> Brazil       <span style="text-decoration: underline;">37</span>737  <span style="text-decoration: underline;">80</span>488</span>
<span class="co">## <span style="color: #BCBCBC;">3</span> China       <span style="text-decoration: underline;">212</span>258 <span style="text-decoration: underline;">213</span>766</span></code></pre>
<div class="sourceCode" id="cb109"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">table4a_long</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span><span class="va">table4a</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">`1999`</span>, <span class="va">`2000`</span><span class="op">)</span>, names_to <span class="op">=</span> <span class="st">"year"</span>, values_to <span class="op">=</span> <span class="st">"cases"</span><span class="op">)</span>
<span class="va">table4a_long</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 6 × 3</span></span>
<span class="co">##   country     year   cases</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Afghanistan 1999     745</span>
<span class="co">## <span style="color: #BCBCBC;">2</span> Afghanistan 2000    <span style="text-decoration: underline;">2</span>666</span>
<span class="co">## <span style="color: #BCBCBC;">3</span> Brazil      1999   <span style="text-decoration: underline;">37</span>737</span>
<span class="co">## <span style="color: #BCBCBC;">4</span> Brazil      2000   <span style="text-decoration: underline;">80</span>488</span>
<span class="co">## <span style="color: #BCBCBC;">5</span> China       1999  <span style="text-decoration: underline;">212</span>258</span>
<span class="co">## <span style="color: #BCBCBC;">6</span> China       2000  <span style="text-decoration: underline;">213</span>766</span></code></pre>
<p>As with almost all tidyverse operations the first argument is the
data frame we’re working on. The second specifies which columns we are
operating on. Here we’ve used a vector with <code><a href="https://rdrr.io/r/base/c.html" class="external-link">c()</a></code> but it is
also quite customary and normally more convenient to use a range of
columns, e.g. <code>`1999`:`2000`</code>. Remember that we have to use
backticks because R doesn’t like variable names starting with
numbers.</p>
<div class="rmdblock">
<p>
<strong>Pivoting operations</strong>
</p>
<p>
<strong><code>pivot_longer(data, cols, names_to = “name”, values_to =
“value”)</code></strong>
</p>
<p>
Pivot data from wide to long, increasing the number of rows and
decreasing the number of columns.
</p>
<p>
<code>table4a_long &lt;- pivot_longer(table4a, c(“1999”, “2000”),
names_to = “year”, values_to = “cases”)</code>
</p>
<p>
<img src="images/pivot_longer.png"></p>
<p>
<strong><code>pivot_wider(data, names_from = name, values_from =
value)</code></strong>
</p>
<p>
Pivot data from long to wide, increasing the number of columns and
decreasing the number of rows.
</p>
<p>
<code>table4a_wide &lt;- pivot_wider(table4a_long, names_from = year,
values_from = cases)</code>
</p>
</div>
<p>The <code>names_to</code> and <code>values_to</code> arguments are so
called because we are taking the names of the columns we specified (1999
and 2000) and putting these in a new column with a name given by
<code>names_to</code>. Likewise, we are taking values in each of our
specified columns and putting these in a new column whose name is given
by <code>values_to</code>.</p>
<p>We can do the same with <code>table4b</code> and join the two
resulting tables together to recreate <code>table1</code>.</p>
<div class="sourceCode" id="cb111"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">table4b_long</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span><span class="va">table4b</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">`1999`</span>, <span class="va">`2000`</span><span class="op">)</span>, names_to <span class="op">=</span> <span class="st">"year"</span>, values_to <span class="op">=</span> <span class="st">"population"</span><span class="op">)</span>
<span class="va">table4b_long</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 6 × 3</span></span>
<span class="co">##   country     year  population</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Afghanistan 1999    19<span style="text-decoration: underline;">987</span>071</span>
<span class="co">## <span style="color: #BCBCBC;">2</span> Afghanistan 2000    20<span style="text-decoration: underline;">595</span>360</span>
<span class="co">## <span style="color: #BCBCBC;">3</span> Brazil      1999   172<span style="text-decoration: underline;">006</span>362</span>
<span class="co">## <span style="color: #BCBCBC;">4</span> Brazil      2000   174<span style="text-decoration: underline;">504</span>898</span>
<span class="co">## <span style="color: #BCBCBC;">5</span> China       1999  <span style="text-decoration: underline;">1</span>272<span style="text-decoration: underline;">915</span>272</span>
<span class="co">## <span style="color: #BCBCBC;">6</span> China       2000  <span style="text-decoration: underline;">1</span>280<span style="text-decoration: underline;">428</span>583</span></code></pre>
<div class="sourceCode" id="cb113"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># join the two pivoted tables to recreate table1</span>
<span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">table4a_long</span>, <span class="va">table4b_long</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"country"</span>, <span class="st">"year"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 6 × 4</span></span>
<span class="co">##   country     year   cases population</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Afghanistan 1999     745   19<span style="text-decoration: underline;">987</span>071</span>
<span class="co">## <span style="color: #BCBCBC;">2</span> Afghanistan 2000    <span style="text-decoration: underline;">2</span>666   20<span style="text-decoration: underline;">595</span>360</span>
<span class="co">## <span style="color: #BCBCBC;">3</span> Brazil      1999   <span style="text-decoration: underline;">37</span>737  172<span style="text-decoration: underline;">006</span>362</span>
<span class="co">## <span style="color: #BCBCBC;">4</span> Brazil      2000   <span style="text-decoration: underline;">80</span>488  174<span style="text-decoration: underline;">504</span>898</span>
<span class="co">## <span style="color: #BCBCBC;">5</span> China       1999  <span style="text-decoration: underline;">212</span>258 <span style="text-decoration: underline;">1</span>272<span style="text-decoration: underline;">915</span>272</span>
<span class="co">## <span style="color: #BCBCBC;">6</span> China       2000  <span style="text-decoration: underline;">213</span>766 <span style="text-decoration: underline;">1</span>280<span style="text-decoration: underline;">428</span>583</span></code></pre>
</div>
<div class="section level3">
<h3 id="pivot_wider">
<code>pivot_wider()</code><a class="anchor" aria-label="anchor" href="#pivot_wider"></a>
</h3>
<p><code><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html" class="external-link">pivot_wider()</a></code> has the opposite effect of
<code><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer()</a></code> and can be used to reverse the pivoting
operation we just performed on <code>table4a</code>.</p>
<div class="sourceCode" id="cb115"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">table4a_long</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 6 × 3</span></span>
<span class="co">##   country     year   cases</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Afghanistan 1999     745</span>
<span class="co">## <span style="color: #BCBCBC;">2</span> Afghanistan 2000    <span style="text-decoration: underline;">2</span>666</span>
<span class="co">## <span style="color: #BCBCBC;">3</span> Brazil      1999   <span style="text-decoration: underline;">37</span>737</span>
<span class="co">## <span style="color: #BCBCBC;">4</span> Brazil      2000   <span style="text-decoration: underline;">80</span>488</span>
<span class="co">## <span style="color: #BCBCBC;">5</span> China       1999  <span style="text-decoration: underline;">212</span>258</span>
<span class="co">## <span style="color: #BCBCBC;">6</span> China       2000  <span style="text-decoration: underline;">213</span>766</span></code></pre>
<div class="sourceCode" id="cb117"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html" class="external-link">pivot_wider</a></span><span class="op">(</span><span class="va">table4a_long</span>, names_from <span class="op">=</span> <span class="st">"year"</span>, values_from <span class="op">=</span> <span class="st">"cases"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 3 × 3</span></span>
<span class="co">##   country     `1999` `2000`</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Afghanistan    745   <span style="text-decoration: underline;">2</span>666</span>
<span class="co">## <span style="color: #BCBCBC;">2</span> Brazil       <span style="text-decoration: underline;">37</span>737  <span style="text-decoration: underline;">80</span>488</span>
<span class="co">## <span style="color: #BCBCBC;">3</span> China       <span style="text-decoration: underline;">212</span>258 <span style="text-decoration: underline;">213</span>766</span></code></pre>
<p><code><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html" class="external-link">pivot_wider()</a></code> has looked for all distinct values in the
<code>year</code> column and created a new column for each. The values
are taken from the <code>cases</code> column. An <code>NA</code> would
be inserted into the new columns for any instances where there is no
observation for a given year. That isn’t the case in this example but we
can show this by removing the last row from
<code>table4a_long</code>.</p>
<div class="sourceCode" id="cb119"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">table4a_long</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">table4a_long</span>, <span class="fl">5</span><span class="op">)</span>
<span class="va">table4a_long</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 5 × 3</span></span>
<span class="co">##   country     year   cases</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Afghanistan 1999     745</span>
<span class="co">## <span style="color: #BCBCBC;">2</span> Afghanistan 2000    <span style="text-decoration: underline;">2</span>666</span>
<span class="co">## <span style="color: #BCBCBC;">3</span> Brazil      1999   <span style="text-decoration: underline;">37</span>737</span>
<span class="co">## <span style="color: #BCBCBC;">4</span> Brazil      2000   <span style="text-decoration: underline;">80</span>488</span>
<span class="co">## <span style="color: #BCBCBC;">5</span> China       1999  <span style="text-decoration: underline;">212</span>258</span></code></pre>
<div class="sourceCode" id="cb121"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html" class="external-link">pivot_wider</a></span><span class="op">(</span><span class="va">table4a_long</span>, names_from <span class="op">=</span> <span class="st">"year"</span>, values_from <span class="op">=</span> <span class="st">"cases"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 3 × 3</span></span>
<span class="co">##   country     `1999` `2000`</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Afghanistan    745   <span style="text-decoration: underline;">2</span>666</span>
<span class="co">## <span style="color: #BCBCBC;">2</span> Brazil       <span style="text-decoration: underline;">37</span>737  <span style="text-decoration: underline;">80</span>488</span>
<span class="co">## <span style="color: #BCBCBC;">3</span> China       <span style="text-decoration: underline;">212</span>258     <span style="color: #BB0000;">NA</span></span></code></pre>
<p>In some cases the wide format is the more human-readable form and
usually it is a more compact way of representing the data. In this
example, there is no duplication of the year value in the wide format of
<code>table4a</code> and <code>table4b</code>. We will see later that
this is much more apparent with larger data tables, e.g. gene expression
matrices.</p>
<p>Let’s look again at <code>table2</code>.</p>
<div class="sourceCode" id="cb123"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">table2</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 12 × 4</span></span>
<span class="co">##    country      year type            count</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> Afghanistan  <span style="text-decoration: underline;">1</span>999 cases             745</span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> Afghanistan  <span style="text-decoration: underline;">1</span>999 population   19<span style="text-decoration: underline;">987</span>071</span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> Afghanistan  <span style="text-decoration: underline;">2</span>000 cases            <span style="text-decoration: underline;">2</span>666</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> Afghanistan  <span style="text-decoration: underline;">2</span>000 population   20<span style="text-decoration: underline;">595</span>360</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> Brazil       <span style="text-decoration: underline;">1</span>999 cases           <span style="text-decoration: underline;">37</span>737</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> Brazil       <span style="text-decoration: underline;">1</span>999 population  172<span style="text-decoration: underline;">006</span>362</span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> Brazil       <span style="text-decoration: underline;">2</span>000 cases           <span style="text-decoration: underline;">80</span>488</span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> Brazil       <span style="text-decoration: underline;">2</span>000 population  174<span style="text-decoration: underline;">504</span>898</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> China        <span style="text-decoration: underline;">1</span>999 cases          <span style="text-decoration: underline;">212</span>258</span>
<span class="co">## <span style="color: #BCBCBC;">10</span> China        <span style="text-decoration: underline;">1</span>999 population <span style="text-decoration: underline;">1</span>272<span style="text-decoration: underline;">915</span>272</span>
<span class="co">## <span style="color: #BCBCBC;">11</span> China        <span style="text-decoration: underline;">2</span>000 cases          <span style="text-decoration: underline;">213</span>766</span>
<span class="co">## <span style="color: #BCBCBC;">12</span> China        <span style="text-decoration: underline;">2</span>000 population <span style="text-decoration: underline;">1</span>280<span style="text-decoration: underline;">428</span>583</span></code></pre>
<p>Are the type and count columns true variables? And what is the
observational unit in this case?</p>
<p>If we consider the observational unit to be a country in a specific
year then <code>table2</code> is not tidy because observations are split
across two rows. Also the count variable contains counts of what are
essentially different things, the number of cases of TB or the total
population.</p>
<p>In tricky situations like this, a tell-tale sign that the data is not
in a tidy format is when we attempt to perform some exploratory data
analysis and visualization and find we’re having to do quite a bit of
work to calculate or plot what we want.</p>
<p>For example, if we wanted to create a bar plot of the numbers of TB
cases in each country, we would have to first remove the rows
corresponding to the populations using a filter operation. Surely that
wouldn’t be necessary if the data were tidy.</p>
<p>Similarly, the rate of new TB cases, i.e. the proportion of the
population infected with TB, is something we should be able to calculate
easily in a simple operation. However, this is actually quite difficult
to do with the data structured as they are in <code>table2</code>.</p>
<p>We can use <code><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html" class="external-link">pivot_wider()</a></code> to sort this out. The
<code>type</code> column contains the variable names so we’d need to set
<code>names_from = "type"</code>, while the values will be taken from
the <code>count</code> column.</p>
<div class="sourceCode" id="cb125"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">table2_fixed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html" class="external-link">pivot_wider</a></span><span class="op">(</span><span class="va">table2</span>, names_from <span class="op">=</span> <span class="st">"type"</span>, values_from <span class="op">=</span> <span class="st">"count"</span><span class="op">)</span>
<span class="va">table2_fixed</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 6 × 4</span></span>
<span class="co">##   country      year  cases population</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Afghanistan  <span style="text-decoration: underline;">1</span>999    745   19<span style="text-decoration: underline;">987</span>071</span>
<span class="co">## <span style="color: #BCBCBC;">2</span> Afghanistan  <span style="text-decoration: underline;">2</span>000   <span style="text-decoration: underline;">2</span>666   20<span style="text-decoration: underline;">595</span>360</span>
<span class="co">## <span style="color: #BCBCBC;">3</span> Brazil       <span style="text-decoration: underline;">1</span>999  <span style="text-decoration: underline;">37</span>737  172<span style="text-decoration: underline;">006</span>362</span>
<span class="co">## <span style="color: #BCBCBC;">4</span> Brazil       <span style="text-decoration: underline;">2</span>000  <span style="text-decoration: underline;">80</span>488  174<span style="text-decoration: underline;">504</span>898</span>
<span class="co">## <span style="color: #BCBCBC;">5</span> China        <span style="text-decoration: underline;">1</span>999 <span style="text-decoration: underline;">212</span>258 <span style="text-decoration: underline;">1</span>272<span style="text-decoration: underline;">915</span>272</span>
<span class="co">## <span style="color: #BCBCBC;">6</span> China        <span style="text-decoration: underline;">2</span>000 <span style="text-decoration: underline;">213</span>766 <span style="text-decoration: underline;">1</span>280<span style="text-decoration: underline;">428</span>583</span></code></pre>
<p>The resulting table is exactly the same as <code>table1</code> and
now the rate of infection can be calculated rather
straightforwardly.</p>
<div class="sourceCode" id="cb127"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span><span class="va">table2_fixed</span>, rate <span class="op">=</span> <span class="va">cases</span> <span class="op">/</span> <span class="va">population</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 6 × 5</span></span>
<span class="co">##   country      year  cases population      rate</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Afghanistan  <span style="text-decoration: underline;">1</span>999    745   19<span style="text-decoration: underline;">987</span>071 0.000<span style="text-decoration: underline;">037</span>3</span>
<span class="co">## <span style="color: #BCBCBC;">2</span> Afghanistan  <span style="text-decoration: underline;">2</span>000   <span style="text-decoration: underline;">2</span>666   20<span style="text-decoration: underline;">595</span>360 0.000<span style="text-decoration: underline;">129</span> </span>
<span class="co">## <span style="color: #BCBCBC;">3</span> Brazil       <span style="text-decoration: underline;">1</span>999  <span style="text-decoration: underline;">37</span>737  172<span style="text-decoration: underline;">006</span>362 0.000<span style="text-decoration: underline;">219</span> </span>
<span class="co">## <span style="color: #BCBCBC;">4</span> Brazil       <span style="text-decoration: underline;">2</span>000  <span style="text-decoration: underline;">80</span>488  174<span style="text-decoration: underline;">504</span>898 0.000<span style="text-decoration: underline;">461</span> </span>
<span class="co">## <span style="color: #BCBCBC;">5</span> China        <span style="text-decoration: underline;">1</span>999 <span style="text-decoration: underline;">212</span>258 <span style="text-decoration: underline;">1</span>272<span style="text-decoration: underline;">915</span>272 0.000<span style="text-decoration: underline;">167</span> </span>
<span class="co">## <span style="color: #BCBCBC;">6</span> China        <span style="text-decoration: underline;">2</span>000 <span style="text-decoration: underline;">213</span>766 <span style="text-decoration: underline;">1</span>280<span style="text-decoration: underline;">428</span>583 0.000<span style="text-decoration: underline;">167</span></span></code></pre>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="splitting-columns">Splitting columns<a class="anchor" aria-label="anchor" href="#splitting-columns"></a>
</h2>
<p>Table 3 contains an example of a column that contains multiple
values. It is a somewhat convoluted example but occasionally you may
come across data like this.</p>
<div class="sourceCode" id="cb129"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">table3</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 6 × 3</span></span>
<span class="co">##   country      year rate             </span>
<span class="co">## <span style="color: #BCBCBC;">*</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            </span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Afghanistan  <span style="text-decoration: underline;">1</span>999 745/19987071     </span>
<span class="co">## <span style="color: #BCBCBC;">2</span> Afghanistan  <span style="text-decoration: underline;">2</span>000 2666/20595360    </span>
<span class="co">## <span style="color: #BCBCBC;">3</span> Brazil       <span style="text-decoration: underline;">1</span>999 37737/172006362  </span>
<span class="co">## <span style="color: #BCBCBC;">4</span> Brazil       <span style="text-decoration: underline;">2</span>000 80488/174504898  </span>
<span class="co">## <span style="color: #BCBCBC;">5</span> China        <span style="text-decoration: underline;">1</span>999 212258/1272915272</span>
<span class="co">## <span style="color: #BCBCBC;">6</span> China        <span style="text-decoration: underline;">2</span>000 213766/1280428583</span></code></pre>
<p>The <code>rate</code> column contains both the number of TB cases and
the population separated by a ‘/’ character. The rate column is a
character type so not terribly useful for doing anything of a
mathematical nature in its current guise.</p>
<div class="section level3">
<h3 id="separate">
<code>separate()</code><a class="anchor" aria-label="anchor" href="#separate"></a>
</h3>
<p>The <code><a href="https://tidyr.tidyverse.org/reference/separate.html" class="external-link">separate()</a></code> function allows us to split a character
column into multiple columns based on a delimiter or separator.</p>
<div class="sourceCode" id="cb131"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">table3_separated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tidyr.tidyverse.org/reference/separate.html" class="external-link">separate</a></span><span class="op">(</span><span class="va">table3</span>, <span class="va">rate</span>, into <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cases"</span>, <span class="st">"population"</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span>
<span class="va">table3_separated</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 6 × 4</span></span>
<span class="co">##   country      year cases  population</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Afghanistan  <span style="text-decoration: underline;">1</span>999 745    19987071  </span>
<span class="co">## <span style="color: #BCBCBC;">2</span> Afghanistan  <span style="text-decoration: underline;">2</span>000 2666   20595360  </span>
<span class="co">## <span style="color: #BCBCBC;">3</span> Brazil       <span style="text-decoration: underline;">1</span>999 37737  172006362 </span>
<span class="co">## <span style="color: #BCBCBC;">4</span> Brazil       <span style="text-decoration: underline;">2</span>000 80488  174504898 </span>
<span class="co">## <span style="color: #BCBCBC;">5</span> China        <span style="text-decoration: underline;">1</span>999 212258 1272915272</span>
<span class="co">## <span style="color: #BCBCBC;">6</span> China        <span style="text-decoration: underline;">2</span>000 213766 1280428583</span></code></pre>
<p>The <code>sep</code> argument takes a regular expression that defines
how to split the values. We’ve mentioned regular expressions before –
they are a language for specifying search patterns used to find
sequences of characters within text and well worth learning. In this
case our separator is just the ‘/’ character.</p>
<p>The resulting data frame is still not quite what we want though. This
becomes apparent as soon as we try to do anything with the new
<code>cases</code> and <code>population</code> columns.</p>
<div class="sourceCode" id="cb133"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span><span class="va">table3_separated</span>, rate <span class="op">=</span> <span class="va">cases</span> <span class="op">/</span> <span class="va">population</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #BBBB00; font-weight: bold;">Error</span><span style="font-weight: bold;"> in `mutate()`:</span></span>
<span class="co">## <span style="color: #BBBB00;">!</span> Problem while computing `rate = cases/population`.</span>
<span class="co">## <span style="font-weight: bold;">Caused by error in `cases / population`:</span></span>
<span class="co">## <span style="color: #BBBB00;">!</span> non-numeric argument to binary operator</span></code></pre>
<p>By default the separated values are character types. We could convert
these using <code>mutate</code> with <code>across</code>.</p>
<div class="sourceCode" id="cb135"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span><span class="va">table3_separated</span>, <span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html" class="external-link">across</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">cases</span>, <span class="va">population</span><span class="op">)</span>, <span class="va">as.integer</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 6 × 4</span></span>
<span class="co">##   country      year  cases population</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Afghanistan  <span style="text-decoration: underline;">1</span>999    745   19<span style="text-decoration: underline;">987</span>071</span>
<span class="co">## <span style="color: #BCBCBC;">2</span> Afghanistan  <span style="text-decoration: underline;">2</span>000   <span style="text-decoration: underline;">2</span>666   20<span style="text-decoration: underline;">595</span>360</span>
<span class="co">## <span style="color: #BCBCBC;">3</span> Brazil       <span style="text-decoration: underline;">1</span>999  <span style="text-decoration: underline;">37</span>737  172<span style="text-decoration: underline;">006</span>362</span>
<span class="co">## <span style="color: #BCBCBC;">4</span> Brazil       <span style="text-decoration: underline;">2</span>000  <span style="text-decoration: underline;">80</span>488  174<span style="text-decoration: underline;">504</span>898</span>
<span class="co">## <span style="color: #BCBCBC;">5</span> China        <span style="text-decoration: underline;">1</span>999 <span style="text-decoration: underline;">212</span>258 <span style="text-decoration: underline;">1</span>272<span style="text-decoration: underline;">915</span>272</span>
<span class="co">## <span style="color: #BCBCBC;">6</span> China        <span style="text-decoration: underline;">2</span>000 <span style="text-decoration: underline;">213</span>766 <span style="text-decoration: underline;">1</span>280<span style="text-decoration: underline;">428</span>583</span></code></pre>
<p>But another option is to specify <code>convert = TRUE</code> when
carrying out the <code><a href="https://tidyr.tidyverse.org/reference/separate.html" class="external-link">separate()</a></code> operation, in which case it
will deduce the type of the values and convert the column
accordingly.</p>
<div class="sourceCode" id="cb137"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">table3_separated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tidyr.tidyverse.org/reference/separate.html" class="external-link">separate</a></span><span class="op">(</span><span class="va">table3</span>, <span class="va">rate</span>, into <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cases"</span>, <span class="st">"population"</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"/"</span>, convert <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span><span class="va">table3_separated</span>, rate <span class="op">=</span> <span class="va">cases</span> <span class="op">/</span> <span class="va">population</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 6 × 5</span></span>
<span class="co">##   country      year  cases population      rate</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> Afghanistan  <span style="text-decoration: underline;">1</span>999    745   19<span style="text-decoration: underline;">987</span>071 0.000<span style="text-decoration: underline;">037</span>3</span>
<span class="co">## <span style="color: #BCBCBC;">2</span> Afghanistan  <span style="text-decoration: underline;">2</span>000   <span style="text-decoration: underline;">2</span>666   20<span style="text-decoration: underline;">595</span>360 0.000<span style="text-decoration: underline;">129</span> </span>
<span class="co">## <span style="color: #BCBCBC;">3</span> Brazil       <span style="text-decoration: underline;">1</span>999  <span style="text-decoration: underline;">37</span>737  172<span style="text-decoration: underline;">006</span>362 0.000<span style="text-decoration: underline;">219</span> </span>
<span class="co">## <span style="color: #BCBCBC;">4</span> Brazil       <span style="text-decoration: underline;">2</span>000  <span style="text-decoration: underline;">80</span>488  174<span style="text-decoration: underline;">504</span>898 0.000<span style="text-decoration: underline;">461</span> </span>
<span class="co">## <span style="color: #BCBCBC;">5</span> China        <span style="text-decoration: underline;">1</span>999 <span style="text-decoration: underline;">212</span>258 <span style="text-decoration: underline;">1</span>272<span style="text-decoration: underline;">915</span>272 0.000<span style="text-decoration: underline;">167</span> </span>
<span class="co">## <span style="color: #BCBCBC;">6</span> China        <span style="text-decoration: underline;">2</span>000 <span style="text-decoration: underline;">213</span>766 <span style="text-decoration: underline;">1</span>280<span style="text-decoration: underline;">428</span>583 0.000<span style="text-decoration: underline;">167</span></span></code></pre>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="example-1-metabric-gene-expression">Example 1: METABRIC gene expression<a class="anchor" aria-label="anchor" href="#example-1-metabric-gene-expression"></a>
</h2>
<p>Although tables 1 to 4 contain real data they are, of course, ‘toy’
data frames created for demonstration and teaching purposes. We’ll now
turn our attention to the METABRIC expression data and see how this
needs to be transformed into a tidier format to open up different
avenues for exploring the data.</p>
<p>We’ll first load the table and then select just the columns we’re
going to need.</p>
<div class="sourceCode" id="cb139"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">metabric_selected</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="st">"https://zenodo.org/record/6450144/files/metabric_clinical_and_expression_data.csv"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">Patient_ID</span>, <span class="va">ER_status</span>, <span class="va">ESR1</span><span class="op">:</span><span class="va">MLPH</span><span class="op">)</span>
<span class="va">metabric_selected</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 1,904 × 10</span></span>
<span class="co">##    Patient_ID ER_status  ESR1 ERBB2   PGR  TP53 PIK3CA GATA3 FOXA1  MLPH</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> MB-0000    Positive   8.93  9.33  5.68  6.34   5.70  6.93  7.95  9.73</span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> MB-0002    Positive  10.0   9.73  7.51  6.19   5.76 11.3  11.8  12.5 </span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> MB-0005    Positive  10.0   9.73  7.38  6.40   6.75  9.29 11.7  10.3 </span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> MB-0006    Positive  10.4  10.3   6.82  6.87   7.22  8.67 11.9  10.5 </span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> MB-0008    Positive  11.3   9.96  7.33  6.34   5.82  9.72 11.6  12.2 </span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> MB-0010    Positive  11.2   9.74  5.95  5.42   6.12  9.79 12.1  11.4 </span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> MB-0014    Positive  10.8   9.28  7.72  5.99   7.48  8.37 11.5  10.8 </span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> MB-0022    Positive  10.4   8.61  5.59  6.17   7.59  7.87 10.7   9.95</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> MB-0028    Positive  12.5  10.7   5.33  6.22   6.25 10.3  12.1  10.9 </span>
<span class="co">## <span style="color: #BCBCBC;">10</span> MB-0035    Positive   7.54 11.5   5.59  6.41   5.99 10.2  12.8  13.5 </span>
<span class="co">## <span style="color: #949494;"># … with 1,894 more rows</span></span></code></pre>
<p>When we first looked at visualization using ggplot2 we created the
following box plot.</p>
<div class="sourceCode" id="cb141"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">metabric_selected</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ER_status</span>, y <span class="op">=</span> <span class="va">GATA3</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="week4_files/figure-html/box_plot_1-1.png" width="480"></p>
<p>But what if we would like to create a series of plots using the
faceting functions in ggplot2 with one plot for each gene?</p>
<p>Faceting requires a categorical variable, which is used to divide the
data into subsets to be used for each plot. In this case we’d need a
gene column. Clearly our data are not structured in this way.</p>
<p>We have gene names for column headings. Are these variables? Well,
maybe, although a more correct name for each of these variables or
column headings would be ‘Expression of ESR1’, ‘Expression of ERBB2’,
etc.</p>
<p>But we could consider that these gene column headings are actually
values of a variable called ‘gene’ or ‘gene symbol’. In this regard,
what we have is a ‘wide format’ table.</p>
<p>Most gene expression matrices have a similar form, although usually
there have rows for each gene and columns for each sample. It should be
said that the gene expression matrix format is a very compact way of
representing the data which could be a consideration when dealing with
tens of thousands of genes and anywhere between a few tens of samples to
a few thousand, such is the case for METABRIC.</p>
<p>Furthermore, there are lots of tools for working with gene expression
data in the form of matrices, including many packages in the
Bioconductor project. Fortunately, as we’ve seen,
<code><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer()</a></code> and <code><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html" class="external-link">pivot_wider()</a></code> provide a
very convenient means of converting between tidy and matrix-like
formats.</p>
<p>We’ll convert our table of ER status and gene expression data to the
tidy format.</p>
<div class="sourceCode" id="cb142"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">metabric_long</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span><span class="va">metabric_selected</span>, <span class="va">ESR1</span><span class="op">:</span><span class="va">MLPH</span>, names_to <span class="op">=</span> <span class="st">"Gene"</span>, values_to <span class="op">=</span> <span class="st">"Expression"</span><span class="op">)</span>
<span class="va">metabric</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 1,904 × 32</span></span>
<span class="co">##    Patient_ID Cohort Age_at_diagnosis Survival_time Survival_status Vital_status</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       </span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> MB-0000         1             75.6         140.  LIVING          Living      </span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> MB-0002         1             43.2          84.6 LIVING          Living      </span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> MB-0005         1             48.9         164.  DECEASED        Died of Dis…</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> MB-0006         1             47.7         165.  LIVING          Living      </span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> MB-0008         1             77.0          41.4 DECEASED        Died of Dis…</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> MB-0010         1             78.8           7.8 DECEASED        Died of Dis…</span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> MB-0014         1             56.4         164.  LIVING          Living      </span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> MB-0022         1             89.1          99.5 DECEASED        Died of Oth…</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> MB-0028         1             86.4          36.6 DECEASED        Died of Oth…</span>
<span class="co">## <span style="color: #BCBCBC;">10</span> MB-0035         1             84.2          36.3 DECEASED        Died of Dis…</span>
<span class="co">## <span style="color: #949494;"># … with 1,894 more rows, and 26 more variables: Chemotherapy &lt;chr&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   Radiotherapy &lt;chr&gt;, Tumour_size &lt;dbl&gt;, Tumour_stage &lt;dbl&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   Neoplasm_histologic_grade &lt;dbl&gt;, Lymph_nodes_examined_positive &lt;dbl&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   Lymph_node_status &lt;dbl&gt;, Cancer_type &lt;chr&gt;, ER_status &lt;chr&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   PR_status &lt;chr&gt;, HER2_status &lt;chr&gt;, HER2_status_measured_by_SNP6 &lt;chr&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   PAM50 &lt;chr&gt;, `3-gene_classifier` &lt;chr&gt;, Nottingham_prognostic_index &lt;dbl&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   Cellularity &lt;chr&gt;, Integrative_cluster &lt;chr&gt;, Mutation_count &lt;dbl&gt;, …</span></span></code></pre>
<p>Note how we specified a range of columns between <code>ESR1</code>
and <code>MLPH</code>, which is a lot easier than naming each column
individually.</p>
<p>We’re now in a position to create our faceted box plot chart.</p>
<div class="sourceCode" id="cb144"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric_long</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ER_status</span>, y <span class="op">=</span> <span class="va">Expression</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html" class="external-link">vars</a></span><span class="op">(</span><span class="va">Gene</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="week4_files/figure-html/box_plot_2-1.png" width="700"></p>
<p>In carrying out this transformation, the observational unit has
changed. The tidy format has <em>one-row-per-gene-per-sample</em>, while
wide format was <em>one-row-per-sample</em>. The tidy format is much
less compact and involves considerable duplication of values in the
first three columns (<code>Patient_ID</code>, <code>ER_status</code> and
<code>Gene</code>).</p>
<p>One of the other plot types we’ve used in exploring these data was a
scatter plot comparing the expression of two genes across all the
samples. For this, the <em>one-row-per-sample</em> representation is the
more appropriate and being able to convert back to this format allows us
to create the plot.</p>
<div class="sourceCode" id="cb145"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">metabric_long</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html" class="external-link">pivot_wider</a></span><span class="op">(</span>names_from <span class="op">=</span> <span class="st">"Gene"</span>, values_from <span class="op">=</span> <span class="st">"Expression"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span>, colour <span class="op">=</span> <span class="va">ER_status</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="week4_files/figure-html/scatter_plot_1-1.png" width="700"></p>
<hr>
</div>
<div class="section level2">
<h2 id="example-2-protein-levels-in-mcf-7-after-treatment-with-tamoxifen">Example 2: Protein levels in MCF-7 after treatment with
tamoxifen<a class="anchor" aria-label="anchor" href="#example-2-protein-levels-in-mcf-7-after-treatment-with-tamoxifen"></a>
</h2>
<p>Our second real example features another data set generated by CRUK
CI scientists (<a href="https://pubmed.ncbi.nlm.nih.gov/29899353" class="external-link">Papachristou <em>et
al.</em>, Nature Communications 9:2311, 2018</a>) in which the dynamics
of endogenous chromatin-associated protein complexes were investigated
using quantitative mass spectrometry.</p>
<p>We’ll look at just one of several tabular data sets made available as
supplementary data, which contains the total level of protein in MCF-7
cells at various time points after treatment with the drug tamoxifen.
MCF-7 is a breast cancer cell line isolated in 1970 from a 69-year old
woman and established for use in breast cancer research by the Michigan
Cancer Foundation-7 institute in Detroit. Tamoxifen is a hormone therapy
used in the treatment of estrogen receptor-positive breast cancer.</p>
<p>The table in question is <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5998130/bin/41467_2018_4619_MOESM11_ESM.xlsx" class="external-link">supplementary
data 9</a>.</p>
<div class="sourceCode" id="cb146"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://readxl.tidyverse.org" class="external-link">readxl</a></span><span class="op">)</span>
<span class="co"># because it's an Excel file located on the web need to download then import</span>
<span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://zenodo.org/record/6450144/files/41467_2018_4619_MOESM11_ESM.xlsx"</span>, destfile <span class="op">=</span> <span class="st">"41467_2018_4619_MOESM11_ESM.xlsx"</span><span class="op">)</span>
<span class="va">protein_levels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readxl.tidyverse.org/reference/read_excel.html" class="external-link">read_excel</a></span><span class="op">(</span><span class="st">"41467_2018_4619_MOESM11_ESM.xlsx"</span>, skip <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">protein_levels</span>, <span class="va">`Uniprot Accession`</span>, <span class="va">`Gene Name`</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">ends_with</a></span><span class="op">(</span><span class="st">"rep01"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 7,943 × 6</span></span>
<span class="co">##    `Uniprot Accession` `Gene Name` vehicle.rep01 tam.2h.rep01 tam.6h.rep01</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> Q09666              AHNAK                17.3         17.3         17.3</span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> P05787              KRT8                 16.9         17.1         17.0</span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> Q15149              PLEC                 16.7         16.7         16.7</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> P21333              FLNA                 16.6         16.5         16.6</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> P05783              KRT18                16.4         16.6         16.5</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> P49327              FASN                 16.0         16.0         16.0</span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> P10809              HSPD1                16.1         16.2         16.1</span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> Q14204              DYNC1H1              16.0         15.9         16.0</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> O75369              FLNB                 16.1         16.1         16.2</span>
<span class="co">## <span style="color: #BCBCBC;">10</span> P35579              MYH9                 15.7         15.6         15.8</span>
<span class="co">## <span style="color: #949494;"># … with 7,933 more rows, and 1 more variable: tam.24h.rep01 &lt;dbl&gt;</span></span></code></pre>
<p>This has a very similar structure to a gene expression matrix having
one row for each protein (or gene) and a column for each sample. We’ve
only shown columns for the first of four replicates for each group
defined by a treatment and a time point. The control group is, in
effect, the ‘untreated’ group in which the cells are treated with the
vehicle (ethanol) alone.</p>
<p>If we wanted to calculate the mean protein levels within each group,
i.e. the average level for the protein measured in the 4 replicates, or
we wanted to show the spread of values for the replicates as a box plot,
then the data as currently structured is not in the most suitable form.
In what follows, we’ll transform the table to allow us to do both these
analyses.</p>
<p>To simplify matters, we’re going to focus on just a few proteins,
those whose levels are markedly reduced 24 hours after treatment with
tamoxifen compared with the vehicle.</p>
<div class="sourceCode" id="cb148"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">protein_levels</span> <span class="op">&lt;-</span> <span class="va">protein_levels</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">`log2FC(24h/veh)`</span> <span class="op">&lt;</span> <span class="op">-</span><span class="fl">0.75</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span>accession <span class="op">=</span> <span class="va">`Uniprot Accession`</span>, gene <span class="op">=</span> <span class="va">`Gene Name`</span>, <span class="va">vehicle.rep01</span><span class="op">:</span><span class="va">tam.24h.rep04</span><span class="op">)</span>
<span class="va">protein_levels</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 10 × 18</span></span>
<span class="co">##    accession gene    vehicle.rep01 tam.2h.rep01 tam.6h.rep01 tam.24h.rep01</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> Q4ZG55    GREB1           12.1         12           11.6           9.78</span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> P15514    AREG            11.4         11.3         11.3           9.92</span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> Q13433    SLC39A6         10.9         10.8         10.7           9.89</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> Q9NPD8    UBE2T           10.4         10.3         10.2           9.62</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> Q6N021    TET2            10.1         10.0          9.59          8.75</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> P06401    PGR              9.97        10.0          9.75          8.75</span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> P25929    NPY1R            9.08         9.16         8.89          7.94</span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> O95084    PRSS23           8.45         8.53         8.27          7.78</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> P08493    MGP              6.76         6.76         6.18          5.09</span>
<span class="co">## <span style="color: #BCBCBC;">10</span> Q96BR1    SGK3             6.33         6.42         6.16          5.5 </span>
<span class="co">## <span style="color: #949494;"># … with 12 more variables: vehicle.rep02 &lt;dbl&gt;, tam.2h.rep02 &lt;dbl&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   tam.6h.rep02 &lt;dbl&gt;, tam.24h.rep02 &lt;dbl&gt;, vehicle.rep03 &lt;dbl&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   tam.2h.rep03 &lt;dbl&gt;, tam.6h.rep03 &lt;dbl&gt;, tam.24h.rep03 &lt;dbl&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   vehicle.rep04 &lt;dbl&gt;, tam.2h.rep04 &lt;dbl&gt;, tam.6h.rep04 &lt;dbl&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   tam.24h.rep04 &lt;dbl&gt;</span></span></code></pre>
<p>This is a fairly typical example of a table with columns for each
sample where the sample names contain quite a lot of information, in
this case:</p>
<ul>
<li>the treatment</li>
<li>the time at which the protein levels are measured after
treatment</li>
<li>the number of the replicate sample for that treatment and time
point</li>
</ul>
<p>To make use of this information we need to pivot the table such that
the sample name is in a column and protein levels are in another column,
and then to split the sample name column into its constituent parts.</p>
<p>For the first part of this transformation we’ll use
<code><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer()</a></code>.</p>
<div class="sourceCode" id="cb150"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">protein_levels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span><span class="va">protein_levels</span>, <span class="va">vehicle.rep01</span><span class="op">:</span><span class="va">tam.24h.rep04</span>, names_to <span class="op">=</span> <span class="st">"sample"</span>, values_to <span class="op">=</span> <span class="st">"protein_level"</span><span class="op">)</span>
<span class="va">protein_levels</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 160 × 4</span></span>
<span class="co">##    accession gene  sample        protein_level</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                 <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> Q4ZG55    GREB1 vehicle.rep01         12.1 </span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> Q4ZG55    GREB1 tam.2h.rep01          12   </span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> Q4ZG55    GREB1 tam.6h.rep01          11.6 </span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> Q4ZG55    GREB1 tam.24h.rep01          9.78</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> Q4ZG55    GREB1 vehicle.rep02         11.3 </span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> Q4ZG55    GREB1 tam.2h.rep02          11.4 </span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> Q4ZG55    GREB1 tam.6h.rep02          11.1 </span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> Q4ZG55    GREB1 tam.24h.rep02          8.98</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> Q4ZG55    GREB1 vehicle.rep03         11.0 </span>
<span class="co">## <span style="color: #BCBCBC;">10</span> Q4ZG55    GREB1 tam.2h.rep03          11.0 </span>
<span class="co">## <span style="color: #949494;"># … with 150 more rows</span></span></code></pre>
<p>Now we can use <code><a href="https://tidyr.tidyverse.org/reference/separate.html" class="external-link">separate()</a></code> to disentangle the components
of the sample name.</p>
<p>It looks like we need to split on the ‘.’ character but that has a
special meaning in a regular expression, i.e. match <em>any</em>
character. Fortunately, the default regular expression used by
<code><a href="https://tidyr.tidyverse.org/reference/separate.html" class="external-link">separate()</a></code> splits on any character that isn’t a letter or a
number, so will do just fine.</p>
<p>But there is another problem. The vehicle sample names don’t follow
the pattern of “treatment.time.replicate”. In actual fact the vehicle
measurements were taken 24 hours after treatment with ethanol alone.
What we should do to correct matters is add the ‘24h’ component. For
that, we’re going to use one of several really useful string
manipulation functions provided by the stringr package.</p>
<div class="rmdblock">
<p>
<strong>Some useful stringr functions</strong>
</p>
<p>
<strong><code>str_c(…, sep = ““)</code></strong>
</p>
<p>
Join multiple strings into a single string.
</p>
<p>
<code>str_c(“tidyverse”, “fab”, sep = ” is “)</code>
</p>
<p>
<code>[1] “tidyverse is fab”</code>
</p>
<p>
<code>str_c(c(“Ashley”, “Matt”), c(“hiking”, “beer”), sep = ” loves
“)</code>
</p>
<p>
<code>[1] “Ashley loves hiking” “Matt loves beer”</code>
</p>
<hr>
<p>
<strong><code>str_replace(string, pattern, replacement)</code></strong>
</p>
<p>
<strong><code>str_replace_all(string, pattern,
replacement)</code></strong>
</p>
<p>
Substitute a matched pattern in a string (or character vector).
</p>
<p>
<code>str_replace(“Oscar is the best cat in the world”, “best”,
“loveliest”)</code>
</p>
<p>
<code>[1] “Oscar is the loveliest cat in the world”</code>
</p>
<p>
<code>str_replace_all(“the cat sat on the mat”, “at”, “AT”)</code>
</p>
<p>
<code>[1] “the cAT sAT on the mAT”</code>
</p>
<hr>
<p>
<strong><code>str_remove(string, pattern)</code></strong>
</p>
<p>
<strong><code>str_remove_all(string, pattern)</code></strong>
</p>
<p>
Remove matched patterns in a string.
</p>
<p>
Alias for <code>str_replace(string, pattern, ““)</code> and
<code>str_replace_all(string, pattern,”“)</code>.
</p>
<hr>
<p>
<strong><code>str_sub(string, start, end)</code></strong>
</p>
<p>
Extract substrings from a character vector at the given start and end
positions.
</p>
<p>
<code>str_sub(“Matthew”, start = 1, end = 4)</code>
</p>
<p>
<code>[1] “Matt”</code>
</p>
<p>
<code>str_sub(c(“tamoxifen”, “vehicle”), 1, 3)</code>
</p>
<p>
<code>[1] “tam” “veh”</code>
</p>
</div>
<p><code><a href="https://stringr.tidyverse.org/reference/str_replace.html" class="external-link">str_replace()</a></code> looks like the function we need so let’s
see how it works on our sample names.</p>
<div class="sourceCode" id="cb152"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_replace.html" class="external-link">str_replace</a></span><span class="op">(</span><span class="st">"vehicle.rep01"</span>, <span class="st">"vehicle"</span>, <span class="st">"vehicle.24h"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "vehicle.24h.rep01"</span></code></pre>
<p>The first argument is the character vector we’re working on, in this
case a single character string. The second argument is the pattern or
substring we want to replace and the third is the string we want to
replace it with.</p>
<p>Looking at the help we can see that, like very many R functions, it
works on a vector of character values (or strings), so let’s try this on
a few of our sample names, say the first 10.</p>
<div class="sourceCode" id="cb154"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_replace.html" class="external-link">str_replace</a></span><span class="op">(</span><span class="va">protein_levels</span><span class="op">$</span><span class="va">sample</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>, <span class="st">"vehicle"</span>, <span class="st">"vehicle.24h"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##  [1] "vehicle.24h.rep01" "tam.2h.rep01"      "tam.6h.rep01"     </span>
<span class="co">##  [4] "tam.24h.rep01"     "vehicle.24h.rep02" "tam.2h.rep02"     </span>
<span class="co">##  [7] "tam.6h.rep02"      "tam.24h.rep02"     "vehicle.24h.rep03"</span>
<span class="co">## [10] "tam.2h.rep03"</span></code></pre>
<p>Finally, we can modify the values in our data frame using
<code><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate()</a></code>.</p>
<div class="sourceCode" id="cb156"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">protein_levels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span><span class="va">protein_levels</span>, sample <span class="op">=</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_replace.html" class="external-link">str_replace</a></span><span class="op">(</span><span class="va">sample</span>, <span class="st">"vehicle"</span>, <span class="st">"vehicle.24h"</span><span class="op">)</span><span class="op">)</span>
<span class="va">protein_levels</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 160 × 4</span></span>
<span class="co">##    accession gene  sample            protein_level</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> Q4ZG55    GREB1 vehicle.24h.rep01         12.1 </span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> Q4ZG55    GREB1 tam.2h.rep01              12   </span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> Q4ZG55    GREB1 tam.6h.rep01              11.6 </span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> Q4ZG55    GREB1 tam.24h.rep01              9.78</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> Q4ZG55    GREB1 vehicle.24h.rep02         11.3 </span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> Q4ZG55    GREB1 tam.2h.rep02              11.4 </span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> Q4ZG55    GREB1 tam.6h.rep02              11.1 </span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> Q4ZG55    GREB1 tam.24h.rep02              8.98</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> Q4ZG55    GREB1 vehicle.24h.rep03         11.0 </span>
<span class="co">## <span style="color: #BCBCBC;">10</span> Q4ZG55    GREB1 tam.2h.rep03              11.0 </span>
<span class="co">## <span style="color: #949494;"># … with 150 more rows</span></span></code></pre>
<p>Now we’re ready to separate the <code>sample</code> column into its
component parts.</p>
<div class="sourceCode" id="cb158"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">protein_levels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tidyr.tidyverse.org/reference/separate.html" class="external-link">separate</a></span><span class="op">(</span><span class="va">protein_levels</span>, <span class="va">sample</span>, into <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"treatment"</span>, <span class="st">"time"</span>, <span class="st">"replicate"</span><span class="op">)</span><span class="op">)</span>
<span class="va">protein_levels</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 160 × 6</span></span>
<span class="co">##    accession gene  treatment time  replicate protein_level</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> Q4ZG55    GREB1 vehicle   24h   rep01             12.1 </span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> Q4ZG55    GREB1 tam       2h    rep01             12   </span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> Q4ZG55    GREB1 tam       6h    rep01             11.6 </span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> Q4ZG55    GREB1 tam       24h   rep01              9.78</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> Q4ZG55    GREB1 vehicle   24h   rep02             11.3 </span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> Q4ZG55    GREB1 tam       2h    rep02             11.4 </span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> Q4ZG55    GREB1 tam       6h    rep02             11.1 </span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> Q4ZG55    GREB1 tam       24h   rep02              8.98</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> Q4ZG55    GREB1 vehicle   24h   rep03             11.0 </span>
<span class="co">## <span style="color: #BCBCBC;">10</span> Q4ZG55    GREB1 tam       2h    rep03             11.0 </span>
<span class="co">## <span style="color: #949494;"># … with 150 more rows</span></span></code></pre>
<p>The groups we want to compare are in fact the combination of the
treatment and the time point. We’ll create a new column called ‘group’
in which we concatenate the treatment and time values using another
stringr function, <code><a href="https://stringr.tidyverse.org/reference/str_c.html" class="external-link">str_c()</a></code>. But we’ll also shorten ‘vehicle’
to ‘veh’ so we have similar length treatment labels (‘veh’ and ‘tam’) in
the plot we’ll eventually create. For that we’ll use
<code><a href="https://stringr.tidyverse.org/reference/str_sub.html" class="external-link">str_sub()</a></code>. Our group variable is categorical so we’ll
convert this to a factor while we’re at it.</p>
<div class="sourceCode" id="cb160"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">protein_levels</span> <span class="op">&lt;-</span> <span class="va">protein_levels</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>group <span class="op">=</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_sub.html" class="external-link">str_sub</a></span><span class="op">(</span><span class="va">treatment</span>, <span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>group <span class="op">=</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_c.html" class="external-link">str_c</a></span><span class="op">(</span><span class="va">group</span>, <span class="va">time</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>group <span class="op">=</span> <span class="fu"><a href="https://forcats.tidyverse.org/reference/as_factor.html" class="external-link">as_factor</a></span><span class="op">(</span><span class="va">group</span><span class="op">)</span><span class="op">)</span>
<span class="va">protein_levels</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 160 × 7</span></span>
<span class="co">##    accession gene  treatment time  replicate protein_level group </span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> </span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> Q4ZG55    GREB1 vehicle   24h   rep01             12.1  veh24h</span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> Q4ZG55    GREB1 tam       2h    rep01             12    tam2h </span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> Q4ZG55    GREB1 tam       6h    rep01             11.6  tam6h </span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> Q4ZG55    GREB1 tam       24h   rep01              9.78 tam24h</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> Q4ZG55    GREB1 vehicle   24h   rep02             11.3  veh24h</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> Q4ZG55    GREB1 tam       2h    rep02             11.4  tam2h </span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> Q4ZG55    GREB1 tam       6h    rep02             11.1  tam6h </span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> Q4ZG55    GREB1 tam       24h   rep02              8.98 tam24h</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> Q4ZG55    GREB1 vehicle   24h   rep03             11.0  veh24h</span>
<span class="co">## <span style="color: #BCBCBC;">10</span> Q4ZG55    GREB1 tam       2h    rep03             11.0  tam2h </span>
<span class="co">## <span style="color: #949494;"># … with 150 more rows</span></span></code></pre>
<p>Computing the mean protein levels within each group is now very
straightforward.</p>
<div class="sourceCode" id="cb162"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">protein_levels</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">accession</span>, <span class="va">gene</span>, <span class="va">group</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>protein_level <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">protein_level</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## `summarise()` has grouped output by 'accession', 'gene'. You can override using</span>
<span class="co">## the `.groups` argument.</span></code></pre>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 40 × 4</span></span>
<span class="co">## <span style="color: #949494;"># Groups:   accession, gene [10]</span></span>
<span class="co">##    accession gene   group  protein_level</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> O95084    PRSS23 veh24h          8.16</span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> O95084    PRSS23 tam2h           8.2 </span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> O95084    PRSS23 tam6h           8.03</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> O95084    PRSS23 tam24h          7.20</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> P06401    PGR    veh24h          9.44</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> P06401    PGR    tam2h           9.66</span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> P06401    PGR    tam6h           9.41</span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> P06401    PGR    tam24h          8.39</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> P08493    MGP    veh24h          5.59</span>
<span class="co">## <span style="color: #BCBCBC;">10</span> P08493    MGP    tam2h           5.69</span>
<span class="co">## <span style="color: #949494;"># … with 30 more rows</span></span></code></pre>
<p>Plotting the protein levels in each group as box plots with each data
point overlaid is only slightly more involved. We can use faceting to
create separate plots for each of the proteins (we’ll use gene symbols
as they’re more recognizable).</p>
<div class="sourceCode" id="cb165"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">protein_levels</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">group</span>, y <span class="op">=</span> <span class="va">protein_level</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"grey60"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html" class="external-link">geom_jitter</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">0.15</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html" class="external-link">vars</a></span><span class="op">(</span><span class="va">gene</span><span class="op">)</span>, scales <span class="op">=</span> <span class="st">"free_y"</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></code></pre></div>
<p><img src="week4_files/figure-html/box_plot_3-1.png" width="700"></p>
<hr>
</div>
<div class="section level2">
<h2 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<p>In this session we have covered the following:</p>
<ul>
<li>Computing summary values for groups of observations</li>
<li>Counting the numbers of observations within categories</li>
<li>Combining data from two tables through join operations</li>
<li>Restructuring data frames using pivoting operations</li>
<li>Splitting compound variables into separate columns</li>
<li>Modifying character columns using various string manipulation
functions</li>
</ul>
</div>
<div class="section level2">
<h2 id="assignment">Assignment<a class="anchor" aria-label="anchor" href="#assignment"></a>
</h2>
<p><a href="data:text/x-markdown;base64,
---
title: "Assignment 4"
---

This week's assignment will involve summarizing values for groups of
observations in the METABRIC data set and combining clinical and mutation data
to perform exploratory analysis involving both.

The mutation data come from targeted sequencing of 2433 primary tumours from the
METABRIC study using a panel of 173 of the most frequently mutated breast cancer
genes.

[Pereira *et al.*, Nature Communications 7:11479, 2016](https://www.ncbi.nlm.nih.gov/pubmed/27161491)

We'll start by loading the two tables we're going to need. The first is the
clinical data that you should by now be quite familiar with. This table does not
contain the additional mRNA expression columns we've used in previous
assignments.

```{r eval = FALSE, message = FALSE, warning = FALSE}
library(tidyverse)
metabric <- read_csv("https://zenodo.org/record/6450144/files/metabric_clinical_data.csv")
# convert the Integrative_cluster variable into a categorical variable with
# levels in the correct order
metabric <- mutate(metabric, Integrative_cluster = factor(Integrative_cluster, levels = c("1", "2", "3", "4ER-", "4ER+", "5", "6", "7", "8", "9", "10")))
```

The second table contains the mutations that were detected in each patient
tumour sample.

```{r eval = FALSE}
mutations <- read_csv("https://zenodo.org/record/6450144/files/metabric_mutations.csv")
mutations
```

The METABRIC cohort contains 2509 patients but not all patient tumour samples
were sequenced and for some patient samples that were sequenced, no mutations
were detected. For one of the exercises we need to know which patients were
sequenced, so we'll also read the case list of samples that were sequenced and
add a column indicating this in the clinical data table.

```{r eval = FALSE}
cases_sequenced <- read_tsv("https://zenodo.org/record/6450144/files/cases_nat_comm_2016.txt", col_names = "Patient_ID")
metabric <- mutate(metabric, Sequenced = Patient_ID %in% cases_sequenced$Patient_ID)
count(metabric, Sequenced)
```

The clinical data, gene expression values and mutation data were all downloaded
from [cBioPortal](https://www.cbioportal.org/study/summary?id=brca_metabric).

---

**1. Compare the average survival time and other attributes between ER-negative and ER-positive patients**

Compute the average survival time for the ER-negative and ER-positive groups.
Note that such a comparison only makes sense for those patients that are
deceased so apply the appropriate filter first. Add a column for the number
of patients in each group.

```{r}

```

Compute the average tumour size, number of positive lymph nodes and Nottingham
prognostic index within ER-negative and ER-positive patients.

```{r}

```

Compute the proportion of tumour samples that are ER positive within each of
PAM50 groups.

```{r}

```

---

**2. Find the top 10 most frequently mutated genes**

Count the numbers of mutations for each gene and display the top 10 most
frequently mutated genes.

```{r}

```

Some genes contain multiple mutations within a single tumour sample so we will
be over-counting the numbers of samples containing a mutated gene. Modify your
calculation to prevent this. The count should be the number of patient samples
in which the gene is mutated.

```{r}

```

_Hint: use the `distinct()` function_

We're usually only interested in non-silent mutations, i.e. those that may have
some functional effect. One of the findings of the METABRIC study is that 7
genes harbour coding mutations in at least 10% of the samples. A total of 2433
patient samples were sequenced.

Calculate the percentages of samples containing non-silent mutations for each
gene and filter those genes mutated in more than 10% of samples. Compare your
gene list and percentages with those given in the METABRIC 2016 publication
(second paragraph in the Results section).

```{r}

```

---

**3. Find mutation hotspots in PIK3CA**

PIK3CA and TP53 are the most frequently mutated genes in breast cancers. But do
the mutations occur randomly within these genes or are they usually found in
specific locations?

Find the most frequently mutated codons in PIK3CA.

```{r}

```

These frequently mutated loci are known as "hotspots".

What type of mutations are found at the most prevalent hotspot in PIK3CA, codon
1047? What changes are occurring at the DNA and protein level?

```{r}

```

_Hint: take a look at what columns are available in the `mutations` table and count the values_

---

**4. Compare the prevalance of TP53 mutations across the Integrative Cluster breast cancer subtypes**

Figure 5a from the METABRIC 2016 paper compares the percentage of samples that
have non-silent mutations for various genes within each of the Integrative
Clusters. In this exercise, we'll reproduce a version of one of these bar charts
for the TP53 gene.

First, we'll need to distinguish between those patients that have a non-silent
TP53 mutation and those that don't. Create a data frame containing the number
of non-silent TP53 mutations in each patient sample. It should have two columns,
`Patient_ID` and `TP53_mutation_count`.

```{r}

```

Now join this table to the `metabric` data frame so that the latter has the
added `TP53_mutation_count` column.

```{r}

```

Create a new column called `Has_TP53_mutation` containing just `TRUE` or `FALSE`
values, i.e. a logical variable.

```{r}

```

_Hint: use `is.na()` function_

Compute the proportion of patient samples within each Integrative Cluster with a
TP53 mutation. Exclude those samples that have not been sequenced and those that
have not been classified into one of the Integrative Cluster subtypes.

```{r}

```

Finally, plot these percentages as a bar chart using `geom_col()`. Look at the
help page for `geom_bar` and `geom_col` to see why we'd want to use the latter
in this case.

```{r}

```


Compare the plot with the equivalent plot in figure 5a from the METABRIC
mutation paper. Do the bars look in roughly the correct proportions? Note that
only 9 of the integrative clusters are shown in figure 5a.

---

**5. Compute the tumour suppressor gene (TSG) score for each gene**

The tumour suppressor gene (TSG) score is the proportion of inactivating
mutations for each gene out of the total number of mutations for that gene.

Inactivating mutations are nonsense SNVs, frameshift substitutions and variants
that affected splice sites.

Compute the TSG score for each gene and find the top 10 tumour suppressor genes,
i.e. those with the highest TSG scores.

```{r}
inactivating_mutation_types <- c("Nonsense_Mutation", "Frame_Shift_Del", "Frame_Shift_Ins", "Splice_Region", "Splice_Site")

```

_Hint: use `%in%` to determine if a mutation is an inactivating mutation_

**Extension**

Download Supplementary Data 3 and create a scatter plot comparing the TSG
scores in the `allSamples_TSGScore` column with those you've calculated.

```{r}

```


If you want more exercises, see here https://raw.githubusercontent.com/PMacDaSci/R4CancerSci/master/assignments/assignment4_extra.Rmd" download="assignment4.Rmd">Download assignment4.Rmd</a></p>
<hr>
</div>
<div class="section level2">
<h2 id="customizing-plots-with-ggplot2-extra">Customizing plots with ggplot2 (EXTRA)<a class="anchor" aria-label="anchor" href="#customizing-plots-with-ggplot2-extra"></a>
</h2>
<p>Optional section. Do this if you’ve completed the previous sections
and the Assignment and want more. In this section, we’ll turn our
attention back to visualization using ggplot2 and how we can customize
our plots by adding or changing titles and labels, changing the scales
used on the x and y axes, and choosing colours.</p>
<div class="section level3">
<h3 id="titles-and-labels">Titles and labels<a class="anchor" aria-label="anchor" href="#titles-and-labels"></a>
</h3>
<p>Adding titles and subtitles to a plot and changing the x- and y-axis
labels is very straightforward using the <code><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs()</a></code>
function.</p>
<div class="sourceCode" id="cb166"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># use the metabric data we started with</span>
<span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span>, colour <span class="op">=</span> <span class="va">ER_status</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.6</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"lm"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>
    title <span class="op">=</span> <span class="st">"mRNA expression in the METABRIC breast cancer data set"</span>,
    subtitle <span class="op">=</span> <span class="st">"GATA3 and ESR1 mRNA expression measured using Illumina HT-12 bead arrays"</span>,
    x <span class="op">=</span> <span class="st">"log2 GATA3 expression"</span>,
    y <span class="op">=</span> <span class="st">"log2 ESR1 expression"</span>,
    colour <span class="op">=</span> <span class="st">"ER status"</span>
  <span class="op">)</span>
<span class="va">plot</span></code></pre></div>
<pre><code><span class="co">## `geom_smooth()` using formula 'y ~ x'</span></code></pre>
<p><img src="week4_files/figure-html/scatter_plot-1.png" width="700"></p>
<p>The labels are another component of the plot object that we’ve
constructed, along with aesthetic mappings and layers (geoms). The plot
object is a list and contains various elements including those mappings
and layers and one element named <code>labels</code>.</p>
<p><code><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs()</a></code> is a simple function for creating a list of
labels you want to specify as name-value pairs as in the above example.
You can name any aesthetic (in this case x and y) to override the
default values (the column names) and you can add a title, subtitle and
caption if you wish. In addition to changing the x- and y-axis labels,
we also removed the underscore from the legend title by setting the
label for the colour aesthetic.</p>
</div>
<div class="section level3">
<h3 id="scales">Scales<a class="anchor" aria-label="anchor" href="#scales"></a>
</h3>
<p>Take a look at the x and y scales in the above plot. ggplot2 has
chosen the x and y scales and where to put breaks and ticks.</p>
<p>Let’s have a look at the elements of the list object we created that
specifies how the plot should be displayed.</p>
<div class="sourceCode" id="cb168"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">plot</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "data"        "layers"      "scales"      "mapping"     "theme"      </span>
<span class="co">## [6] "coordinates" "facet"       "plot_env"    "labels"</span></code></pre>
<p>One of the components of the plot is called <code>scales</code>.
ggplot2 automatically adds default scales behind the scenes equivalent
to the following:</p>
<div class="sourceCode" id="cb170"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span>, colour <span class="op">=</span> <span class="va">ER_status</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.6</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"lm"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_colour_discrete.html" class="external-link">scale_colour_discrete</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>Note that we have three aesthetics and ggplot2 adds a scale for
each.</p>
<div class="sourceCode" id="cb171"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plot</span><span class="op">$</span><span class="va">mapping</span></code></pre></div>
<pre><code><span class="co">## Aesthetic mapping: </span>
<span class="co">## * `x`      -&gt; `GATA3`</span>
<span class="co">## * `y`      -&gt; `ESR1`</span>
<span class="co">## * `colour` -&gt; `ER_status`</span></code></pre>
<p>The x and y variables (<code>GATA3</code> and <code>ESR1</code>) are
continuous so ggplot2 adds a continuous scale for each.
<code>ER_status</code> is a discrete variable in this case so ggplot2
adds a discrete scale for colour.</p>
<p>Generalizing, the scales that are required follow the naming
scheme:</p>
<pre><code>scale_&lt;NAME_OF_AESTHETIC&gt;_&lt;NAME_OF_SCALE&gt;</code></pre>
<p>Look at the help page for <code>scale_y_continuous</code> to see what
we can change about the y-axis scale.</p>
<p>First we’ll change the breaks, i.e. where ggplot2 puts ticks and
numeric labels, on the y axis.</p>
<div class="sourceCode" id="cb174"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span>, colour <span class="op">=</span> <span class="va">ER_status</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.6</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"lm"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">15</span>, by <span class="op">=</span> <span class="fl">2.5</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## `geom_smooth()` using formula 'y ~ x'</span></code></pre>
<p><img src="week4_files/figure-html/scatter_plot_2-1.png" width="700"></p>
<p><code><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq()</a></code> is a useful function for generating regular
sequences of numbers. In this case we wanted numbers from 5 to 15 going
up in steps of 2.5.</p>
<div class="sourceCode" id="cb176"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">15</span>, by <span class="op">=</span> <span class="fl">2.5</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1]  5.0  7.5 10.0 12.5 15.0</span></code></pre>
<p>We could do the same thing for the x axis using
<code><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous()</a></code>.</p>
<p>We can also adjust the extents of the x or y axis.</p>
<div class="sourceCode" id="cb178"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span>, colour <span class="op">=</span> <span class="va">ER_status</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.6</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"lm"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">15</span>, by <span class="op">=</span> <span class="fl">2.5</span><span class="op">)</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">12</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## `geom_smooth()` using formula 'y ~ x'</span></code></pre>
<pre><code><span class="co">## Warning: Removed 160 rows containing non-finite values (stat_smooth).</span></code></pre>
<pre><code><span class="co">## Warning: Removed 160 rows containing missing values (geom_point).</span></code></pre>
<p><img src="week4_files/figure-html/scatter_plot_3-1.png" width="700"></p>
<p>Here, just for demonstration purposes, we set the upper limit to be
less than the largest values of ESR1 expression and ggplot2 warned us
that some rows have been removed from the plot.</p>
<p>We can change the minor breaks, e.g. to add more lines that act as
guides. These are shown as thin white lines when using the default
theme.</p>
<div class="sourceCode" id="cb182"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span>, colour <span class="op">=</span> <span class="va">ER_status</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.6</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"lm"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">12.5</span>, by <span class="op">=</span> <span class="fl">2.5</span><span class="op">)</span>, minor_breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">13.5</span>, <span class="fl">0.5</span><span class="op">)</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">13.5</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## `geom_smooth()` using formula 'y ~ x'</span></code></pre>
<p><img src="week4_files/figure-html/scatter_plot_4-1.png" width="700"></p>
<p>Or we can remove the minor breaks entirely.</p>
<div class="sourceCode" id="cb184"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span>, colour <span class="op">=</span> <span class="va">ER_status</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.6</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"lm"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">6</span>, <span class="fl">14</span>, by <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>, minor_breaks <span class="op">=</span> <span class="cn">NULL</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">13.5</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## `geom_smooth()` using formula 'y ~ x'</span></code></pre>
<p><img src="week4_files/figure-html/scatter_plot_5-1.png" width="700"></p>
<p>Similarly we could remove all breaks entirely.</p>
<div class="sourceCode" id="cb186"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span>, colour <span class="op">=</span> <span class="va">ER_status</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.6</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"lm"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## `geom_smooth()` using formula 'y ~ x'</span></code></pre>
<p><img src="week4_files/figure-html/scatter_plot_6-1.png" width="700"></p>
<p>A more typical scenario would be to keep the breaks, because we want
to display the ticks and their lables, but remove the grid lines.
Somewhat confusingly the position of grid lines are controlled by a
scale but preventing these from being displayed requires changing the
theme. The theme controls the way in which non-data components are
displayed – we’ll look at how these can be customized later. For now,
though, here’s an example of turning off the display of all grid lines
for major and minor breaks for both axes.</p>
<div class="sourceCode" id="cb188"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span>, colour <span class="op">=</span> <span class="va">ER_status</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.6</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"lm"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">14</span>, by <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">14</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>panel.grid <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## `geom_smooth()` using formula 'y ~ x'</span></code></pre>
<p><img src="week4_files/figure-html/scatter_plot_7-1.png" width="700"></p>
<p>By default, the scales are expanded by 5% of the range on either
side. We can add or reduce the space as follows.</p>
<div class="sourceCode" id="cb190"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span>, colour <span class="op">=</span> <span class="va">ER_status</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.6</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"lm"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html" class="external-link">expand_scale</a></span><span class="op">(</span>mult <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html" class="external-link">expand_scale</a></span><span class="op">(</span>mult <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Warning: `expand_scale()` is deprecated; use `expansion()` instead.</span>

<span class="co">## Warning: `expand_scale()` is deprecated; use `expansion()` instead.</span></code></pre>
<pre><code><span class="co">## `geom_smooth()` using formula 'y ~ x'</span></code></pre>
<p><img src="week4_files/figure-html/scatter_plot_8-1.png" width="700"></p>
<p>Here we only added 1% (0.01) of the range of GATA3 expression values
on either side along the x axis but we added 25% (0.25) of the range of
ESR1 expression on either side along the y axis.</p>
<p>We can move the axis to the other side of the plot –- not sure why
you’d want to do this but with ggplot2 just about anything is
possible.</p>
<div class="sourceCode" id="cb193"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span>, colour <span class="op">=</span> <span class="va">ER_status</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.6</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"lm"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>position <span class="op">=</span> <span class="st">"top"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## `geom_smooth()` using formula 'y ~ x'</span></code></pre>
<p><img src="week4_files/figure-html/scatter_plot_9-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="colours">Colours<a class="anchor" aria-label="anchor" href="#colours"></a>
</h3>
<p>The colour asthetic is used with a categorical variable,
<code>ER_status</code>, in the scatter plots we’ve been customizing. The
default colour scale used by ggplot2 for categorical variables is
<code>scale_colour_discrete</code>. We can manually set the colours we
wish to use using <code>scale_colour_manual</code> instead.</p>
<div class="sourceCode" id="cb195"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span>, colour <span class="op">=</span> <span class="va">ER_status</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.6</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"lm"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dodgerblue2"</span>, <span class="st">"firebrick2"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## `geom_smooth()` using formula 'y ~ x'</span></code></pre>
<p><img src="week4_files/figure-html/scatter_plot_10-1.png" width="700"></p>
<p>Setting colours manually is ok when we only have two or three
categories but when we have a larger number it would be handy to be able
to choose from a selection of carefully-constructed colour palettes.
Helpfully, ggplot2 provides access to the <a href="http://colorbrewer2.org" class="external-link">ColorBrewer</a> palettes through the
functions <code><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_colour_brewer()</a></code> and
<code><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_brewer()</a></code>.</p>
<div class="sourceCode" id="cb197"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span>, colour <span class="op">=</span> <span class="va">`3-gene_classifier`</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.6</span>, alpha <span class="op">=</span> <span class="fl">0.5</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span></code></pre></div>
<p><img src="week4_files/figure-html/scatter_plot_11-1.png" width="700"></p>
<p>Look at the help page for <code>scale_colour_brewer</code> to see
what other colour palettes are available and visit the <a href="http://colorbrewer2.org" class="external-link">ColorBrewer</a> website to see what these
look like.</p>
<p>Interestingly, you can set other attributes other than just the
colours at the same time.</p>
<div class="sourceCode" id="cb198"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span>, colour <span class="op">=</span> <span class="va">ER_status</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.6</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"lm"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dodgerblue2"</span>, <span class="st">"firebrick2"</span><span class="op">)</span>, labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ER-negative"</span>, <span class="st">"ER-positive"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span>  <span class="co"># remove legend title for colour now that the labels are self-explanatory</span></code></pre></div>
<pre><code><span class="co">## `geom_smooth()` using formula 'y ~ x'</span></code></pre>
<p><img src="week4_files/figure-html/scatter_plot_12-1.png" width="700"></p>
<p>We have applied our own set of mappings from levels in the data to
aesthetic values.</p>
<p>For continuous variables we may wish to be able to change the colours
used in the colour gradient. To demonstrate this we’ll correct the
Nottingham prognostic index (NPI) values and use this to colour points
in the scatter plot of ESR1 vs GATA3 expression on a continuous
scale.</p>
<div class="sourceCode" id="cb200"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Nottingham_prognostic_index is incorrectly calculated in the data downloaded from cBioPortal</span>
<span class="va">metabric</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span><span class="va">metabric</span>, Nottingham_prognostic_index <span class="op">=</span> <span class="fl">0.02</span> <span class="op">*</span> <span class="va">Tumour_size</span> <span class="op">+</span> <span class="va">Lymph_node_status</span> <span class="op">+</span> <span class="va">Neoplasm_histologic_grade</span><span class="op">)</span>
<span class="co">#</span>
<span class="va">metabric</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">Nottingham_prognostic_index</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span>, colour <span class="op">=</span> <span class="va">Nottingham_prognostic_index</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></code></pre></div>
<p><img src="week4_files/figure-html/scatter_plot_13-1.png" width="700"></p>
<p>Higher NPI scores correspond to worse prognosis and lower chance of 5
year survival. We’ll emphasize those points on the scatter plot by
adjusting our colour scale.</p>
<div class="sourceCode" id="cb201"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">metabric</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">Nottingham_prognostic_index</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span>, colour <span class="op">=</span> <span class="va">Nottingham_prognostic_index</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_colour_gradient</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"white"</span>, high <span class="op">=</span> <span class="st">"firebrick2"</span><span class="op">)</span></code></pre></div>
<p><img src="week4_files/figure-html/scatter_plot_14-1.png" width="700"></p>
<p>In some cases it might make sense to specify two colour gradients
either side of a mid-point.</p>
<div class="sourceCode" id="cb202"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">metabric</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">Nottingham_prognostic_index</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span>, colour <span class="op">=</span> <span class="va">Nottingham_prognostic_index</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_colour_gradient2</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"dodgerblue1"</span>, mid <span class="op">=</span> <span class="st">"grey90"</span>, high <span class="op">=</span> <span class="st">"firebrick1"</span>, midpoint <span class="op">=</span> <span class="fl">4.5</span><span class="op">)</span></code></pre></div>
<p><img src="week4_files/figure-html/scatter_plot_15-1.png" width="700"></p>
<p>As before we can override the default labels and other aspects of the
colour scale within the scale function.</p>
<div class="sourceCode" id="cb203"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">metabric</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">Nottingham_prognostic_index</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span>, colour <span class="op">=</span> <span class="va">Nottingham_prognostic_index</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_colour_gradient</a></span><span class="op">(</span>
    low <span class="op">=</span> <span class="st">"lightblue"</span>, high <span class="op">=</span> <span class="st">"darkblue"</span>,
    name <span class="op">=</span> <span class="st">"NPI"</span>,
    breaks <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">6</span>,
    limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1.5</span>, <span class="fl">6.5</span><span class="op">)</span>
  <span class="op">)</span></code></pre></div>
<p><img src="week4_files/figure-html/scatter_plot_16-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="themes">Themes<a class="anchor" aria-label="anchor" href="#themes"></a>
</h3>
<p>Themes can be used to customize non-data components of a plot. Let’s
create a plot showing the expression of estrogen receptor alpha (ESR1)
for each of the Integrative cluster breast cancer subtypes.</p>
<div class="sourceCode" id="cb204"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># read in the METABRIC data, convert the Integrative_cluster variable into a</span>
<span class="co"># categorical variable with the levels in the correct order, and select just</span>
<span class="co"># the columns and rows we're going to use</span>
<span class="va">metabric</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="st">"https://zenodo.org/record/6450144/files/metabric_clinical_and_expression_data.csv"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>Integrative_cluster <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">Integrative_cluster</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1"</span>, <span class="st">"2"</span>, <span class="st">"3"</span>, <span class="st">"4ER-"</span>, <span class="st">"4ER+"</span>, <span class="st">"5"</span>, <span class="st">"6"</span>, <span class="st">"7"</span>, <span class="st">"8"</span>, <span class="st">"9"</span>, <span class="st">"10"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>`3-gene_classifier` <span class="op">=</span> <span class="fu"><a href="https://tidyr.tidyverse.org/reference/replace_na.html" class="external-link">replace_na</a></span><span class="op">(</span><span class="va">`3-gene_classifier`</span>, <span class="st">"Unclassified"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">Patient_ID</span>, <span class="va">ER_status</span>, <span class="va">PR_status</span>, <span class="va">`3-gene_classifier`</span>, <span class="va">Integrative_cluster</span>, <span class="va">ESR1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">Integrative_cluster</span><span class="op">)</span>, <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">ESR1</span><span class="op">)</span><span class="op">)</span>
<span class="va">metabric</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 1,904 × 6</span></span>
<span class="co">##    Patient_ID ER_status PR_status `3-gene_classifier`   Integrative_clust…  ESR1</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                 <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> MB-0000    Positive  Negative  ER-/HER2-             4ER+                8.93</span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> MB-0002    Positive  Positive  ER+/HER2- High Prolif 4ER+               10.0 </span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> MB-0005    Positive  Positive  Unclassified          3                  10.0 </span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> MB-0006    Positive  Positive  Unclassified          9                  10.4 </span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> MB-0008    Positive  Positive  ER+/HER2- High Prolif 9                  11.3 </span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> MB-0010    Positive  Positive  ER+/HER2- High Prolif 7                  11.2 </span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> MB-0014    Positive  Positive  Unclassified          3                  10.8 </span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> MB-0022    Positive  Negative  Unclassified          3                  10.4 </span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> MB-0028    Positive  Negative  ER+/HER2- High Prolif 9                  12.5 </span>
<span class="co">## <span style="color: #BCBCBC;">10</span> MB-0035    Positive  Negative  ER+/HER2- High Prolif 3                   7.54</span>
<span class="co">## <span style="color: #949494;"># … with 1,894 more rows</span></span></code></pre>
<div class="sourceCode" id="cb206"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># plot the ESR1 expression for each integrative cluster</span>
<span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Integrative_cluster</span>, y <span class="op">=</span> <span class="va">ESR1</span>, fill <span class="op">=</span> <span class="va">Integrative_cluster</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Integrative cluster"</span>, y <span class="op">=</span> <span class="st">"ESR1 expression"</span><span class="op">)</span>
<span class="va">plot</span></code></pre></div>
<p><img src="week4_files/figure-html/box_plot_4-1.png" width="700"></p>
<p>The default theme has the characteristic grey background which isn’t
particularly suitable for printing on paper. We can change to one of a
number of alternative themes available in the ggplot2 package, e.g. the
black and white theme.</p>
<div class="sourceCode" id="cb207"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plot</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="week4_files/figure-html/box_plot_5-1.png" width="700"></p>
<p>The help page for the available themes, which can be accessed using
<code><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">?theme_bw</a></code>, lists each and when you might want to use them.
It states that <code>black_bw</code> may work better for presentations
displayed on a projector.</p>
<p>Each of these themes is really just a collection of attributes
relating to how various non-data elements of the plot will be displayed.
We can override any of these individual settings using the
<code><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme()</a></code> function. A look at the help page
(<code><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">?theme</a></code>) shows that there are a very large number of
settings that you can change. The following example demonstrates a few
of these.</p>
<div class="sourceCode" id="cb208"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plot</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>
    panel.grid.major.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,
    axis.ticks.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,
    legend.position <span class="op">=</span> <span class="st">"none"</span>
  <span class="op">)</span></code></pre></div>
<p><img src="week4_files/figure-html/box_plot_6-1.png" width="700"></p>
<p>Here’s another example that also involves customizing the labels,
scales and colours.</p>
<div class="sourceCode" id="cb209"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">`3-gene_classifier`</span>, fill <span class="op">=</span> <span class="va">ER_status</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">700</span><span class="op">)</span>, breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">700</span>, <span class="fl">100</span><span class="op">)</span>, expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html" class="external-link">expand_scale</a></span><span class="op">(</span>mult <span class="op">=</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"firebrick2"</span>, <span class="st">"dodgerblue2"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="cn">NULL</span>, y <span class="op">=</span> <span class="st">"samples"</span>, fill <span class="op">=</span> <span class="st">"ER status"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>
    panel.border <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,
    panel.grid <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,
    axis.ticks.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,
    axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,
    axis.line.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_line</a></span><span class="op">(</span><span class="op">)</span>,
    axis.ticks.length.y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">0.2</span>, <span class="st">"cm"</span><span class="op">)</span>,
    legend.position <span class="op">=</span> <span class="st">"bottom"</span>
  <span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Warning: `expand_scale()` is deprecated; use `expansion()` instead.</span></code></pre>
<p><img src="week4_files/figure-html/bar_plot_1-1.png" width="480"></p>
<p>The ggthemes package contains some extra themes and might be fun to
check out. Here’s an example of a plot that uses the
<code>theme_gdocs</code> theme that resembles the default look of charts
in Google Docs.</p>
<div class="sourceCode" id="cb211"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jrnold/ggthemes" class="external-link">ggthemes</a></span><span class="op">)</span>
<span class="va">metabric</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">`3-gene_classifier`</span> <span class="op">==</span> <span class="st">"HER2+"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">PR_status</span>, y <span class="op">=</span> <span class="va">ESR1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html" class="external-link">geom_jitter</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">PR_status</span><span class="op">)</span>, width <span class="op">=</span> <span class="fl">0.25</span>, alpha <span class="op">=</span> <span class="fl">0.4</span>, show.legend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"PR status"</span>, y <span class="op">=</span> <span class="st">"ESR1 expression"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggthemes/man/theme_gdocs.html" class="external-link">theme_gdocs</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="week4_files/figure-html/box_plot_7-1.png" width="480"></p>
</div>
<div class="section level3">
<h3 id="position-adjustments">Position adjustments<a class="anchor" aria-label="anchor" href="#position-adjustments"></a>
</h3>
<p>All geoms in ggplot2 have a position adjustment that can be set using
the <code>position</code> argument. This has different effects for
different types of plot but essentially this resolves how overlapping
geoms are displayed.</p>
<p>For example, let’s consider the stacked bar plot we created earlier
showing the numbers of patients in each of the 3-gene classifications
subdivided by ER status. The default position value for
<code><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar()</a></code> is “stack” which is why the plot is shown as a
stacked bar chart. An alternative way of representing these data would
be to show separate bars for each ER status side-by-side by setting
<code>position = "dodge"</code>.</p>
<div class="sourceCode" id="cb212"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">`3-gene_classifier`</span>, fill <span class="op">=</span> <span class="va">ER_status</span><span class="op">)</span>, position <span class="op">=</span> <span class="st">"dodge"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">700</span><span class="op">)</span>, breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">700</span>, <span class="fl">100</span><span class="op">)</span>, expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html" class="external-link">expand_scale</a></span><span class="op">(</span>mult <span class="op">=</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"firebrick2"</span>, <span class="st">"dodgerblue2"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="cn">NULL</span>, y <span class="op">=</span> <span class="st">"samples"</span>, fill <span class="op">=</span> <span class="st">"ER status"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>
    panel.border <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,
    panel.grid <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,
    axis.ticks.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,
    axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,
    axis.line.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_line</a></span><span class="op">(</span><span class="op">)</span>,
    axis.ticks.length.y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">0.2</span>, <span class="st">"cm"</span><span class="op">)</span>
  <span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Warning: `expand_scale()` is deprecated; use `expansion()` instead.</span></code></pre>
<p><img src="week4_files/figure-html/bar_plot_2-1.png" width="700"></p>
<p>Another position adjustment we’ve come across is
<code><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html" class="external-link">geom_jitter()</a></code>, which is just a convenient shortcut for
<code>geom_point(position = "jitter")</code>. A variation on this,
<code><a href="https://ggplot2.tidyverse.org/reference/position_jitterdodge.html" class="external-link">position_jitterdodge()</a></code>, comes in handy when we are
overlaying points on top of a box plot. We show an example of just such
a plot in which first use <code>postion = "jitter"</code>.</p>
<div class="sourceCode" id="cb214"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">metabric</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">`3-gene_classifier`</span>, y <span class="op">=</span> <span class="va">ESR1</span>, colour <span class="op">=</span> <span class="va">PR_status</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>position <span class="op">=</span> <span class="st">"jitter"</span>, size <span class="op">=</span> <span class="fl">0.5</span>, alpha <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"3-gene classification"</span>, y <span class="op">=</span> <span class="st">"ESR1 expression"</span>, colour <span class="op">=</span> <span class="st">"PR status"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_color_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>
    panel.grid.major.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,
    axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,
    axis.ticks.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>
  <span class="op">)</span></code></pre></div>
<p><img src="week4_files/figure-html/box_plot_8-1.png" width="700"></p>
<p>The PR-negative and PR-positive points have distinct colours but are
overlapping in a way that is aesthetically displeasing. What we want is
for the points to have both jitter and to be dodged in the same way as
the boxes. With <code><a href="https://ggplot2.tidyverse.org/reference/position_jitterdodge.html" class="external-link">position_jitterdodge()</a></code> we get a better
looking plot.</p>
<div class="sourceCode" id="cb215"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">metabric</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">`3-gene_classifier`</span>, y <span class="op">=</span> <span class="va">ESR1</span>, colour <span class="op">=</span> <span class="va">PR_status</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>position <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/position_jitterdodge.html" class="external-link">position_jitterdodge</a></span><span class="op">(</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">0.5</span>, alpha <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"3-gene classification"</span>, y <span class="op">=</span> <span class="st">"ESR1 expression"</span>, colour <span class="op">=</span> <span class="st">"PR status"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_color_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>
    panel.grid.major.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,
    axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,
    axis.ticks.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>
  <span class="op">)</span></code></pre></div>
<p><img src="week4_files/figure-html/box_plot_9-1.png" width="700"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Maria Doyle.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
