<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Session 1 -- Introduction to RStudio • R4CancerSci</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Session 1 -- Introduction to RStudio">
<meta property="og:description" content="R4CancerSci">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">R4CancerSci</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Material
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/week0.html">Week 0 - Getting ready</a>
    </li>
    <li>
      <a href="../articles/week1.html">Week 1 - Introduction to RStudio</a>
    </li>
    <li>
      <a href="../articles/week2.html">Week 2 - Data visualization with ggplot2</a>
    </li>
    <li>
      <a href="../articles/week3.html">Week 3 - Data manipulation using dplyr</a>
    </li>
    <li>
      <a href="../articles/week4.html">Week 4 - Grouping, combining and restructuring data for analysis</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/mblue9/Bite-size-R-v2" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Session 1 – Introduction to RStudio</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mblue9/Bite-size-R-v2/blob/HEAD/vignettes/week1.Rmd" class="external-link"><code>vignettes/week1.Rmd</code></a></small>
      <div class="hidden name"><code>week1.Rmd</code></div>

    </div>

    
    
<blockquote>
<h4 id="aims">Aims</h4>
<ul>
<li>To introduce RStudio which we will be using to write R code in this
course</li>
<li>To demonstrate generating a visualisation and report</li>
<li>To review some basic R commands</li>
</ul>
<h4 id="learning-objectives">Learning objectives</h4>
<ul>
<li>Describe the main panes in RStudio</li>
<li>Generate a visualisation and report using R</li>
<li>Edit code to update visualisations and reports</li>
</ul>
</blockquote>
<div class="section level2">
<h2 id="introduction-to-r-and-rstudio">Introduction to R and RStudio<a class="anchor" aria-label="anchor" href="#introduction-to-r-and-rstudio"></a>
</h2>
</div>
<div class="section level2">
<h2 id="why-learn-r">Why learn R?<a class="anchor" aria-label="anchor" href="#why-learn-r"></a>
</h2>
<ul>
<li><p>R involves creating &amp; using scripts which makes the steps you
used in your analysis clear and can be inspected by someone else for
feedback and error-checking.</p></li>
<li><p>R code is great for reproducibility. An increasing number of
journals and funding agencies expect analyses to be reproducible, so
knowing R will give you an edge with these requirements.</p></li>
<li><p>R integrates with other tools to generate manuscripts from your
code. This document (RMarkdown a .Rmd file) is a case in point.</p></li>
<li><p>R is interdisciplinary and extensible and has thousands of
installable packages to extend its capabilities. R has packages for
image analysis, GIS, time series, population genetics, and a lot
more.</p></li>
<li><p>R scales well to work on data of all shapes and sizes.</p></li>
<li><p>R can connect to spreadsheets, databases, and many other data
formats, on your computer or on the web.</p></li>
<li><p>R produces high-quality graphics suitable for publication in
journals or the web.</p></li>
<li><p>R has a large and welcoming community - Thousands use R daily and
many of them are willing to help you through websites such as Stack
Overflow or the RStudio community.</p></li>
<li><p>Not only is R free, but it is also open-source and
cross-platform.</p></li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 id="rstudio-a-brief-tour">RStudio a brief tour<a class="anchor" aria-label="anchor" href="#rstudio-a-brief-tour"></a>
</h2>
<p>RStudio provides us with a friendly interface to the R statistical
programming language. It consists of four main “Panes”. These can be
re-sized and moved around to suit how you like to work.</p>
<p><img src="images/Rstudioataglance.png" width="100%"></p>
<div class="section level3">
<h3 id="editing-pane">Editing pane<a class="anchor" aria-label="anchor" href="#editing-pane"></a>
</h3>
<p>By default the top left-hand pane is one for creating, editing &amp;
running R scripts.</p>
<p><img src="images/Rstudio_scripting.png" width="100%"></p>
<p>A script is an R program that you have written. A good practice is
for that script to perform only one role in your analysis workflow and
so you may have several R scripts which you call, in a particular
sequence, to analyse your data.</p>
<p>As you will see, a script is basically a text file that contains R
commands and (ideally) comments to explain what the codes function is
(as a documentation process).</p>
<p>As well as R scripts, there are many types of RStudio document
including Markdown files which we will use in the teaching of this
course. These can provide interactive reports, workbooks or pdf and web
documents to name but a few possible outcomes.</p>
</div>
<div class="section level3">
<h3 id="console">Console<a class="anchor" aria-label="anchor" href="#console"></a>
</h3>
<p>Coming down the screen to the bottom left-hand pane we find the
console window. This is where we can find output produced by running our
R scripts.</p>
<p><img src="images/Rstudio_console.png" width="100%"></p>
<p>We can also try out snippets of R code here. Those of you who have
only used graphical interfaces like Windows or MacOS where you click on
commands using a mouse may find this aspect of R somewhat different. We
type in commands to R using the command line.</p>
<p>This area can also be used like a calculator. Let’s just type in
something like <code>23 + 45</code> followed by the return key and see
what happens. You should get the following:</p>
<pre><code>&gt; 23 + 45
[1] 68</code></pre>
<p>Now 68 is clearly the answer but what is that 1 in brackets?</p>
<p>Here is another example to explain. If we type <code>1:36</code> and
press enter, what happens? R generates output counting from 1 to 36 but
cannot fit all the output on one line and so starts another like
this:</p>
<pre><code>&gt; 1:36
 [1]  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25
[26] 26 27 28 29 30 31 32 33 34 35 36</code></pre>
<p>Now we have two lines beginning with a number in square brackets.
Note that the number of values displayed on each line may differ on your
computer; it largely depends on the width of your console pane and the
font size. Try creating a larger sequence of numbers,
e.g. <code>1:100</code>, if all 36 numbers fit on a single line in your
case.</p>
<p>This is just R helping us to keep tabs on which number we are looking
at. <code>[1]</code> denotes that the line starts with the first result
and the <code>[26]</code> denotes that this line starts with the 26th
number. Let’s try another one and generate a sequence incrementing in
steps of 2:</p>
<pre><code>&gt; 1:36 * 2
 [1]  2  4  6  8 10 12 14 16 18 20 22 24 26 28 30 32 34 36 38 40 42 44 46 48 50
[26] 52 54 56 58 60 62 64 66 68 70 72</code></pre>
<p>There are other tabs on this pane but we shall not be covering these
on this course.</p>
</div>
<div class="section level3">
<h3 id="environment">Environment<a class="anchor" aria-label="anchor" href="#environment"></a>
</h3>
<p>Next we move to the top right-hand corner pane. Here we have even
more tabs (of which we will only consider two Environment and
History).</p>
<p><img src="images/Rstudio_console.png" width="100%"></p>
<p>Environment keeps track on R variables which we create (more on those
later) and their contents. History is like a tally role of all the R
commands we have entered in our session.</p>
</div>
<div class="section level3">
<h3 id="files-plots-help">Files, Plots &amp; Help<a class="anchor" aria-label="anchor" href="#files-plots-help"></a>
</h3>
<p>Our final bottom right-hand pane also has several tabs. The Files tab
is a file explorer that enable us to move around our directories and
select which files we wish to work on. We can also change the default
working directory that Rstudio will use.</p>
<p><img src="images/Rstudio_filexpl.png" width="100%"></p>
<p>The Plots tab is where any graphs that we create in R will appear. We
can move through them using the arrow buttons and the export button will
convert them to different graphics formats e.g. for publication in a
paper or for the web.</p>
<p>The Packages tab shows which R packages are installed (These expand
R’s functionallity and again will be covered later) and can also install
new packages.</p>
<p>The Help tab is a <em>massively useful</em> tab which enables you to
search R help index to get help pages on R functions and provide example
code to help you use them in your R scripts.</p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="our-first-look-at-an-r-markdown-report">Our first look at an R Markdown report<a class="anchor" aria-label="anchor" href="#our-first-look-at-an-r-markdown-report"></a>
</h2>
<p>Our overall goal for this course is to give you the ability to import
your data into R, format it, select a subset of the data most of
interest for a given analysis, carry out an analysis to summarize these
data, create visualizations of the data and generate a reproducible
report.</p>
<p>So let’s see an example of what we’ll be learning and generate our
first visualisation and report. By the end of this course we should be
able to explain what each line of the code does.</p>
<p><a href="data:text/x-markdown;base64,LS0tCnRpdGxlOiAiUHJvdGVpbiBsZXZlbHMgaW4gYnJlYXN0IGNhbmNlciBjZWxscyB0cmVhdGVkIHdpdGggdGFtb3hpZmVuIgphdXRob3I6ICJNYXJpYSBEb3lsZSIKb3V0cHV0OiBodG1sX2RvY3VtZW50Ci0tLQoKIyMgSW50cm9kdWN0aW9uCgpXaGljaCBwcm90ZWlucyBjaGFuZ2UgaW4gYWJ1bmRhbmNlIGluIHRoZSBicmVhc3QgY2FuY2VyIGNlbGwgbGluZSwgTUNGLTcsIHdpdGggdGFtb3hpZmVuIHRyZWF0bWVudD8gSG93IGRvIHRoZWlyIGxldmVscyBjaGFuZ2Ugd2l0aCB0cmVhdG1lbnQgdGltZT8gQW5zd2VyaW5nIHRoZXNlIHF1ZXN0aW9ucyAoYXQgYSBoaWdoIGxldmVsKSBpcyB0aGUgZm9jdXMgb2YgdGhpcyBhbmFseXNpcy4KCiMjIyBQYWNrYWdlcwoKV2Ugd2lsbCB1c2UgdGhlICoqdGlkeXZlcnNlKiogY29sbGVjdGlvbiBvZiBwYWNrYWdlcyBmb3IgZGF0YSBpbXBvcnQsIHdyYW5nbGluZyBhbmQgdmlzdWFsaXphdGlvbi4KCmBgYHtyIGxvYWQtcGFja2FnZXMsIHdhcm5pbmc9RkFMU0UsIG1lc3NhZ2U9RkFMU0V9CmxpYnJhcnkodGlkeXZlcnNlKQpgYGAKCiMjIyBEYXRhCgpUaGUgZGF0YSB3ZSdyZSB1c2luZyB3YXMgZ2VuZXJhdGVkIGJ5IENSVUsgQ0kgc2NpZW50aXN0cyAoW1BhcGFjaHJpc3RvdSBfZXQgYWwuXywgTmF0dXJlIENvbW11bmljYXRpb25zIDk6MjMxMSwgMjAxOF0oaHR0cHM6Ly9wdWJtZWQubmNiaS5ubG0ubmloLmdvdi8yOTg5OTM1MykpLgpJdCBpcyB0aGUgYWJ1bmRhbmNlIG9mIHByb3RlaW5zIGluIGEgYnJlYXN0IGNhbmNlciBjZWxsIGxpbmUgKE1DRi03KSBhdAp2YXJpb3VzIHRpbWUgcG9pbnRzIGFmdGVyIHRyZWF0bWVudCB3aXRoIHRoZSBkcnVnIHRhbW94aWZlbiwgbWVhc3VyZWQgdXNpbmcgbWFzcyBzcGVjdHJvbWV0cnkuIEluIHRoZSBjaHVuayBiZWxvdyB3ZSBpbXBvcnQgdGhlIEV4Y2VsIGZpbGUgY29udGFpbmluZyB0aGUgZGF0YSwgc2tpcHBpbmcgdGhlIGZpcnN0IGhlYWRlciByb3cuCgpgYGB7ciBtZXNzYWdlPUZBTFNFfQpwcm90ZWluX2xldmVscyA8LSByZWFkX2NzdigiaHR0cHM6Ly96ZW5vZG8ub3JnL3JlY29yZC82NDUwMTQ0L2ZpbGVzLzQxNDY3XzIwMThfNDYxOV9NT0VTTTExX0VTTS5jc3YiLCBza2lwID0gMSkKYGBgCgojIyBQcm90ZWluIGxldmVscyBhZnRlciB0cmVhdG1lbnQgd2l0aCB0YW1veGlmZW4KCkxldCdzIGNyZWF0ZSBhIGRhdGEgdmlzdWFsaXNhdGlvbiB0aGF0IGRpc3BsYXlzIHRoZSBsZXZlbHMgZm9yIHByb3RlaW5zIHRoYXQgYXJlIGRlY3JlYXNlZCAobG9nMmZjIDwgLTAuNzUpIGluIHRoZSAyNGggdGFtb3hpZmVuIHRyZWF0ZWQgc2FtcGxlIHZlcnN1cyBjb250cm9sICh2ZWgpLgoKYGBge3IgcGxvdC1wcm90ZWluLWxldmVscywgZmlnLndpZHRoPTEwLCBmaWcuaGVpZ2h0PTYsIG1lc3NhZ2U9RkFMU0V9CnByb3RlaW5fbGV2ZWxzICU+JQogIGZpbHRlcihgbG9nMkZDKDI0aC92ZWgpYCA8IC0wLjc1KSAlPiUKICBzZWxlY3QoYWNjZXNzaW9uID0gYFVuaXByb3QgQWNjZXNzaW9uYCwgZ2VuZSA9IGBHZW5lIE5hbWVgLCB2ZWhpY2xlLnJlcDAxOnRhbS4yNGgucmVwMDQpICU+JQogIHBpdm90X2xvbmdlcih2ZWhpY2xlLnJlcDAxOnRhbS4yNGgucmVwMDQsIG5hbWVzX3RvID0gInNhbXBsZSIsIHZhbHVlc190byA9ICJwcm90ZWluX2xldmVsIikgJT4lCiAgbXV0YXRlKHNhbXBsZSA9IHN0cl9yZXBsYWNlKHNhbXBsZSwgInZlaGljbGUiLCAidmVoaWNsZS4yNGgiKSkgJT4lCiAgc2VwYXJhdGUoc2FtcGxlLCBpbnRvID0gYygidHJlYXRtZW50IiwgInRpbWUiLCAicmVwbGljYXRlIikpICU+JQogIG11dGF0ZShncm91cCA9IHN0cl9zdWIodHJlYXRtZW50LCAxLCAzKSkgJT4lCiAgbXV0YXRlKGdyb3VwID0gc3RyX2MoZ3JvdXAsIHRpbWUpKSAlPiUKICBtdXRhdGUoZ3JvdXAgPSBhc19mYWN0b3IoZ3JvdXApKSAlPiUKICBnZ3Bsb3QobWFwcGluZyA9IGFlcyh4ID0gZ3JvdXAsIHkgPSBwcm90ZWluX2xldmVsLCBmaWxsID0gZ3JvdXApKSArCiAgZ2VvbV9ib3hwbG90KCkgKwogIGdlb21faml0dGVyKHdpZHRoID0gMC4xNSkgKwogIGZhY2V0X3dyYXAodmFycyhnZW5lKSwgc2NhbGVzID0gImZyZWVfeSIsIG5jb2wgPSAzKSArCiAgbGFicygKICAgIHRpdGxlID0gIlByb3RlaW4gbGV2ZWxzIGluIE1DRi03IHRyZWF0ZWQgd2l0aCB0YW1veGlmZW4iLAogICAgeSA9ICJQcm90ZWluIGxldmVsIiwKICAgIHggPSAiIgogICkKYGBgCgoxLiBDbGljayBLbml0IHRvIGdlbmVyYXRlIGEgaHRtbCByZXBvcnQgY29udGFpbmluZyB0aGUgcGxvdC4gIAoyLiBXZSBjYW4gZWFzaWx5IGNoYW5nZSB3aGljaCBwcm90ZWlucyBhcmUgYmVpbmcgcGxvdHRlZCBieSBjaGFuZ2luZyB0aGUgdGhyZXNob2xkIHRoZSBjb2RlIGFib3ZlIGBmaWx0ZXJgcyBmb3IuIFRyeSBjaGFuZ2luZyB0aGUgZmlsdGVyIHRvIGA+IDAuNzVgICB0aGVuIEtuaXQgYWdhaW4uCgojIyBSZWZlcmVuY2VzCgoxLiAgQ2FuY2VyIFJlc2VhcmNoIFVLIENhbWJyaWRnZSBJbnN0aXR1dGUsIFVuaXZlcnNpdHkgb2YgQ2FtYnJpZGdlICJJbnRyb2R1Y3Rpb24gdG8gUiBmb3IgQ2FuY2VyIFNjaWVudGlzdHMiCjIuICBNaW5lIENldGlua2F5YS1SdW5kZWwgIkRhdGEgU2NpZW5jZSBpbiBhIEJveCIgCg==" download="example-report.Rmd">Download example-report.Rmd</a></p>
<ul>
<li>Download the example report file in the link above</li>
<li>In your RStudio top menu go to File &gt; Open File and open the
example report file</li>
<li>Click the Knit button to generate the report</li>
<li>Try out the exercise to edit the file.</li>
</ul>
<p>Next we’ll explain some of the things you can see in the file.</p>
<hr>
</div>
<div class="section level2">
<h2 id="packages">Packages<a class="anchor" aria-label="anchor" href="#packages"></a>
</h2>
<p>There are thousands of R packages that extend R’s core functionality
to provide specialized functions for doing all kinds of things. There
are R packages that provide various kinds of visualizations and plotting
functions, others that let us manipulate data in interesting and useful
ways, and still others that give us access to cutting-edge machine
learning algorithms. Several packages have been developed by
bioinformaticians at Peter Mac, mainly focusing on genomic and
transcriptomic analysis methods.</p>
<p>You need to load a package using the
<strong><code><a href="https://rdrr.io/r/base/library.html" class="external-link">library()</a></code></strong> function in order to be able to
use the functions and data sets that the package provides. We can load
the core tidyverse packages in one go as follows.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## ── <span style="font-weight: bold;">Attaching packages</span> ─────────────────────────────────────── tidyverse 1.3.1 ──</span></code></pre>
<pre><code><span class="co">## <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">ggplot2</span> 3.3.5     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">purrr  </span> 0.3.4</span>
<span class="co">## <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">tibble </span> 3.1.6     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">dplyr  </span> 1.0.8</span>
<span class="co">## <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">tidyr  </span> 1.2.0     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">stringr</span> 1.4.0</span>
<span class="co">## <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">readr  </span> 2.1.2     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">forcats</span> 0.5.1</span></code></pre>
<pre><code><span class="co">## ── <span style="font-weight: bold;">Conflicts</span> ────────────────────────────────────────── tidyverse_conflicts() ──</span>
<span class="co">## <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">dplyr</span>::<span style="color: #00BB00;">filter()</span> masks <span style="color: #0000BB;">stats</span>::filter()</span>
<span class="co">## <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">dplyr</span>::<span style="color: #00BB00;">lag()</span>    masks <span style="color: #0000BB;">stats</span>::lag()</span></code></pre>
<p>The output from this command tells us which packages were loaded, one
of which is <code>ggplot2</code> that provides excellent plotting
functions for visualizing our data and will be the main focus next
week.</p>
<p>We also got a warning about some conflicts. This tells us that there
are functions loaded as part of the <code>dplyr</code> package that have
the same name as functions that are in the <code>stats</code> package.
The <code>stats</code> package was already loaded, from when we first
started our R session, and the functions referred to in the warning are
now masked; to use these instead of their <code>dplyr</code> namesakes
you need to specify them fully by adding the package name and
<code>::</code> as shown in the warning message.</p>
<div class="section level3">
<h3 id="installing-packages">Installing packages<a class="anchor" aria-label="anchor" href="#installing-packages"></a>
</h3>
<p>If you got the following error message when trying to load the
tidyverse, then you haven’t yet installed it.</p>
<pre><code>&gt; library(tidyverse)
Error in library(tidyverse) : there is no package called ‘tidyverse’</code></pre>
<p>Installing packages should be relatively straightforward using the
<code><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages()</a></code> function.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"tidyverse"</span><span class="op">)</span></code></pre></div>
<p>If you’re using R at a regular console within a terminal window,
i.e. not RStudio, you may be asked to select one of several mirrors of
the CRAN R package repository hosted at different locations around the
world.</p>
<p>For some packages you may be asked to choose between a binary and a
source package, particularly if the latest version of the package is
only available in source form and that package contains code written in
another language such as C. This could be a problem if you don’t have a
C compiler installed on your computer.</p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="importing-data-into-r">Importing data into R<a class="anchor" aria-label="anchor" href="#importing-data-into-r"></a>
</h2>
<p>Another of the tidyverse packages loaded when we ran
<code><a href="https://tidyverse.tidyverse.org" class="external-link">library(tidyverse)</a></code> is the <code>readr</code> package, which
provides functions for reading tabular data into a tibble and writing
tibbles to tab-delmited and comma-separated value (CSV) files.</p>
<p>For example, the clinical data for the METABRIC breast cancer dataset
generated at CRUK CI can be downloaded as a tab-delimited file from <a href="https://www.cbioportal.org/study/clinicalData?id=brca_metabric" class="external-link">cBioPortal</a>.
Targeted sequencing was performed for 2509 primary breast tumours, along
with 548 matched normals, using a panel of 173 of the most frequently
mutated breast cancer genes as part of the METABRIC study (<a href="https://www.ncbi.nlm.nih.gov/pubmed/27161491" class="external-link">Pereira <em>et
al.</em>, Nature Communications 7:11479, 2016</a>).</p>
<p>This file contains information in tabular form in which values are
separated by tab characters so we’ll use the
<strong><code><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv()</a></code></strong> to read these data.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">metabric</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="st">"https://zenodo.org/record/6450144/files/brca_metabric_clinical_data.tsv"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">2509</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">20</span></span>
<span class="co">## <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span>
<span class="co">## <span style="font-weight: bold;">Delimiter:</span> "\t"</span>
<span class="co">## <span style="color: #BB0000;">chr</span> (14): Patient ID, Sample ID, Cancer Type, Cancer Type Detailed, 3-Gene c...</span>
<span class="co">## <span style="color: #00BB00;">dbl</span>  (6): Mutation Count, Age at Diagnosis, Cohort, Lymph nodes examined pos...</span>
<span class="co">## </span>
<span class="co">## <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span>
<span class="co">## <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.</span></code></pre>
<p><code><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv()</a></code> prints a message telling us which types it
has assigned to each of the columns it has read, although in this case
the table has quite a lot of columns and it only outputs details about
the first few.</p>
<p>If we print the clinical data set we can see that
<code><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv()</a></code> created a tibble.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">metabric</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 2,509 × 20</span></span>
<span class="co">##    `Patient ID` `Sample ID` `Cancer Type` `Cancer Type Detail…` `Mutation Count`</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> MB-0000      MB-0000     Breast Cancer Breast Invasive Duct…               <span style="color: #BB0000;">NA</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> MB-0002      MB-0002     Breast Cancer Breast Invasive Duct…                2</span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> MB-0005      MB-0005     Breast Cancer Breast Invasive Duct…                2</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> MB-0006      MB-0006     Breast Cancer Breast Mixed Ductal …                1</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> MB-0008      MB-0008     Breast Cancer Breast Mixed Ductal …                2</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> MB-0010      MB-0010     Breast Cancer Breast Invasive Duct…                4</span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> MB-0014      MB-0014     Breast Cancer Breast Invasive Duct…                4</span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> MB-0020      MB-0020     Breast Cancer Breast Invasive Lobu…               <span style="color: #BB0000;">NA</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> MB-0022      MB-0022     Breast Cancer Breast Mixed Ductal …                1</span>
<span class="co">## <span style="color: #BCBCBC;">10</span> MB-0025      MB-0025     Breast Cancer Breast Invasive Duct…                5</span>
<span class="co">## <span style="color: #949494;"># … with 2,499 more rows, and 15 more variables:</span></span>
<span class="co">## <span style="color: #949494;">#   `3-Gene classifier subtype` &lt;chr&gt;, `Age at Diagnosis` &lt;dbl&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   Cellularity &lt;chr&gt;, Chemotherapy &lt;chr&gt;, Cohort &lt;dbl&gt;, `ER Status` &lt;chr&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   `ER status measured by IHC` &lt;chr&gt;, `HER2 Status` &lt;chr&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   `HER2 status measured by SNP6` &lt;chr&gt;, `Hormone Therapy` &lt;chr&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   `Inferred Menopausal State` &lt;chr&gt;, `Integrative Cluster` &lt;chr&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   `Lymph nodes examined positive` &lt;dbl&gt;, `Neoplasm Histologic Grade` &lt;dbl&gt;, …</span></span></code></pre>
<div class="rmdblock">
<p>
<strong><code><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv()</a></code></strong> and
<strong><code><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv()</a></code></strong>
</p>
<p>
<code><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv()</a></code> reads data from a comma-separated value (CSV)
file into a tibble. <code><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv()</a></code> is the equivalent function
that works on tab-delimited files.
</p>
<p>
These functions expect the first line to contain column names and try to
make sensible guesses at the type of data in each column. You can change
this by specifying various arguments, e.g. to skip comment lines
beginning with a specific character (e.g. “#”) or to tell the function
what the column types are.
</p>
<p>
patients &lt;- read_csv(“patients.csv”)
</p>
</div>
<p>We are now in a position to explore these data. We can see all column
names with R’s built-in <strong><code><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames()</a></code></strong>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">metabric</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##  [1] "Patient ID"                    "Sample ID"                    </span>
<span class="co">##  [3] "Cancer Type"                   "Cancer Type Detailed"         </span>
<span class="co">##  [5] "Mutation Count"                "3-Gene classifier subtype"    </span>
<span class="co">##  [7] "Age at Diagnosis"              "Cellularity"                  </span>
<span class="co">##  [9] "Chemotherapy"                  "Cohort"                       </span>
<span class="co">## [11] "ER Status"                     "ER status measured by IHC"    </span>
<span class="co">## [13] "HER2 Status"                   "HER2 status measured by SNP6" </span>
<span class="co">## [15] "Hormone Therapy"               "Inferred Menopausal State"    </span>
<span class="co">## [17] "Integrative Cluster"           "Lymph nodes examined positive"</span>
<span class="co">## [19] "Neoplasm Histologic Grade"     "Nottingham prognostic index"</span></code></pre>
<p>We can access individual columns with
<code>table_name$column_name</code>, for example
<code>metabric$Chemotherapy</code></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">metabric</span><span class="op">$</span><span class="va">Chemotherapy</span></code></pre></div>
<pre><code><span class="co">##    [1] "NO"  "NO"  "YES" "YES" "YES" "NO"  "YES" "YES" "NO"  NA    "NO"  "NO" </span>
<span class="co">##   [13] "NO"  "NO"  "YES" "NO"  "YES" "YES" "NO"  "NO"  "NO"  "NO"  "YES" "YES"</span>
<span class="co">##   [25] "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "YES"</span>
<span class="co">##   [37] "NO"  "YES" "YES" "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "YES"</span>
<span class="co">##   [49] "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "YES" "NO" </span>
<span class="co">##   [61] "YES" "NO"  "YES" "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">##   [73] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">##   [85] "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "YES" "YES" "YES" "NO"  "NO" </span>
<span class="co">##   [97] "NO"  "NO"  "NO"  "YES" "NO"  "YES" "NO"  "YES" "NO"  "NO"  "YES" "NO" </span>
<span class="co">##  [109] "NO"  "NO"  "NO"  "YES" "YES" "NO"  "NO"  "NO"  "YES" "NO"  "YES" "YES"</span>
<span class="co">##  [121] "NO"  "NO"  "YES" "YES" "YES" "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO" </span>
<span class="co">##  [133] "YES" "NO"  "YES" "NO"  "NO"  "YES" "YES" "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">##  [145] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "YES" "YES" "NO"  "NO"  "NO" </span>
<span class="co">##  [157] "YES" "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "YES" "NO"  "YES" "NO" </span>
<span class="co">##  [169] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO" </span>
<span class="co">##  [181] "YES" "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO" </span>
<span class="co">##  [193] "NO"  "YES" "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES"</span>
<span class="co">##  [205] "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO" </span>
<span class="co">##  [217] "NO"  "NO"  "NO"  "YES" "NO"  "YES" "NO"  NA    "NO"  "NO"  NA    "YES"</span>
<span class="co">##  [229] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "YES" "NO"  "NO" </span>
<span class="co">##  [241] "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "YES" "NO" </span>
<span class="co">##  [253] "YES" "NO"  "NO"  "NO"  "YES" "YES" "YES" "NO"  "YES" "NO"  "NO"  "NO" </span>
<span class="co">##  [265] "NO"  "YES" "YES" "NO"  "YES" "YES" "NO"  "NO"  "YES" "NO"  "YES" "YES"</span>
<span class="co">##  [277] "YES" "YES" "YES" "YES" "NO"  "NO"  "YES" "NO"  "NO"  "YES" "YES" "YES"</span>
<span class="co">##  [289] "NO"  "NO"  "NO"  "NO"  "YES" "YES" "NO"  "YES" "NO"  "YES" "NO"  "NO" </span>
<span class="co">##  [301] "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "YES" "NO"  "YES" "NO"  "NO"  "NO" </span>
<span class="co">##  [313] "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "YES" "NO"  "YES" "NO"  "NO"  "NO" </span>
<span class="co">##  [325] "YES" "YES" "YES" "YES" "NO"  "NO"  "YES" "NO"  "NO"  "YES" "NO"  "NO" </span>
<span class="co">##  [337] "NO"  "NO"  "YES" "NO"  "YES" "YES" "YES" "YES" "NO"  "NO"  "NO"  "YES"</span>
<span class="co">##  [349] "NO"  "YES" "NO"  "YES" "NO"  "YES" "YES" "YES" "YES" "YES" "YES" "NO" </span>
<span class="co">##  [361] "NO"  "YES" "NO"  "NO"  "NO"  "YES" "YES" "YES" "NO"  "NO"  "YES" "YES"</span>
<span class="co">##  [373] "YES" "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">##  [385] "YES" "YES" "NO"  "YES" "NO"  "NO"  "YES" "YES" "NO"  "NO"  "YES" "YES"</span>
<span class="co">##  [397] "YES" "YES" "YES" "NO"  "NO"  "YES" "YES" "NO"  "NO"  "YES" "YES" "NO" </span>
<span class="co">##  [409] "YES" "NO"  "YES" NA    "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "YES"</span>
<span class="co">##  [421] "YES" "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "YES" "NO" </span>
<span class="co">##  [433] "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "YES" "NO" </span>
<span class="co">##  [445] "NO"  "YES" "NO"  "NO"  "YES" "NO"  "YES" "NO"  "YES" "NO"  "NO"  "NO" </span>
<span class="co">##  [457] "YES" "NO"  "YES" "NO"  "NO"  "YES" "YES" "NO"  "NO"  "NO"  "NO"  "YES"</span>
<span class="co">##  [469] "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "YES" "NO"  "NO"  "NO" </span>
<span class="co">##  [481] "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">##  [493] "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "YES" "NO"  "YES" "NO"  "NO"  "YES"</span>
<span class="co">##  [505] "YES" "YES" "NO"  "YES" "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "YES" "NO" </span>
<span class="co">##  [517] "NO"  "YES" "NO"  "NO"  "YES" "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">##  [529] "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES"</span>
<span class="co">##  [541] "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO" </span>
<span class="co">##  [553] "NO"  "NO"  "YES" "NO"  "YES" "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">##  [565] "YES" "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">##  [577] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">##  [589] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "YES" "NO"  "NO" </span>
<span class="co">##  [601] "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO" </span>
<span class="co">##  [613] "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "YES"</span>
<span class="co">##  [625] "NO"  "YES" "NO"  "YES" "YES" "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">##  [637] "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">##  [649] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO" </span>
<span class="co">##  [661] "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "YES" "NO"  "NO"  "YES" "NO" </span>
<span class="co">##  [673] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "YES" "NO"  "NO"  "YES" "NO" </span>
<span class="co">##  [685] "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">##  [697] "NO"  "YES" "YES" "YES" "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO" </span>
<span class="co">##  [709] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "YES" "NO" </span>
<span class="co">##  [721] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO" </span>
<span class="co">##  [733] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "YES" "NO" </span>
<span class="co">##  [745] "NO"  "NO"  "YES" "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "YES" "NO" </span>
<span class="co">##  [757] "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "YES" "YES" "NO"  "NO"  "NO" </span>
<span class="co">##  [769] "YES" "NO"  "NO"  "YES" "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">##  [781] "NO"  "NO"  "NO"  "NO"  "YES" "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "YES"</span>
<span class="co">##  [793] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">##  [805] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">##  [817] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">##  [829] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO" </span>
<span class="co">##  [841] "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">##  [853] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO" </span>
<span class="co">##  [865] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">##  [877] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO" </span>
<span class="co">##  [889] "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO" </span>
<span class="co">##  [901] "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">##  [913] "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO" </span>
<span class="co">##  [925] "NO"  "NO"  "NO"  "NO"  "YES" "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">##  [937] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">##  [949] "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">##  [961] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">##  [973] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">##  [985] "YES" "NO"  "YES" "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">##  [997] "NO"  "NO"  "NO"  "YES" "YES" "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1009] "NO"  "NO"  "YES" "YES" "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "YES"</span>
<span class="co">## [1021] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1033] "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1045] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1057] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1069] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1081] "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "YES" "NO"  "NO"  "NO" </span>
<span class="co">## [1093] "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1105] "NO"  "NO"  "NO"  "YES" "NO"  "YES" "NO"  "NO"  "NO"  "YES" "NO"  "NO" </span>
<span class="co">## [1117] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1129] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO" </span>
<span class="co">## [1141] "NO"  "NO"  "NO"  "NO"  "YES" "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1153] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1165] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1177] "YES" "NO"  "NO"  "NO"  "NO"  "YES" "YES" "NO"  "NO"  "YES" "NO"  "NO" </span>
<span class="co">## [1189] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "YES" "NO" </span>
<span class="co">## [1201] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1213] "YES" "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1225] "YES" "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1237] "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1249] "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1261] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1273] "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1285] "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO" </span>
<span class="co">## [1297] "NO"  "NO"  "YES" "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1309] "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1321] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1333] "NO"  "NO"  "YES" "NO"  "NO"  "YES" "YES" "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1345] "NO"  "NO"  "NO"  "YES" "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "YES"</span>
<span class="co">## [1357] "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1369] "NO"  "YES" "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1381] "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "YES" "YES" "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1393] "NO"  "YES" "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1405] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "YES"</span>
<span class="co">## [1417] "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO" </span>
<span class="co">## [1429] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "YES" "NO"  "YES"</span>
<span class="co">## [1441] "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO" </span>
<span class="co">## [1453] "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO" </span>
<span class="co">## [1465] "NO"  "YES" "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1477] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1489] "NO"  "NO"  "NO"  "YES" "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO" </span>
<span class="co">## [1501] "NO"  "NO"  "NO"  "YES" "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1513] "YES" "YES" "NO"  "NO"  "NO"  "YES" "YES" "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1525] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1537] "YES" "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "YES"</span>
<span class="co">## [1549] "NO"  "NO"  "NO"  "NO"  "YES" "YES" "NO"  "NO"  "NO"  "NO"  "YES" "NO" </span>
<span class="co">## [1561] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO" </span>
<span class="co">## [1573] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1585] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1597] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES"</span>
<span class="co">## [1609] "YES" "YES" "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1621] "NO"  "YES" "NO"  "YES" "NO"  "NO"  "YES" "YES" "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1633] "NO"  "YES" "YES" "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1645] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1657] "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1669] "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1681] NA    "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "YES" "YES" "NO" </span>
<span class="co">## [1693] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "YES"</span>
<span class="co">## [1705] "NO"  "YES" "YES" "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "YES"</span>
<span class="co">## [1717] "YES" "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "YES" "NO" </span>
<span class="co">## [1729] "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "YES" "NO" </span>
<span class="co">## [1741] "YES" "YES" "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1753] "NO"  "NO"  "YES" "NO"  "YES" "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "YES"</span>
<span class="co">## [1765] "YES" "NO"  "YES" "NO"  "YES" "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO" </span>
<span class="co">## [1777] "YES" "NO"  "NO"  "YES" "YES" "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1789] "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "YES" "NO"  "NO" </span>
<span class="co">## [1801] "YES" "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO" </span>
<span class="co">## [1813] "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1825] "YES" "YES" "NO"  "NO"  "NO"  "YES" "NO"  "YES" "YES" "NO"  "NO"  "YES"</span>
<span class="co">## [1837] "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1849] "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "YES" "NO"  "NO"  "NO" </span>
<span class="co">## [1861] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1873] "NO"  "NO"  "YES" "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES"</span>
<span class="co">## [1885] "NO"  "YES" "YES" "NO"  "NO"  "NO"  "YES" "NO"  "YES" "NO"  "NO"  "NO" </span>
<span class="co">## [1897] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "NO"  "NO"  "YES" "NO"  "YES"</span>
<span class="co">## [1909] "NO"  "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "YES" "YES"</span>
<span class="co">## [1921] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "YES" "NO"  "NO"  "NO" </span>
<span class="co">## [1933] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1945] "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "YES" "YES" "YES" "NO"  "NO"  "NO" </span>
<span class="co">## [1957] "NO"  "YES" "NO"  "YES" "YES" "YES" "YES" "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1969] "YES" "YES" "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO"  "NO" </span>
<span class="co">## [1981] "NO"  "NO"  "NO"  "NO"  "NO"  NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [1993] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2005] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2017] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2029] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2041] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2053] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2065] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2077] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2089] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2101] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2113] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2125] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2137] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2149] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2161] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2173] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2185] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2197] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2209] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2221] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2233] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2245] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2257] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2269] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2281] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2293] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2305] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2317] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2329] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2341] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2353] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2365] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2377] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2389] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2401] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2413] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2425] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2437] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2449] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2461] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2473] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2485] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2497] NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA   </span>
<span class="co">## [2509] NA</span></code></pre>
<p>We can use R’s built-in <strong><code><a href="https://rdrr.io/r/base/table.html" class="external-link">table()</a></code></strong>
function to count the number of patients in this cohort that had
chemotherapy and those that did not.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">metabric</span><span class="op">$</span><span class="va">Chemotherapy</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">##   NO  YES </span>
<span class="co">## 1568  412</span></code></pre>
<p>We can summarize the numbers of mutations found per patient using R’s
built-in <strong><code><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary()</a></code></strong> function.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">metabric</span><span class="op">$</span><span class="va">`Mutation Count`</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's </span>
<span class="co">##   1.000   3.000   5.000   5.579   7.000  80.000     152</span></code></pre>
<p>Note that several of the column names have multiple words separated
by spaces. We have to use backticks ( <strong>`</strong> ) to refer to
column names such as these.</p>
<p>We can create a scatter plot to see if there is any relationship
between the age of the patient at diagnosis of breast cancer and the
Nottingham prognostic index (whatever that is) using R’s built-in
<strong><code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code></strong> function.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">metabric</span><span class="op">$</span><span class="va">`Age at Diagnosis`</span>, <span class="va">metabric</span><span class="op">$</span><span class="va">`Nottingham prognostic index`</span>,
     xlab <span class="op">=</span> <span class="st">"Age at diagnosis"</span>, ylab <span class="op">=</span> <span class="st">"Nottingham prognostic index"</span><span class="op">)</span></code></pre></div>
<p><img src="week1_files/figure-html/metabric_plot-1.png" width="700"></p>
<p>Hmm, what a horrible looking plot. Thank goodness we’re moving on to
<strong><code>ggplot2</code></strong> next week.</p>
<div class="section level3">
<h3 id="working-directory">Working directory<a class="anchor" aria-label="anchor" href="#working-directory"></a>
</h3>
<p>When importing the METABRIC dataset we specified the file using a URL
<code>https://zenodo.org/record/6450144/files/brca_metabric_clinical_data.tsv</code>
as it’s located on the web. If you have a file on your computer you can
specify the path to where it is located. For example, if we paste the
URL above into our web browser we can download the file and then import
it.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># assumes file is in our Downloads folder, "~" means our home directory e.g./home/maria</span>
<span class="va">metabric_downloaded</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="st">"~/Downloads/brca_metabric_clinical_data.tsv"</span><span class="op">)</span></code></pre></div>
<p>The <strong><code><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd()</a></code></strong> function will tell you what
the current working directory is.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>If you open RStudio by clicking on its icon in the task bar or its
entry in the start menu, your working directory will probably be your
home folder.</p>
<p>You can specify a full path for the file you want to read, which is
also known as an absolute path.</p>
<p>You can explicitly set your working directory using
<strong><code><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd()</a></code></strong>.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="st">"/Users/matt/projects/20200116_JB_amplicon_sequencing"</span><span class="op">)</span></code></pre></div>
<p>If you’re working with an R Markdown report (Rmd file) your working
directory will be the directory in which the report is located. In this
case you will probably want to read files from locations that are
relative to the directory in which you run the report.</p>
<p>If you’re writing an R report in RStudio, you can set the working
directory to be the same directory as that which contains your file
using the menu: <strong>Session &gt; Set Working Directory &gt; To
Source File Location</strong>. You can also navigate to the directory of
your choosing in the Files tab pane and set this to be the working
directory by clicking on the <strong>More</strong> menu button and
selecting <strong>Set As Working Directory</strong>.</p>
<p>Alternatively (and better) you can create an RStudio Project (File
&gt; New Project) as described <a href="https://r4ds.had.co.nz/workflow-projects.html#rstudio-projects" class="external-link">here</a>
to automatically keep track of and set the working directory.</p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="missing-values">Missing values<a class="anchor" aria-label="anchor" href="#missing-values"></a>
</h2>
<p>The METABRIC clinical dataset has some missing values. If you look
back to the table we printed out earlier you can see 2 missing values
for mutation counts within the first 10 rows. We can specified rows of a
column to view with
<strong><code>[firstrow:lastrow]</code></strong>.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">metabric</span><span class="op">$</span><span class="va">`Mutation Count`</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">##  [1] NA  2  2  1  2  4  4 NA  1  5</span></code></pre>
<div class="rmdblock">
<p>
<strong>Missing values</strong> (<strong><code>NA</code></strong>)
</p>
<p>
Missing values in R are represented as <strong><code>NA</code></strong>,
which stands for ‘not available’.
</p>
</div>
<p>The <code><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary()</a></code> function reported that there are 152
missing mutation counts. <code><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary()</a></code> reports the mean and
median values for the remaining observations, i.e. those for which a
mutation count <em>is</em> available. Not all functions do this
though.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">metabric</span><span class="op">$</span><span class="va">`Mutation Count`</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] NA</span></code></pre>
<p>The <code><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean()</a></code> function, for example, takes a different
approach in how it treats missing values. It returns <code>NA</code>,
i.e. a missing value, because it takes the view that it cannot compute a
mean for a set of values where some are not available. You can specify
the argument <code>na.rm = TRUE</code> when calling <code><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean()</a></code>
to instruct it to ignore the missing values and calculate the mean of
the remaining values.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">metabric</span><span class="op">$</span><span class="va">`Mutation Count`</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 5.578702</span></code></pre>
<p>There are many such functions in R that would be similarly affected
by missing values and which offer the <code>na.rm</code> argument.</p>
<p>You can test for missing values using the <code><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na()</a></code>
function.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">first_ten_mutation_counts</span> <span class="op">&lt;-</span> <span class="va">metabric</span><span class="op">$</span><span class="va">`Mutation Count`</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">first_ten_mutation_counts</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##  [1]  TRUE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE</span></code></pre>
<p>One final handy tip: summing the result of <code><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na()</a></code> will
count the number of missing values in a vector.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">metabric</span><span class="op">$</span><span class="va">`Mutation Count`</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 152</span></code></pre>
<p>Can you see why this works? <em>(hint: look at what kind of vector
<code><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum()</a></code> expects and think about how a logical vector will be
coerced into such a type)</em>.</p>
<hr>
</div>
<div class="section level2">
<h2 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<p>In this session we have covered the following concepts:</p>
<ul>
<li>The RStudio environment and the functions associated with each
pane</li>
<li>Generating a visualisation and report using R</li>
<li>How to install and use packages that extend R’s basic
functionality</li>
<li>How to read tabular data into R</li>
<li>Missing values and how R handles these</li>
</ul>
<p>We also cover how to work within an R markdown document and run code
therein in this week’s assignment.</p>
<hr>
</div>
<div class="section level2">
<h2 id="assignment">Assignment<a class="anchor" aria-label="anchor" href="#assignment"></a>
</h2>
<p><a href="data:text/x-markdown;base64,Ci0tLQp0aXRsZTogIkFzc2lnbm1lbnQgMSIKLS0tCgoKIyMjIFdvcmtpbmcgd2l0aCBkYXRhIGZyb20gYW4gRXhjZWwgc3ByZWFkc2hlZXQKCioqUmVhZCBpbnRvIFIgb25lIG9mIHlvdXIgb3duIEV4Y2VsIHNwcmVhZHNoZWV0cyBhbmQgZXhwbG9yZSB0aGUgZGF0YSB1c2luZyBgc3VtbWFyeSgpYCwgYHRhYmxlKClgIGFuZC9vciBgcGxvdCgpYCoqCgpXZSBoYXZlbid0IHlldCBzaG93ZWQgaG93IHRvIHJlYWQgRXhjZWwgc3ByZWFkc2hlZXRzIGludG8gUi4gVGhlcmUgYXJlIHNldmVyYWwgcGFja2FnZSBwcm92aWRpbmcgdGhpcyBjYXBhYmlsaXR5IGJ1dCBmb3IgdGhpcyBleGVyY2lzZSB1c2UgdGhlIGByZWFkeGxgIHBhY2thZ2UKZnJvbSB0aGUgdGlkeXZlcnNlLgoKYHJlYWR4bGAgc2hvdWxkIGhhdmUgYmVlbiBpbnN0YWxsZWQgYXMgcGFydCBvZiB0aGUgdGlkeXZlcnNlIGJ1dCBpZiB5b3UgZmluZAp0aGF0J3Mgbm90IHRoZSBjYXNlLCB5b3Ugd2lsbCBoYXZlIHRvIGluc3RhbGwgaXQgc2VwYXJhdGVseS4KCmByZWFkeGxgIGlzIG5vdCBsb2FkZWQgYXMgb25lIG9mIHRoZSBjb3JlIHRpZHl2ZXJzZSBwYWNrYWdlcyB3aGVuIHlvdSBydW4KYGxpYnJhcnkodGlkeXZlcnNlKWAgc28geW91J2xsIGhhdmUgdG8gbG9hZCB0aGlzIHNlcGFyYXRlbHkuCgpJZiB5b3UgZG9uJ3QgaGF2ZSBhbiBFeGNlbCBzcHJlYWRzaGVldCB0byBoYW5kLCB5b3UgY2FuIHVzZSB0aGUgY2xpbmljYWwKYW5ub3RhdGlvbiBkYXRhIGZvciB0aGUgTkNJLTYwIGNlbGwgbGluZSBkYXRhIHNldCAoaHR0cHM6Ly9iaW9pbmZvcm1hdGljcy1jb3JlLXNoYXJlZC10cmFpbmluZy5naXRodWIuaW8vci1pbnRyby9kYXRhL2NlbGxsaW5lX25jaTYwX2NsaW5pY2FsX2RhdGEueGxzeCkuCgpgYGB7cn0KCmBgYAoKLS0tCgojIyMgQ3JlYXRlIGEgcmVwb3J0IGZvciB5b3VyIGFzc2lnbm1lbnQKCkNsaWNrIG9uIHRoZSAnS25pdCcgbWVudSBhdCB0aGUgdG9wIG9mIHRoaXMgZmlsZSBhbmQgc2VsZWN0IGVpdGhlciB3aGljaGV2ZXIKb3B0aW9uIHlvdSBwcmVmZXIgdG8gY3JlYXRlIGFuIEhUTUwsIFBERiBvciBXb3JkIGRvY3VtZW50IHZlcnNpb24gb2YgeW91cgphc3NpZ25tZW50LiBUaGlzIHdpbGwgcnVuIGFsbCB0aGUgY29kZSBjaHVua3MgYW5kICJrbml0IiB0aGUgcmVzdWx0aW5nCnJlc3VsdHMgd2l0aCB0aGUgc3Vycm91bmRpbmcgdGV4dCB0byBwcm9kdWNlIGEgcmVwb3J0LgoKIyMjIyBBZGRpdGlvbmFsIG1hcmtkb3duIGZlYXR1cmVzCgoqKl9FbXBoYXNpc18qKgoKQ2FuIHVzZSBlaXRoZXIgdW5kZXJzY29yZXMgb3IgYXN0ZXJpc2tzIGFuZCBjb21iaW5hdGlvbiBvZiBib3RoIGZvciBib3RoIGJvbGQKYW5kIGl0YWxpYyBlbXBoYXNpcy4KCmBgYAoqaXRhbGljKiAgICoqYm9sZCoqCl9pdGFsaWNfICAgX19ib2xkX18KYGBgCgoqKl9JbmxpbmUgQ29kZV8qKgoKVXNlIG9mIGJhY2t0aWNrcyAoYCkuCgpgYGAKYG1lYW4oc3RhcndhciRoZWlnaHQpYApgYGAKCioqX0hvcml6b250YWwgUnVsZV8qKgoKVGhyZWUgb2YgbW9yZSBhc3Rlcmlza3Mgb3IgZGFzaGVzLgoKYGBgCioqKioqKioKCi0tLS0tLS0KYGBgCgoqKl9QbGFpbiBDb2RlIEJsb2Nrc18qKgoKUGxhaW4gY29kZSBibG9ja3MgYXJlIGRpc3BsYXllZCBpbiBhIGZpeGVkLXdpZHRoIGZvbnQgYnV0IG5vdCBldmFsdWF0ZWQuCgpgYGAKVGhpcyB0ZXh0IGlzIGRpc3BsYXllZCB2ZXJiYXRpbSAvIHByZWZvcm1hdHRlZApgYGAKCkhhdmUgYSBsb29rIGF0IGV4YW1wbGVzIG9mIHRoZXNlIGluIHRoZSB0ZXh0IHBhcnRzIG9mIHRoaXMgUiBtYXJrZG93biBkb2N1bWVudAphbmQgaG93IHRoZXNlIGFyZSByZW5kZXJlZCB3aGVuIHlvdSBydW4gdGhlIGtuaXQgb3BlcmF0aW9uLgoKRm9yIG1vcmUgZmVhdHVyZXMsIHNlZSB0aGUgaGVscCBwYWdlIGluIFJTdHVkaW8gdGhhdCBjYW4gYmUgYWNjZXNzZWQgZnJvbSB0aGUKdGhlIGhlbHAgbWVudTogKipIZWxwID4gTWFya2Rvd24gUXVpY2sgUmVmZXJlbmNlKioKCi0tLQoKIyMjIEFkdmFuY2VkOiBDbGluaWNhbCBjYW5jZXIgZGF0YQoKKipIb3cgbWFueSBzYW1wbGVzIGluIHRoZSBNRVRBQlJJQyBkYXRhc2V0IGhhdmUgaGlnaCBjZWxsdWxhcml0eSBidXQgaGF2ZSBubyByZWNvcmRlZCBjbGFzc2lmaWNhdGlvbiB3aXRoIHRoZSAzLWdlbmUgY2xhc3NpZmllcj8qKgoKYGBge3J9CgpgYGAKCipIaW50OiogdGhpcyBpcyBhY3R1YWxseSBxdWl0ZSB0cmlja3kgLS0geW91IG1heSB3YW50IHRvIGNoZWNrIGZvciBtaXNzaW5nCnZhbHVlcyBpbiB0aGUgY2VsbHVsYXJpdHkgY29sdW1uIGFzIHdlbGwgYXMgdGhlIGNvbHVtbiBmb3IgdGhlIDMtZ2VuZQpjbGFzc2lmaWVyLgoKLS0tCg==" download="assignment1.Rmd">Download assignment1.Rmd</a></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Maria Doyle.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a>
2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
