<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Session 2 -- Visualizing tabular data with ggplot2 • R4CancerSci</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Session 2 -- Visualizing tabular data with ggplot2">
<meta property="og:description" content="R4CancerSci">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">R4CancerSci</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Material
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/week0.html">Week 0 - Getting ready</a>
    </li>
    <li>
      <a href="../articles/week1.html">Week 1 - Introduction to RStudio</a>
    </li>
    <li>
      <a href="../articles/week2.html">Week 2 - Data visualization with ggplot2</a>
    </li>
    <li>
      <a href="../articles/week3.html">Week 3 - Data manipulation using dplyr</a>
    </li>
    <li>
      <a href="../articles/week4.html">Week 4 - Grouping, combining and restructuring data for analysis</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/pmacdasci/R4CancerSci" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Session 2 – Visualizing tabular data with
ggplot2</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/R4CancerSci/R4CancerSci/blob/HEAD/vignettes/week2.Rmd" class="external-link"><code>vignettes/week2.Rmd</code></a></small>
      <div class="hidden name"><code>week2.Rmd</code></div>

    </div>

    
    
<blockquote>
<h4 id="learning-objectives">Learning objectives</h4>
<ul>
<li>Understand ggplot2’s grammar of graphics</li>
<li>Apply the grammar to create a variety of plots including scatter
plots, box plots and histograms</li>
<li>Learn about how R handles categorical data</li>
</ul>
</blockquote>
<div class="rmdblock">
<p>
A video that goes through this session’s material can be found <a href="https://www.youtube.com/watch?v=pVkdTcrCj9M" class="external-link">here</a>.
</p>
</div>
<div class="section level2">
<h2 id="visualizing-data-with-ggplot2">Visualizing data with ggplot2<a class="anchor" aria-label="anchor" href="#visualizing-data-with-ggplot2"></a>
</h2>
<p>In this session we will focus our attention on the popular
<strong>ggplot2</strong> package for visualizing tabular data and will
be learning how to create various different kinds of plot.</p>
<div class="section level3">
<h3 id="metabric-data-set">METABRIC data set<a class="anchor" aria-label="anchor" href="#metabric-data-set"></a>
</h3>
<p>We’ll be using an extended version of the METABRIC data set (from
last week) in which columns have been added for the mRNA expression
values for selected genes, including estrogen receptor alpha (ESR1),
progesterone receptor (PGR), GATA3 and FOXA1.</p>
<p>The METABRIC study characterized the genomic mutations and gene
expression profiles for over 2000 primary breast tumours. In addition to
the gene expression data generated using microarrays, genome-wide copy
number profiles were obtained using SNP microarrays and targeted
sequencing was performed using a panel of 40 driver-mutation genes to
detect mutations (single nucleotide variants).</p>
<p><a href="https://pubmed.ncbi.nlm.nih.gov/22522925" class="external-link">Curtis <em>et
al.</em>, Nature 486:346-52, 2012</a></p>
<p><a href="https://www.ncbi.nlm.nih.gov/pubmed/27161491" class="external-link">Pereira <em>et
al.</em>, Nature Communications 7:11479, 2016</a></p>
<p>Both the clinical data and the gene expression values were downloaded
from <a href="https://www.cbioportal.org/study/summary?id=brca_metabric" class="external-link">cBioPortal</a>.</p>
<p>These have been combined into a single table using techniques we will
cover later in the course. We’ve removed observations for patient tumour
samples for which expression data are not available, so our table has
fewer rows.</p>
<p>In addition, we have replaced spaces in column names with underscores
so we don’t have to constantly refer to these using backticks (
<strong><code>`</code></strong> ). Note, however, that even after
renaming some of the column names, we have one column, ‘3-gene
classifier’, that is not strictly a valid name in R as it starts with a
number and contains a hyphen. Valid names for variables in R contain
only letters, numbers and the dot (<strong><code>.</code></strong>) and
underscore (<strong><code>_</code></strong>) characters and should not
begin with a number. So when we refer to the ‘3-gene classifier’ column
we’ll have to use backticks,
e.g. <code>metabric$`3-gene_classifier`</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span>
<span class="va">metabric</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="st">"https://zenodo.org/record/6450144/files/metabric_clinical_and_expression_data.csv"</span><span class="op">)</span>
<span class="va">metabric</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 1,904 × 32</span></span>
<span class="co">##    Patient_ID Cohort Age_at_diagnosis Survival_time Survival_status Vital_status</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       </span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> MB-0000         1             75.6         140.  LIVING          Living      </span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> MB-0002         1             43.2          84.6 LIVING          Living      </span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> MB-0005         1             48.9         164.  DECEASED        Died of Dis…</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> MB-0006         1             47.7         165.  LIVING          Living      </span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> MB-0008         1             77.0          41.4 DECEASED        Died of Dis…</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> MB-0010         1             78.8           7.8 DECEASED        Died of Dis…</span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> MB-0014         1             56.4         164.  LIVING          Living      </span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> MB-0022         1             89.1          99.5 DECEASED        Died of Oth…</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> MB-0028         1             86.4          36.6 DECEASED        Died of Oth…</span>
<span class="co">## <span style="color: #BCBCBC;">10</span> MB-0035         1             84.2          36.3 DECEASED        Died of Dis…</span>
<span class="co">## <span style="color: #949494;"># … with 1,894 more rows, and 26 more variables: Chemotherapy &lt;chr&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   Radiotherapy &lt;chr&gt;, Tumour_size &lt;dbl&gt;, Tumour_stage &lt;dbl&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   Neoplasm_histologic_grade &lt;dbl&gt;, Lymph_nodes_examined_positive &lt;dbl&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   Lymph_node_status &lt;dbl&gt;, Cancer_type &lt;chr&gt;, ER_status &lt;chr&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   PR_status &lt;chr&gt;, HER2_status &lt;chr&gt;, HER2_status_measured_by_SNP6 &lt;chr&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   PAM50 &lt;chr&gt;, `3-gene_classifier` &lt;chr&gt;, Nottingham_prognostic_index &lt;dbl&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   Cellularity &lt;chr&gt;, Integrative_cluster &lt;chr&gt;, Mutation_count &lt;dbl&gt;, …</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="our-first-ggplot---a-scatter-plot">Our first ggplot - a scatter plot<a class="anchor" aria-label="anchor" href="#our-first-ggplot---a-scatter-plot"></a>
</h2>
<p>First we’ll create a simple scatter plot much like the one we created
in the previous session using R’s <code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code> function. We’ll
plot the expression of estrogen receptor alpha (ESR1) against that of
the transcription factor, GATA3.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="week2_files/figure-html/scatter_plot_1-1.png" width="700"></p>
<p>We specified 3 things to create this plot:</p>
<ol style="list-style-type: decimal">
<li>The <strong>data</strong> – needs to be a data frame (or a
tibble)</li>
<li>The type of plot – this is called a <strong>geom</strong> in ggplot2
terminology</li>
<li>The mapping of variables in the data to visual properties of objects
in the plot - these are called <strong>aesthetics</strong> in
ggplot2</li>
</ol>
<p>In this case, the type of plot is a
<strong><code>geom_point</code></strong>, ggplot2’s function for a
scatter plot, and the expression of the genes GATA3 and ESR1 (estrogen
receptor alpha) are mapped to the <strong><code>x</code></strong> and
<strong><code>y</code></strong> coordinates, i.e. the positions of
points on the scatter plot.</p>
<p>Other aesthetics include the size, shape and colour of the points. It
might seem a little surprising that the x and y coordinates are treated
in the same way as we wouldn’t normally think about these as having
aesthetic qualities but this is indeed the case with ggplot2.</p>
</div>
<div class="section level2">
<h2 id="aesthetic-mappings">Aesthetic mappings<a class="anchor" aria-label="anchor" href="#aesthetic-mappings"></a>
</h2>
<p>Look up the <code>geom_point</code> help page to see the list of
aesthetics that are available for this geom.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="op">?</span><span class="va">geom_point</span></code></pre></div>
<p>If you scroll down to the aesthetics section in the help page for
<code>geom_point</code>, you’ll see that the
<strong><code>x</code></strong> and <strong><code>y</code></strong>
aesthetics are in bold face which tells us that these are required,
i.e. you have to supply them, otherwise you’ll get an error message from
ggplot.</p>
<p>Other aesthetics such as size, shape, colour and transparency
(<code>alpha</code>) are optional.</p>
<p>So what about those other aesthetics? Colours are usually fun – let’s
try to use the <code>colour</code> aesthetic.</p>
<p>We need to map a variable to a colour just like we did for x and y.
We’ll use the ER status as an example.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span>, colour <span class="op">=</span> <span class="va">ER_status</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="week2_files/figure-html/scatter_plot_2-1.png" width="700"></p>
<p>What about colouring points based on a continuous variable?</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span>, colour <span class="op">=</span> <span class="va">Nottingham_prognostic_index</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="week2_files/figure-html/scatter_plot_3-1.png" width="700"></p>
<p>ggplot2 uses a colour scale for a continuous variable but discrete
colours for discrete or categorical values.</p>
<p>Size is not usually a good aesthetic to map to a variable. This is
particularly the case where we have such a large number of
observations.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span>, size <span class="op">=</span> <span class="va">ER_status</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Warning: Using size for a discrete variable is not advised.</span></code></pre>
<p><img src="week2_files/figure-html/scatter_plot_4-1.png" width="700"></p>
<p>Some aesthetics, such as size, are designed to be used with
continuous variables. If you map these to a discrete variable as we have
done here, then ggplot warns you that this isn’t such a good idea.</p>
<p>It would however be useful to be able to decrease the size of the
points given that there are around 2000 in this data set. We can do so
by setting the size to a single value rather than mapping it to one of
the variables in the data set - this has to be done outside the
aesthetic mappings, i.e. outside the <code><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes()</a></code> bit.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span>, colour <span class="op">=</span> <span class="va">Nottingham_prognostic_index</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></code></pre></div>
<p><img src="week2_files/figure-html/scatter_plot_5-1.png" width="700"></p>
<p>Some aesthetics can only be used with categorical variables,
e.g. shape.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span>, shape <span class="op">=</span> <span class="va">Survival_time</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>Error: A continuous variable can not be mapped to shape</code></pre>
<p>Here we set the shape based on the 3-gene classifier.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span>, shape <span class="op">=</span> <span class="va">`3-gene_classifier`</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">0.9</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Warning: Removed 204 rows containing missing values (geom_point).</span></code></pre>
<p><img src="week2_files/figure-html/scatter_plot_6-1.png" width="700"></p>
<p>Note that some of the patient samples have not been classified and
ggplot has removed those points with missing values for the 3-gene
classifier.</p>
<p>It is usually preferable to use colours to distinguish between
different categories but sometimes colour and shape are used together
where we want to show which group a data point belongs to in two
different categorical variables.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span>, colour <span class="op">=</span> <span class="va">PAM50</span>, shape <span class="op">=</span> <span class="va">`3-gene_classifier`</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">0.9</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Warning: Removed 204 rows containing missing values (geom_point).</span></code></pre>
<p><img src="week2_files/figure-html/scatter_plot_7-1.png" width="700"></p>
<p>Transparency can be useful when we have a large number of points as
we can more easily tell when points are overlaid, but like size, it is
not usually mapped to a variable and sits outside the
<code><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes()</a></code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span>, colour <span class="op">=</span> <span class="va">`3-gene_classifier`</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">0.75</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></code></pre></div>
<p><img src="week2_files/figure-html/scatter_plot_8-1.png" width="700"></p>
<p>Colours are more commonly used than shapes, sizes or
transparency.</p>
<p>Virtually every aspect of the plots we’ve created can be customized.
We will learn how to change the colours used as well as other display
parameters later in the course.</p>
</div>
<div class="section level2">
<h2 id="ggplot2-grammar">ggplot2 grammar<a class="anchor" aria-label="anchor" href="#ggplot2-grammar"></a>
</h2>
<p>The “gg” in ggplot2 stands for “grammar of graphics”. This grammar is
a coherent system for describing and building graphs.</p>
<p>The basic template is given in the summary box below. This template
can be used to create any of a wide range of graph types with
ggplot2.</p>
<div class="rmdblock">
<p>
<strong>ggplot2 grammar of graphics</strong>
</p>
<p>
<strong><code>ggplot(data = &lt;DATA&gt;) +
&lt;GEOM_FUNCTION&gt;(mapping = aes(&lt;MAPPINGS&gt;))</code></strong>
</p>
<p>
<code><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot()</a></code> creates the plot object.
</p>
<p>
Layers can be added by adding geoms such as <code><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point()</a></code>
using the <code>+</code> operator that ggplot2 has overridden for this
purpose.
</p>
</div>
</div>
<div class="section level2">
<h2 id="another-plot-type---bar-chart">Another plot type - bar chart<a class="anchor" aria-label="anchor" href="#another-plot-type---bar-chart"></a>
</h2>
<p>The range of geoms available in ggplot2 can be obtained by navigating
to the ggplot2 package in the Packages tab pane in RStudio (bottom
right-hand corner) and scrolling down the list of functions sorted
alphabetically to the <code>geom_...</code> functions.</p>
<p>The METABRIC study redefined how we think about breast cancer by
identifying and characterizing several new subtypes, referred to as
<em>integrative clusters</em>. Let’s create a bar chart of the number of
patients whose cancers fall within each subtype in the METABRIC
cohort.</p>
<p><code>geom_bar</code> is the geom we need and it requires a single
aesthetic mapping of the categorical variable of interest to
<code>x</code>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Integrative_cluster</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="week2_files/figure-html/bar_plot_1-1.png" width="700"></p>
<p>The dark grey bars are a big ugly - what if we want each bar to be a
different colour?</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Integrative_cluster</span>, colour <span class="op">=</span> <span class="va">Integrative_cluster</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="week2_files/figure-html/bar_plot_2-1.png" width="700"></p>
<p>Colouring the edges wasn’t quite what we had in mind. Look at the
help for <code>geom_bar</code> to see what other aesthetic we should
have used.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Integrative_cluster</span>, fill <span class="op">=</span> <span class="va">Integrative_cluster</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="week2_files/figure-html/bar_plot_3-1.png" width="700"></p>
<p>What happens if we colour (fill) with something other than the
integrative cluster?</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Integrative_cluster</span>, fill <span class="op">=</span> <span class="va">ER_status</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="week2_files/figure-html/bar_plot_4-1.png" width="700"></p>
<p>We get a stacked bar plot.</p>
<p>Note the similarity in what we did here to what we did with the
scatter plot - there is a common grammar.</p>
<p>Let’s try another stacked bar plot, this time with a categorical
variable with more than two categories.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Integrative_cluster</span>, fill <span class="op">=</span> <span class="va">`3-gene_classifier`</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="week2_files/figure-html/bar_plot_5-1.png" width="700"></p>
<p>What if want all the bars to be the same colour but not dark grey,
e.g. blue?</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Integrative_cluster</span>, fill <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="week2_files/figure-html/bar_plot_6-1.png" width="700"></p>
<p>That doesn’t look right - why not?</p>
<p>You can set the aesthetics to a fixed value but this needs to be
outside the mapping, just like we did before for size and transparency
in the scatter plots.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Integrative_cluster</span><span class="op">)</span>, fill <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span></code></pre></div>
<p><img src="week2_files/figure-html/bar_plot_7-1.png" width="700"></p>
<p>Setting this inside the <code><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes()</a></code> mapping told ggplot2 to
map the colour aesthetic to some variable in the data frame, one that
doesn’t really exist but which is created on-the-fly with a value of
“blue” for every observation.</p>
<div class="section level3">
<h3 id="statistical-transformations-advanced">Statistical transformations (ADVANCED)<a class="anchor" aria-label="anchor" href="#statistical-transformations-advanced"></a>
</h3>
<p>You may have noticed that ggplot2 didn’t just plot values from our
data set but had to do some calculation first for the bar chart, i.e. it
had to sum the number of observations in each category.</p>
<p>Each geom has a <strong>statistical transformation</strong>. In the
case of the scatter plot, <code>geom_point</code> uses the “identity”
transformation which means just use the values as they are, i.e. not
really a transformation at all. The statistical transformation for
<code>geom_bar</code> is “count”, which means it will count the number
of observations for each category in the variable mapped to the x
aesthetic.</p>
<p>You can see which statistical transformation is being used by a geom
by looking at the <code>stat</code> argument in the help page for that
geom.</p>
<p>There are some circumstances where you’d want to change the
<code>stat</code>, for example if we already had count values in our
table.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>
  `3-gene_classifier` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ER-/HER2-"</span>, <span class="st">"ER+/HER2- High Prolif"</span>, <span class="st">"ER+/HER2- Low Prolif"</span>, <span class="st">"HER2+"</span><span class="op">)</span>,
  count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">290</span>, <span class="fl">603</span>, <span class="fl">619</span>, <span class="fl">188</span><span class="op">)</span>
<span class="op">)</span>
<span class="va">counts</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 4 × 2</span></span>
<span class="co">##   `3-gene_classifier`   count</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                 <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> ER-/HER2-               290</span>
<span class="co">## <span style="color: #BCBCBC;">2</span> ER+/HER2- High Prolif   603</span>
<span class="co">## <span style="color: #BCBCBC;">3</span> ER+/HER2- Low Prolif    619</span>
<span class="co">## <span style="color: #BCBCBC;">4</span> HER2+                   188</span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">counts</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">`3-gene_classifier`</span>, y <span class="op">=</span> <span class="va">count</span><span class="op">)</span>, stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span></code></pre></div>
<p><img src="week2_files/figure-html/bar_plot_8-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="multiple-layers">Multiple layers<a class="anchor" aria-label="anchor" href="#multiple-layers"></a>
</h2>
<p>Consider again the ggplot2 grammar:</p>
<pre><code>ggplot(data = &lt;DATA&gt;) +
  &lt;GEOM_FUNCTION&gt;(mapping = aes(&lt;MAPPINGS&gt;))</code></pre>
<p>Why do we have the ggplot part and geom part joined by a “+”
symbol?</p>
<p>The <strong>‘+’</strong> operator has been overridden by ggplot2 to
add layers (geoms) and other components (scales, themes, etc.) to the
plot.</p>
<p>We can have multiple geoms in the same plot. Each is a different
layer.</p>
<p>Let’s create a plot with two geoms.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## `geom_smooth()` using method = 'gam' and formula 'y ~ s(x, bs = "cs")'</span></code></pre>
<p><img src="week2_files/figure-html/scatter_line_plot_1-1.png" width="700"></p>
<p>What is <strong><code><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth()</a></code></strong> doing? Look up
the help page.</p>
<p>Note that the shaded area surrounding blue line represents the
standard error bounds on the fitted model.</p>
<p>It is important when breaking these ggplot statements across multiple
lines to ensure that the <code>+</code> is at the end of the line so
that R knows you haven’t finished constructing your plot.</p>
<p>There is some annoying duplication of code used to create this plot.
We’ve repeated the exact same aesthetic mapping for both geoms. We can
avoid this by putting the mappings in the <code><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot()</a></code> function
instead.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## `geom_smooth()` using method = 'gam' and formula 'y ~ s(x, bs = "cs")'</span></code></pre>
<p><img src="week2_files/figure-html/scatter_line_plot_2-1.png" width="700"></p>
<p>Let’s make the plot look a bit prettier by reducing the size of the
points and making them transparent. We’re not mapping <code>size</code>
or <code>alpha</code> to any variables, just setting them to constant
values, and we only want these settings to apply to the points, so we
set them inside <code><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point()</a></code>.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.5</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## `geom_smooth()` using method = 'gam' and formula 'y ~ s(x, bs = "cs")'</span></code></pre>
<p><img src="week2_files/figure-html/scatter_line_plot_3-1.png" width="700"></p>
<p>Aesthetics set in the <code>ggplot</code> function are treated as
global mappings for the plot and are inherited by all geoms added to the
plot.</p>
<p>Let’s say we still want to colour the points by ER status.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span>, colour <span class="op">=</span> <span class="va">ER_status</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.5</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## `geom_smooth()` using method = 'gam' and formula 'y ~ s(x, bs = "cs")'</span></code></pre>
<p><img src="week2_files/figure-html/scatter_line_plot_4-1.png" width="700"></p>
<p>Well, that wasn’t expected, although it’s pretty neat. We only wanted
to apply the colour aesthetic to the points.</p>
<p>We can add the colour aesthetic to the <code><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point()</a></code>
function instead of as a global mapping in the <code><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot()</a></code>
function.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">ER_status</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">0.5</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## `geom_smooth()` using method = 'gam' and formula 'y ~ s(x, bs = "cs")'</span></code></pre>
<p><img src="week2_files/figure-html/scatter_line_plot_5-1.png" width="700"></p>
<p>Or suppose you’ve spent a bit of time getting your scatter plot just
right and decide to add another layer but you’re a bit worried about
interfering with the the code you so lovingly crafted, you can set the
<code>inherit.aes</code> option to <code>FALSE</code> and set the
aesthetic mappings explicitly for your new layer.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span>, colour <span class="op">=</span> <span class="va">ER_status</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.5</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span><span class="op">)</span>, inherit.aes <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## `geom_smooth()` using method = 'gam' and formula 'y ~ s(x, bs = "cs")'</span></code></pre>
<p><img src="week2_files/figure-html/scatter_line_plot_6-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="box-plots">Box plots<a class="anchor" aria-label="anchor" href="#box-plots"></a>
</h2>
<p>Box plots are a particular favourite seen in many seminars and
papers.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ER_status</span>, y <span class="op">=</span> <span class="va">GATA3</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="week2_files/figure-html/box_plot_1-1.png" width="700"></p>
<p>See <code>geom_boxplot</code> help to explain how the box and
whiskers are constructed and how it decides which points are outliers
and should be displayed as points.</p>
<p>How about adding another layer to display all the points?</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ER_status</span>, y <span class="op">=</span> <span class="va">GATA3</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="week2_files/figure-html/box_plot_2-1.png" width="700"></p>
<p>Ideally, we’d like these points to be spread out a bit. The
<code>geom_point</code> help points to <code>geom_jitter</code> as more
suitable when one of the variables is categorical.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ER_status</span>, y <span class="op">=</span> <span class="va">GATA3</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html" class="external-link">geom_jitter</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="week2_files/figure-html/box_plot_3-1.png" width="700"></p>
<p>Well, that’s a bit of a mess. We need to reduce the spread or jitter
and make the points smaller and transparent.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ER_status</span>, y <span class="op">=</span> <span class="va">GATA3</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html" class="external-link">geom_jitter</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">0.3</span>, size <span class="op">=</span> <span class="fl">0.5</span>, alpha <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span></code></pre></div>
<p><img src="week2_files/figure-html/box_plot_4-1.png" width="700"></p>
<p>Displaying points in this way makes much more sense when we only have
a few observations and where the box plot masks the fact, perhaps giving
the false impression that the sample size is larger than it actually is.
Here it makes less sense as we have very many observations.</p>
<p>Let’s try a colour aesthetic to also look at how estrogen receptor
expression differs between HER2 positive and negative tumours.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ER_status</span>, y <span class="op">=</span> <span class="va">GATA3</span>, colour <span class="op">=</span> <span class="va">HER2_status</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="week2_files/figure-html/box_plot_5-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="histograms">Histograms<a class="anchor" aria-label="anchor" href="#histograms"></a>
</h2>
<p>The geom for creating histograms is, rather unsurprisingly,
<code><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram()</a></code>.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Age_at_diagnosis</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></code></pre>
<p><img src="week2_files/figure-html/histogram_plot_1-1.png" width="700"></p>
<p>The warning message hints at picking a more optimal number of bins by
specifying the <code>binwidth</code> argument.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Age_at_diagnosis</span><span class="op">)</span>, binwidth <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></code></pre></div>
<p><img src="week2_files/figure-html/histogram_plot_2-1.png" width="700"></p>
<p>Or we can set the number of bins.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Age_at_diagnosis</span><span class="op">)</span>, bins <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></code></pre></div>
<p><img src="week2_files/figure-html/histogram_plot_3-1.png" width="700"></p>
<p>These histograms are not very pleasing, aesthetically speaking - how
about some better aesthetics?</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Age_at_diagnosis</span><span class="op">)</span>, bins <span class="op">=</span> <span class="fl">20</span>, colour <span class="op">=</span> <span class="st">"darkblue"</span>, fill <span class="op">=</span> <span class="st">"grey"</span><span class="op">)</span></code></pre></div>
<p><img src="week2_files/figure-html/histogram_plot_4-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="categorical-variables-factors">Categorical variables – factors<a class="anchor" aria-label="anchor" href="#categorical-variables-factors"></a>
</h2>
<p>Several of the variables in the METABRIC data set are categorical,
i.e. they can take one of a limited set of values. Some of these have
been read into R as character types (e.g. the 3-gene classifier), other
as numerical values, e.g. tumour stage. We also have some binary
variables that are essentially categorical variables but with only 2
possible values, e.g. ER status.</p>
<p>In many of the plots given above, ggplot2 has treated character
variables as categorical in situtations where a categorical variable is
expected. For example, when we displayed points on a scatter plot using
different colours for each 3-gene classification, or when we created
separate box plots in the same graph for ER positive and negative
patients.</p>
<p>But what about when our categorical variable has been read into R as
a continuous variable, e.g. <code>Tumour_stage</code>, which is read in
as a double type.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span>, colour <span class="op">=</span> <span class="va">Tumour_stage</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="week2_files/figure-html/scatter_plot_9-1.png" width="700"></p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">metabric</span><span class="op">$</span><span class="va">Tumour_stage</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">##   0   1   2   3   4 </span>
<span class="co">##   4 475 800 115   9</span></code></pre>
<p>Tumour stage has only 5 discrete states but ggplot2 doesn’t know
these are supposed to be a restricted set of values and has used a
colour scale to show them as if they were continuous. We need to tell R
that these are categorical.</p>
<p>R has a special type for categorical variables called a
<strong>factor</strong>. You can read more about factors <a href="https://r4ds.had.co.nz/factors.html" class="external-link">here</a>. The other R main
data types are numeric (numbers), character (text) and logical
(TRUE/FALSE), which you can read more about in the R for Data Science
section on vectors we mentioned before <a href="https://r4ds.had.co.nz/vectors.html?#important-types-of-atomic-vector" class="external-link">here</a>.
Let’s convert our tumour stage variable to a factor using the
<code><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor()</a></code> function. You should already have seen
<code><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select()</a></code> in the Introduction to Tidyverse course and we
will discuss it in more detail next week.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">metabric</span><span class="op">$</span><span class="va">Tumour_stage</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">metabric</span><span class="op">$</span><span class="va">Tumour_stage</span><span class="op">)</span>
<span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">metabric</span>, <span class="va">Patient_ID</span>, <span class="va">Tumour_stage</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 1,904 × 2</span></span>
<span class="co">##    Patient_ID Tumour_stage</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>       </span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> MB-0000    2           </span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> MB-0002    1           </span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> MB-0005    2           </span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> MB-0006    2           </span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> MB-0008    2           </span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> MB-0010    4           </span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> MB-0014    2           </span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> MB-0022    2           </span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> MB-0028    2           </span>
<span class="co">## <span style="color: #BCBCBC;">10</span> MB-0035    2           </span>
<span class="co">## <span style="color: #949494;"># … with 1,894 more rows</span></span></code></pre>
<p>R actually stores categorical variables as integers but with some
additional metadata about which of the integer values, or ‘levels’,
corresponds to each category.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/typeof.html" class="external-link">typeof</a></span><span class="op">(</span><span class="va">metabric</span><span class="op">$</span><span class="va">Tumour_stage</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "integer"</span></code></pre>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">metabric</span><span class="op">$</span><span class="va">Tumour_stage</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "factor"</span></code></pre>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">metabric</span><span class="op">$</span><span class="va">Tumour_stage</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "0" "1" "2" "3" "4"</span></code></pre>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">GATA3</span>, y <span class="op">=</span> <span class="va">ESR1</span>, colour <span class="op">=</span> <span class="va">Tumour_stage</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="week2_files/figure-html/scatter_plot_10-1.png" width="700"></p>
<p>In this case the order of the levels makes sense but for other
variables you may wish for more control over the ordering. Take the
integrative cluster variable for example. We created a bar plot of the
numbers of patients in the METABRIC cohort within each integrative
cluster. Did you notice the ordering of the clusters? 10 came just after
1 and before 2. That looked a bit odd as we’d have naturally expected it
to come last of all. R, on the other hand, is treating this vector as a
character vector (mainly because of the ‘ER-’ and ‘ER+’ subtypes of
cluster 4, and sorts the values into alphanumerical order.</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">metabric</span><span class="op">$</span><span class="va">Integrative_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">metabric</span><span class="op">$</span><span class="va">Integrative_cluster</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">metabric</span><span class="op">$</span><span class="va">Integrative_cluster</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##  [1] "1"    "10"   "2"    "3"    "4ER-" "4ER+" "5"    "6"    "7"    "8"   </span>
<span class="co">## [11] "9"</span></code></pre>
<p>A look at the help page for <code>as.factor</code> shows that we can
create a factor using the <code><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor()</a></code> function and specify the
levels using the <code>levels</code> argument.</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">metabric</span><span class="op">$</span><span class="va">Integrative_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">metabric</span><span class="op">$</span><span class="va">Integrative_cluster</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1"</span>, <span class="st">"2"</span>, <span class="st">"3"</span>, <span class="st">"4ER-"</span>, <span class="st">"4ER+"</span>, <span class="st">"5"</span>, <span class="st">"6"</span>, <span class="st">"7"</span>, <span class="st">"8"</span>, <span class="st">"9"</span>, <span class="st">"10"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">metabric</span><span class="op">$</span><span class="va">Integrative_cluster</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##  [1] "1"    "2"    "3"    "4ER-" "4ER+" "5"    "6"    "7"    "8"    "9"   </span>
<span class="co">## [11] "10"</span></code></pre>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Integrative_cluster</span>, fill <span class="op">=</span> <span class="va">Integrative_cluster</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="week2_files/figure-html/bar_plot_9-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="saving-plot-images">Saving plot images<a class="anchor" aria-label="anchor" href="#saving-plot-images"></a>
</h2>
<p>Use <code><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave()</a></code> to save the last plot you displayed.</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="st">"integrative_cluster.png"</span><span class="op">)</span></code></pre></div>
<p>You can alter the width and height of the plot and can change the
image file type.</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="st">"integrative_cluster.pdf"</span>, width <span class="op">=</span> <span class="fl">20</span>, height <span class="op">=</span> <span class="fl">12</span>, units <span class="op">=</span> <span class="st">"cm"</span><span class="op">)</span></code></pre></div>
<p>You can also pass in a plot object you have created instead of using
the last plot displayed. See the help page (<code><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">?ggsave</a></code>) for
more details.</p>
</div>
<div class="section level2">
<h2 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<p>In this session we have covered the following concepts:</p>
<ul>
<li>How to use ggplot2 to create different types of plot using a common
grammar</li>
<li>How to map variables to aesthetic features of a plot</li>
<li>How to construct a plot with multiple layers (geoms)</li>
<li>Various types of plot including scatter plots, box plots, bar plots
and histograms</li>
<li>Categorical variables and how R handles these with its factor
type</li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 id="assignment">Assignment<a class="anchor" aria-label="anchor" href="#assignment"></a>
</h2>
<p><a href="data:text/x-markdown;base64,Ci0tLQp0aXRsZTogIkFzc2lnbm1lbnQgMiIKLS0tCgpUaGUgd2VlaydzIGFzc2lnbm1lbnQgd2lsbCB0ZXN0IHlvdXIgYWJpbGl0eSB0byBjcmVhdGUgdmFyaW91cyBkaWZmZXJlbnQgdHlwZXMKb2YgcGxvdC4gSXQgaW5jbHVkZXMgc29tZSBleHRlbnNpb24gZXhlcmNpc2VzIHdoaWNoIHlvdSBjYW4gY29uc2lkZXIgb3B0aW9uYWwuCgpgYGB7ciBldmFsID0gRkFMU0UsIG1lc3NhZ2UgPSBGQUxTRSwgd2FybmluZyA9IEZBTFNFfQpsaWJyYXJ5KHRpZHl2ZXJzZSkKbWV0YWJyaWMgPC0gcmVhZF9jc3YoImh0dHBzOi8vemVub2RvLm9yZy9yZWNvcmQvNjQ1MDE0NC9maWxlcy9tZXRhYnJpY19jbGluaWNhbF9hbmRfZXhwcmVzc2lvbl9kYXRhLmNzdiIpCmBgYAoKLS0tCgoqKjEuIENyZWF0ZSBhIGJhciBwbG90IHNob3dpbmcgdGhlIG51bWJlciBvZiBwYXRpZW50cyB3aXRoIHR1bW91cnMgYXQgZWFjaCBvZiB0aGUgZGlmZmVyZW50IG5lb3BsYXNtIGhpc3RvbG9naWMgZ3JhZGVzKioKCmBgYHtyfQoKYGBgCgpUdXJuIHRoaXMgaW50byBhIHN0YWNrZWQgYmFyIHBsb3Qgc2hvd2luZyB0aGUgcHJvcG9ydGlvbnMgb2YgcGF0aWVudHMgd2l0aCBlYWNoCmhpc3RvbG9naWMgZ3JhZGUgdGhhdCBoYWQgY2hlbW90aGVyYXB5CgpgYGB7cn0KCmBgYAoKKipFeHRlbnNpb24qKgoKZ2dwbG90IGdpdmVzIGEgd2FybmluZyBhYm91dCBub24tZmluaXRlIHZhbHVlcywgaS5lLiB0aGUgbWlzc2luZyB2YWx1ZXMuIEZpbmQKb3V0IGhvdyB0byBnZXQgcmlkIG9mIHRoaXMgd2FybmluZy4KCmBgYHtyfQoKYGBgCgpMZXQncyBzYXkgd2UgYXJlIGludGVyZXN0ZWQgaW4gZGlzcGxheWluZyB0aGUgbWlzc2luZyB2YWx1ZXMgYXMgYSBzZXBhcmF0ZQpjYXRlZ29yeSBhbmQgcmVjcmVhdGUgdGhlIGJhciBwbG90IHdpdGggYW4gYWRkaXRpb25hbCBjb2x1bW4uCgpgYGB7cn0KCmBgYAoKLS0tCgoqKjIuIENyZWF0ZSBhIGRlbnNpdHkgcGxvdCBvZiB0aGUgZXN0cm9nZW4gcmVjZXB0b3IgKEVTUjEpIGV4cHJlc3Npb24gbGV2ZWwqKgoKYGBge3J9CgpgYGAKCl9IaW50OiB0aGlzIGlzIGEgbmV3IHBsb3QgdHlwZSAtLSBsb29rIHVwIHRoZSBnZW9tIHlvdSBuZWVkIGJ5IGxvb2tpbmcgYXQgdGhlCmxpc3Qgb2YgZnVuY3Rpb25zIGZvciB0aGUgZ2dwbG90MiBwYWNrYWdlIGluIHRoZSBQYWNrYWdlcyB0YWIgaW4gUlN0dWRpby5fCgpDcmVhdGUgc2VwYXJhdGUgYnV0IG92ZXJsYXlpbmcgZGVuc2l0eSBwbG90cyBmb3IgRVNSMSBleHByZXNzaW9uIGZvciBIRVIyCnBvc2l0aXZlIGFuZCBuZWdhdGl2ZSBwYXRpZW50cy4gVHJ5IHRvIGVuc3VyZSB0aGF0IGJvdGggZGVuc2l0eSBwbG90cyBhcmUKdmlzaWJsZS4KCmBgYHtyfQoKYGBgCgotLS0KCioqMy4gQ3JlYXRlIGJveCBwbG90cyBmb3IgdGhlIHN1cnZpdmFsIHRpbWUgb2YgcGF0aWVudHMgd2l0aCBkaWZmZXJlbnQgdHVtb3VyIHN0YWdlcyBvbmx5IGluY2x1ZGluZyBkZWNlYXNlZCBwYXRpZW50cyoqCgpgYGB7cn0KCmBgYAoKX0hpbnQ6IGlmIHlvdSBnZXQgYSB3YXJuaW5nIG1lc3NhZ2UgYW5kIG9ubHkgaGF2ZSBhIHNpbmdsZSBib3ggZGlzcGxheWVkCmNvbnNpZGVyIHRoZSB0eXBlIG9mIHRoZSB0dW1vdXIgc3RhZ2UgdmFyaWFibGUgYXMgcmVhZCBpbnRvIFIgYW5kIHdoYXQga2luZCBvZgp2YXJpYWJsZSBpcyBleHBlY3RlZCBmb3IgdGhlIHggYWVzdGhldGljLl8KCkNvbnZlcnQgdGhpcyBpbnRvIGEgdmlvbGluIHBsb3QgYW5kIGRpc3BsYXkgRVIgbmVnYXRpdmUgYW5kIEVSIHBvc2l0aXZlIHBhdGllbnRzCnNlcGFyYXRlbHkuCgpgYGB7cn0KCmBgYAoKX0hpbnQ6IGxvb2sgYXQgdGhlIGhlbHAgZm9yIGdlb21fYm94cGxvdC5fCgotLS0KCioqNC4gRXhwbG9yZSB0aGUgY29ycmVsYXRpb24gYmV0d2VlbiBGT1hBMSBhbmQgTUxQSCoqCgpBIHJlY2VudGx5LXB1Ymxpc2hlZCBwYXBlciBsb29rcyBhdCBhc3NvY2lhdGlvbnMgb2YgZXhwcmVzc2lvbiBhbmQgZ2Vub21pYyBkYXRhCmluIHRoZSBNRVRBQlJJQyBkYXRhIHNldC4gSXQgZm91bmQgYSBzdHJvbmcgY29ycmVsYXRpb24gYmV0d2VlbiBGT1hBMSBhbmQgb25lIG9mIHRoZSBvdGhlciBnZW5lcyBpbmNsdWRlZCBpbiBvdXIgdGFibGUsIE1MUEggKHNlZSB0aGUgb3V0cHV0IG9mIGBjb3IudGVzdChtZXRhYnJpYyRGT1hBMSwgIG1ldGFicmljJE1MUEgpYCkuCgoKQ3JlYXRlIGEgc2NhdHRlciBwbG90IG9mIHRoZSBleHByZXNzaW9uIG9mIEZPWEExIGFuZCBNTFBIIHdpdGggYSBsaW5lIG9mIGJlc3QgZml0LgoKYGBge3J9CgpgYGAKCkNoYW5nZSB0aGUgbWV0aG9kIHVzZWQgdG8gZHJhdyB0aGUgbGluZSBvZiBiZXN0IGZpdCB0byBiZSBhIHN0cmFpZ2h0IGxpbmUuCgpgYGB7cn0KCmBgYAoKCi0tLQoKKipFWFRSQS4gSWYgeW91IGRpZCB0aGUgb3B0aW9uYWwgRVhUUkEgcGFydCBpbiB0aGUgbWF0ZXJpYWwuIEV4cGxvcmUgdGhlIEFwcGxlIE1vYmlsaXR5IGRhdGEgb24gdHJhbnNwb3J0IHVzZSBkdXJpbmcgdGhlIGNvcm9uYXZpcnVzIGxvY2tkb3duIHVzaW5nIGxpbmUgcGxvdHMqKgoKVGhlIEdvdmVybm1lbnQgaW5jbHVkZWQgaW4gdGhlaXIgZGFpbHkgcHJlc3MgY29uZmVyZW5jZSBhIG5ldyBzZXQKb2YgZGF0YSBvZiB0cmFuc3BvcnQgdXNlIGFzIG1lYXN1cmVkIGJ5IGRpcmVjdGlvbiByZXF1ZXN0cyBvbiBBcHBsZSBNYXBzLgoKU2VhcmNoIGZvciBhbmQgZG93bmxvYWQgdGhlIFVLIEdvdmVybm1lbnQgZGF0YSBmb3IgdGhlIFN1bmRheSBicmllZmluZyAoMjYgQXByaWwKMjAyMCkuCgpfSGludDogdXNlIHRoZSBzZWFyY2ggdGVybXMgInVrIGdvdmVybm1lbnQgY29yb25hdmlydXMgcHJlc3MgY29uZmVyZW5jZSBkYXRhIi5fCgpMb2FkIHRoZSAnQXBwbGUgTW9iaWxpdHknIHNoZWV0IGludG8gUi4gWW91IHdpbGwgbmVlZCB0byBzcGVjaWZ5IHRoZQp3b3Jrc2hlZXQgYW5kIGEgZmV3IG90aGVyIGFyZ3VtZW50cyBhcyB0aGUgZGF0YSBhcmUgbm90IGluIHRoZSBtb3N0IGNvbnZlbmllbnQKZm9ybWF0IChzZWUgaG93IHRoZSB0cmFuc3BvcnQgdXNlIHdvcmtzaGVldCB3YXMgcmVhZCBpbnRvIFIgaW4gdGhlIGNvdXJzZQptYXRlcmlhbHMpLgoKYGBge3J9CgpgYGAKClBsb3QgdGhlIDctZGF5IHJvbGxpbmcgYXZlcmFnZSBhZ2FpbnN0IGRhdGUgYXMgYSBsaW5lIGdyYXBoIHdpdGggc2VwYXJhdGUgbGluZXMKZm9yIGVhY2ggbW9kZSBvZiB0cmFuc3BvcnQuCgpgYGB7cn0KCmBgYAoKV2hhdCBpcyB0aGUgbWFpbiAoYW5kIG9idmlvdXMpIGNvbmNsdXNpb24gdG8gZHJhdyBmcm9tIHRoaXMgcGxvdD8KCk5vdyBwbG90IHRoZSBkYWlseSBwZXJjZW50YWdlIGluc3RlYWQgb2YgdGhlIDctZGF5IHJvbGxpbmcgYXZlcmFnZS4KCmBgYHtyfQoKYGBgCgpXaGF0IGRvIHlvdSBub3RpY2UgYWJvdXQgZGFpbHkgZmx1Y3R1YXRpb25zPwoKKipFeHRlbnNpb24qKgoKQ3JlYXRlIGEgbmV3IGNvbHVtbiBmb3IgdGhlIHdlZWtkYXkgKHRha2luZyB2YWx1ZXMgU3VuZGF5LCBNb25kYXksIFR1ZXNkYXksCmV0Yy4pIGluIHRoZSBBcHBsZSBNb2JpbGl0eSB0YWJsZSBhbmQgcGxvdCBqdXN0IHRoZSBkYWlseSBwZXJjZW50YWdlIGZvciB3YWxraW5nCmJ1dCBub3cgYXMgc2VwYXJhdGUgbGluZXMgZm9yIGVhY2ggZGF5IG9mIHRoZSB3ZWVrLgoKYGBge3J9CgpgYGAKCl9IaW50OiB1c2UgYSBmdW5jdGlvbiBmcm9tIHRoZSBsdWJyaWRhdGUgcGFja2FnZSwgYW5vdGhlciBvZiB0aGUgdGlkeXZlcnNlCnBhY2thZ2VzIChpbnN0YWxsIGlmIG5lY2Vzc2FyeSkuXwo=" download="assignment2.Rmd">Download assignment2.Rmd</a></p>
<hr>
</div>
<div class="section level2">
<h2 id="time-series-data-extra">Time series data (EXTRA)<a class="anchor" aria-label="anchor" href="#time-series-data-extra"></a>
</h2>
<p>Optional section. Do this if you’ve completed the previous sections
and the Assignment and want more. In this section we’re going to look a
<strong>line plot</strong>. These are created using similar ggplot
grammar to what we’ve already seen. Line plots are a common
visualization for time series data but we don’t have data of this kind
in the METABRIC data set so we’ll turn our attention to a more topical
data set.</p>
<div class="section level3">
<h3 id="covid-19-data-set">COVID-19 data set<a class="anchor" aria-label="anchor" href="#covid-19-data-set"></a>
</h3>
<p>Every day the UK Government presents a series of plots showing data
relating to the coronavirus (COVID-19) pandemic. Both the plots and the
data are published on the Government’s <a href="https://www.gov.uk/government/collections/slides-and-datasets-to-accompany-coronavirus-press-conferences" class="external-link">website</a>.</p>
<p>One of the plots shows the reduction in different types of transport
use over time in the UK in the week leading up to and since the lockdown
restrictions were introduced in response to the pandemic.</p>
<p>Reading these data into R from the downloaded Excel spreadsheet
requires a bit more care as the spreadsheet contains multiple sheets and
each table has header lines and footnotes. The date column doesn’t have
a column heading, so in the code below we’ve skipped that row as well
and told <code><a href="https://readxl.tidyverse.org/reference/read_excel.html" class="external-link">read_excel()</a></code> what column names to use. We’ve also
given it a heads-up about which type of data are in each column. Look up
the help page for <code><a href="https://readxl.tidyverse.org/reference/read_excel.html" class="external-link">read_excel()</a></code> for more details about the
arguments we’ve used below.</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://readxl.tidyverse.org" class="external-link">readxl</a></span><span class="op">)</span>
<span class="co"># because it's an Excel file located on the web need to download then import</span>
<span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://zenodo.org/record/6450144/files/2020-04-26_COVID-19_Press_Conference_Data.xlsx"</span>, destfile <span class="op">=</span> <span class="st">"2020-04-26_COVID-19_Press_Conference_Data.xlsx"</span><span class="op">)</span>
<span class="va">transport_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readxl.tidyverse.org/reference/read_excel.html" class="external-link">read_excel</a></span><span class="op">(</span><span class="st">"2020-04-26_COVID-19_Press_Conference_Data.xlsx"</span>,
                            sheet <span class="op">=</span> <span class="st">"Transport use"</span>,
                            col_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"date"</span>, <span class="st">"transport_type"</span>, <span class="st">"percentage"</span><span class="op">)</span>,
                            col_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"date"</span>, <span class="st">"text"</span>, <span class="st">"numeric"</span><span class="op">)</span>,
                            skip <span class="op">=</span> <span class="fl">4</span>,
                            n_max <span class="op">=</span> <span class="fl">186</span><span class="op">)</span>
<span class="va">transport_use</span><span class="op">$</span><span class="va">percentage</span> <span class="op">&lt;-</span> <span class="va">transport_use</span><span class="op">$</span><span class="va">percentage</span> <span class="op">*</span> <span class="fl">100</span>
<span class="va">transport_use</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 186 × 3</span></span>
<span class="co">##    date                transport_type     percentage</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;dttm&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> 2020-03-16 <span style="color: #949494;">00:00:00</span> All motor vehicles         98</span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> 2020-03-17 <span style="color: #949494;">00:00:00</span> All motor vehicles         89</span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> 2020-03-18 <span style="color: #949494;">00:00:00</span> All motor vehicles         83</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> 2020-03-19 <span style="color: #949494;">00:00:00</span> All motor vehicles         82</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> 2020-03-20 <span style="color: #949494;">00:00:00</span> All motor vehicles         81</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> 2020-03-21 <span style="color: #949494;">00:00:00</span> All motor vehicles         73</span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> 2020-03-22 <span style="color: #949494;">00:00:00</span> All motor vehicles         69</span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> 2020-03-23 <span style="color: #949494;">00:00:00</span> All motor vehicles         69</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> 2020-03-24 <span style="color: #949494;">00:00:00</span> All motor vehicles         49</span>
<span class="co">## <span style="color: #BCBCBC;">10</span> 2020-03-25 <span style="color: #949494;">00:00:00</span> All motor vehicles         42</span>
<span class="co">## <span style="color: #949494;"># … with 176 more rows</span></span></code></pre>
<p>In this data set each row represents an observation of the current
traffic volume for various modes of transport (motor vehicle, bus,
train, etc.) since 16 March as a percentage of the volume measured
during the first week of February.</p>
<p>So let’s get a feel for this data set. Firstly, what modes of
transport are included and how many observations do we have for
each?</p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">transport_use</span><span class="op">$</span><span class="va">transport_type</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">## All motor vehicles Bus (excl. London)      National rail          TfL - Bus </span>
<span class="co">##                 40                 40                 40                 34 </span>
<span class="co">##         TfL - Tube </span>
<span class="co">##                 32</span></code></pre>
<p>We’ll focus solely on rail use to start with. You should already have
seen <code><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter()</a></code> in the Introduction to Tidyverse course and
we will discuss it in more detail next week.</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">rail_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">transport_use</span>, <span class="va">transport_type</span> <span class="op">==</span> <span class="st">"National rail"</span><span class="op">)</span>

<span class="va">rail_use</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 40 × 3</span></span>
<span class="co">##    date                transport_type percentage</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;dttm&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> 2020-03-16 <span style="color: #949494;">00:00:00</span> National rail          77</span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> 2020-03-17 <span style="color: #949494;">00:00:00</span> National rail          69</span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> 2020-03-18 <span style="color: #949494;">00:00:00</span> National rail          59</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> 2020-03-19 <span style="color: #949494;">00:00:00</span> National rail          52</span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> 2020-03-20 <span style="color: #949494;">00:00:00</span> National rail          44</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> 2020-03-21 <span style="color: #949494;">00:00:00</span> National rail          38</span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> 2020-03-22 <span style="color: #949494;">00:00:00</span> National rail          35</span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> 2020-03-23 <span style="color: #949494;">00:00:00</span> National rail          29</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> 2020-03-24 <span style="color: #949494;">00:00:00</span> National rail          24</span>
<span class="co">## <span style="color: #BCBCBC;">10</span> 2020-03-25 <span style="color: #949494;">00:00:00</span> National rail          18</span>
<span class="co">## <span style="color: #949494;"># … with 30 more rows</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="dates">Dates<a class="anchor" aria-label="anchor" href="#dates"></a>
</h3>
<p>There is something unfamiliar about the <code>date</code> column that
needs some explanation. Let’s check the type and class of
<code>date</code>.</p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/typeof.html" class="external-link">typeof</a></span><span class="op">(</span><span class="va">rail_use</span><span class="op">$</span><span class="va">date</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "double"</span></code></pre>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">rail_use</span><span class="op">$</span><span class="va">date</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "POSIXct" "POSIXt"</span></code></pre>
<p>So the <code>date</code> column is stored as a double and is a
suitable continuous variable for a line plot but it’s also got some
special properties, afforded to it by its <code>POSIXct</code> class,
that allows it to be treated as a date.</p>
<p>The <code>POSIXct</code> class is how R represents dates, where POSIX
is a standardization scheme in the world of computers, spanning multiple
operating systems and programming languages, and the <code>ct</code>
part stands for calendar time. We won’t go into the details here but
note that the numeric representation of a date, i.e. the
<code>double</code> value, is the number of seconds since 1 January
1970. For example, the date in our first row is:</p>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">rail_use</span><span class="op">$</span><span class="va">date</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">## [1] "2020-03-16 UTC"</span></code></pre>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/double.html" class="external-link">as.double</a></span><span class="op">(</span><span class="va">rail_use</span><span class="op">$</span><span class="va">date</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 1584316800</span></code></pre>
</div>
<div class="section level3">
<h3 id="line-plot">Line plot<a class="anchor" aria-label="anchor" href="#line-plot"></a>
</h3>
<p>We’re going to plot the volume of traffic on each date with each
measurement represented as a point on our graph and observations on
successive dates connected by a line. Looking at the geoms available in
ggplot2 (see the list of functions available for ggplot2 by navigating
to the ggplot2 package in the Packages tab in RStudio), it looks like
<strong><code>geom_line</code></strong> is the one we need.</p>
<p>The aesthetics section in the help page for <code>geom_line</code>
tells us we need an x and y mapping. Our x aesthetic will be
<code>date</code> and the y aesthetic will be
<code>percentage</code>.</p>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">rail_use</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">date</span>, y <span class="op">=</span> <span class="va">percentage</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="week2_files/figure-html/line_plot_1-1.png" width="700"></p>
<p>We can return to our original <code>transport_use</code> table
containing data for all transport modes and use another aesthetic,
e.g. colour, to plot a line for each.</p>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">transport_use</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">date</span>, y <span class="op">=</span> <span class="va">percentage</span>, colour <span class="op">=</span> <span class="va">transport_type</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="week2_files/figure-html/line_plot_2-1.png" width="700"></p>
<p>Another aesthetic available for <code>geom_line</code> is
<code>linetype</code>.</p>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">transport_use</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">date</span>, y <span class="op">=</span> <span class="va">percentage</span>, linetype <span class="op">=</span> <span class="va">transport_type</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="week2_files/figure-html/line_plot_3-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="the-ggplot-object-a-peek-under-the-hood-advanced">The ggplot object – a peek under the hood (ADVANCED)<a class="anchor" aria-label="anchor" href="#the-ggplot-object-a-peek-under-the-hood-advanced"></a>
</h2>
<p>Optional section. Do this if you’d like to understand more about how
ggplot works.</p>
<p>Let’s build a ggplot2 plot up in stages to understand what’s really
going on.</p>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span><span class="op">)</span></code></pre></div>
<p>What is the plot object we’ve just created?</p>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/typeof.html" class="external-link">typeof</a></span><span class="op">(</span><span class="va">plot</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "list"</span></code></pre>
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">plot</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 9</span></code></pre>
<p>The plot object is a list containing 9 elements. It’s actually a
special type of list.</p>
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">plot</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "gg"     "ggplot"</span></code></pre>
<p>What are the 9 things in the list?</p>
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">plot</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "data"        "layers"      "scales"      "mapping"     "theme"      </span>
<span class="co">## [6] "coordinates" "facet"       "plot_env"    "labels"</span></code></pre>
<p>So far we’ve constructed ggplots from data, layers (geoms) and
aesthetic mappings but do you notice the labels, scales, facet and theme
elements? We’ll be looking at how to customize plots later, including
changing the labels, altering the scales of axes, and applying a
different visual theme.</p>
<p>The <code>data</code> element is just the metabric tibble that we
provided to the <code><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot()</a></code> function.</p>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plot</span><span class="op">$</span><span class="va">data</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 1,904 × 32</span></span>
<span class="co">##    Patient_ID Cohort Age_at_diagnosis Survival_time Survival_status Vital_status</span>
<span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       </span>
<span class="co">## <span style="color: #BCBCBC;"> 1</span> MB-0000         1             75.6         140.  LIVING          Living      </span>
<span class="co">## <span style="color: #BCBCBC;"> 2</span> MB-0002         1             43.2          84.6 LIVING          Living      </span>
<span class="co">## <span style="color: #BCBCBC;"> 3</span> MB-0005         1             48.9         164.  DECEASED        Died of Dis…</span>
<span class="co">## <span style="color: #BCBCBC;"> 4</span> MB-0006         1             47.7         165.  LIVING          Living      </span>
<span class="co">## <span style="color: #BCBCBC;"> 5</span> MB-0008         1             77.0          41.4 DECEASED        Died of Dis…</span>
<span class="co">## <span style="color: #BCBCBC;"> 6</span> MB-0010         1             78.8           7.8 DECEASED        Died of Dis…</span>
<span class="co">## <span style="color: #BCBCBC;"> 7</span> MB-0014         1             56.4         164.  LIVING          Living      </span>
<span class="co">## <span style="color: #BCBCBC;"> 8</span> MB-0022         1             89.1          99.5 DECEASED        Died of Oth…</span>
<span class="co">## <span style="color: #BCBCBC;"> 9</span> MB-0028         1             86.4          36.6 DECEASED        Died of Oth…</span>
<span class="co">## <span style="color: #BCBCBC;">10</span> MB-0035         1             84.2          36.3 DECEASED        Died of Dis…</span>
<span class="co">## <span style="color: #949494;"># … with 1,894 more rows, and 26 more variables: Chemotherapy &lt;chr&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   Radiotherapy &lt;chr&gt;, Tumour_size &lt;dbl&gt;, Tumour_stage &lt;fct&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   Neoplasm_histologic_grade &lt;dbl&gt;, Lymph_nodes_examined_positive &lt;dbl&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   Lymph_node_status &lt;dbl&gt;, Cancer_type &lt;chr&gt;, ER_status &lt;chr&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   PR_status &lt;chr&gt;, HER2_status &lt;chr&gt;, HER2_status_measured_by_SNP6 &lt;chr&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   PAM50 &lt;chr&gt;, `3-gene_classifier` &lt;chr&gt;, Nottingham_prognostic_index &lt;dbl&gt;,</span></span>
<span class="co">## <span style="color: #949494;">#   Cellularity &lt;chr&gt;, Integrative_cluster &lt;fct&gt;, Mutation_count &lt;dbl&gt;, …</span></span></code></pre>
<p>This <code>plot</code> list object is really only a specification for
how ggplot2 should create the plot. ggplot2 only renders the plot when
we ask it to by printing it out to the screen by typing
<code>print(plot)</code> or simply just <code>plot</code>.</p>
<p>We haven’t added any layers yet so what does our plot look like at
this point?</p>
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plot</span></code></pre></div>
<p><img src="week2_files/figure-html/building_up_plot_1-1.png" width="700"></p>
<p>ggplot2 doesn’t know what to plot yet as we haven’t added a layer
(geom).</p>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plot</span><span class="op">$</span><span class="va">mapping</span></code></pre></div>
<pre><code><span class="co">## Aesthetic mapping: </span>
<span class="co">## &lt;empty&gt;</span></code></pre>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plot</span><span class="op">$</span><span class="va">layers</span></code></pre></div>
<pre><code><span class="co">## list()</span></code></pre>
<p>Lets add an aesthetic mapping.</p>
<div class="sourceCode" id="cb102"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">metabric</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Nottingham_prognostic_index</span>, y <span class="op">=</span> <span class="va">ESR1</span>, colour <span class="op">=</span> <span class="va">ER_status</span><span class="op">)</span><span class="op">)</span>
<span class="va">plot</span></code></pre></div>
<p><img src="week2_files/figure-html/building_up_plot_2-1.png" width="700"></p>
<p>ggplot2 has automatically added scales for x and y based on the
ranges of values for the Nottingham prognostic index and ESR1
expression. Still nothing has been plotted as we haven’t yet specified
the type of plot (geom) to add as a layer.</p>
<div class="sourceCode" id="cb103"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plot</span><span class="op">$</span><span class="va">mapping</span></code></pre></div>
<pre><code><span class="co">## Aesthetic mapping: </span>
<span class="co">## * `x`      -&gt; `Nottingham_prognostic_index`</span>
<span class="co">## * `y`      -&gt; `ESR1`</span>
<span class="co">## * `colour` -&gt; `ER_status`</span></code></pre>
<div class="sourceCode" id="cb105"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plot</span><span class="op">$</span><span class="va">layers</span></code></pre></div>
<pre><code><span class="co">## list()</span></code></pre>
<p>Finally, let’s add a <code>geom_point</code> layer to create a
scatter plot.</p>
<div class="sourceCode" id="cb107"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plot</span> <span class="op">&lt;-</span> <span class="va">plot</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">plot</span></code></pre></div>
<p><img src="week2_files/figure-html/building_up_plot_3-1.png" width="700"></p>
<div class="sourceCode" id="cb108"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plot</span><span class="op">$</span><span class="va">layers</span></code></pre></div>
<pre><code><span class="co">## [[1]]</span>
<span class="co">## geom_point: na.rm = FALSE</span>
<span class="co">## stat_identity: na.rm = FALSE</span>
<span class="co">## position_identity</span></code></pre>
<p>We touched on statistical transformations earlier. The
<code>stat</code> associated with a <code>geom_point</code> is
<code>stat_identity</code> which leaves values unchanged. In the case of
a scatter plot, we already have the x and y values – they don’t need to
be transformed, just plotted on the x and y axes.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Maria Doyle.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
